{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import IPython\n",
    "\n",
    "import pygal\n",
    "from pygal.style import DarkStyle\n",
    "\n",
    "import importlib\n",
    "import svgwrite\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#B400FF', '#3232FF'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    "    \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "IPython.utils.text.SList"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = !head -1 ~/Dropbox/LiLab/RNAseq/CSV/zfUVvsS_all.csv\n",
    "gene = !grep opn1sw1 ~/Dropbox/LiLab/RNAseq/CSV/zfUVvsS_all.csv\n",
    "type(gene)\n",
    "# pd.DataFrame(data=data[0],columns=names[0])\n",
    "gdf=pd.DataFrame(data=gene)[0].str.split(',',expand=True)\n",
    "ndf=pd.DataFrame(data=names)[0].str.split(',',expand=True) \n",
    "ndf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>UV01</th>\n",
       "      <th>UV02</th>\n",
       "      <th>UV03</th>\n",
       "      <th>UV04</th>\n",
       "      <th>S01</th>\n",
       "      <th>S02</th>\n",
       "      <th>S03</th>\n",
       "      <th>S04</th>\n",
       "      <th>chr-rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gabrr3a</td>\n",
       "      <td>gamma-aminobutyric acid (GABA) A receptor, rho 3a</td>\n",
       "      <td>11.81</td>\n",
       "      <td>-0.1694</td>\n",
       "      <td>0.9730</td>\n",
       "      <td>0.614828</td>\n",
       "      <td>0.999194</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>5206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rho</td>\n",
       "      <td>rhodopsin</td>\n",
       "      <td>207561.67</td>\n",
       "      <td>-0.0154</td>\n",
       "      <td>0.7499</td>\n",
       "      <td>0.976022</td>\n",
       "      <td>0.999194</td>\n",
       "      <td>395257</td>\n",
       "      <td>106284</td>\n",
       "      <td>66731</td>\n",
       "      <td>252358</td>\n",
       "      <td>367166</td>\n",
       "      <td>93345</td>\n",
       "      <td>272102</td>\n",
       "      <td>107250</td>\n",
       "      <td>9938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gabrr2a</td>\n",
       "      <td>gamma-aminobutyric acid (GABA) A receptor, rho 2a</td>\n",
       "      <td>34.91</td>\n",
       "      <td>0.1464</td>\n",
       "      <td>0.9957</td>\n",
       "      <td>0.612222</td>\n",
       "      <td>0.999194</td>\n",
       "      <td>26</td>\n",
       "      <td>186</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>15603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gstr</td>\n",
       "      <td>glutathione S-transferase rho</td>\n",
       "      <td>281.96</td>\n",
       "      <td>0.4081</td>\n",
       "      <td>1.0452</td>\n",
       "      <td>0.303268</td>\n",
       "      <td>0.999194</td>\n",
       "      <td>889</td>\n",
       "      <td>426</td>\n",
       "      <td>440</td>\n",
       "      <td>94</td>\n",
       "      <td>48</td>\n",
       "      <td>357</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>26994</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    symbol                                           genename   baseMean  \\\n",
       "0  gabrr3a  gamma-aminobutyric acid (GABA) A receptor, rho 3a      11.81   \n",
       "1      rho                                          rhodopsin  207561.67   \n",
       "2  gabrr2a  gamma-aminobutyric acid (GABA) A receptor, rho 2a      34.91   \n",
       "3     gstr                      glutathione S-transferase rho     281.96   \n",
       "\n",
       "   log2FoldChange   lfcSE    pvalue      padj    UV01    UV02   UV03    UV04  \\\n",
       "0         -0.1694  0.9730  0.614828  0.999194       0       0     27       0   \n",
       "1         -0.0154  0.7499  0.976022  0.999194  395257  106284  66731  252358   \n",
       "2          0.1464  0.9957  0.612222  0.999194      26     186      0       0   \n",
       "3          0.4081  1.0452  0.303268  0.999194     889     426    440      94   \n",
       "\n",
       "      S01    S02     S03     S04  chr-rank  \n",
       "0       0     29      20      17      5206  \n",
       "1  367166  93345  272102  107250      9938  \n",
       "2       0      0      68       0     15603  \n",
       "3      48    357       0       0     26994  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "!head -1 ~/Dropbox/LiLab/RNAseq/CSV/zfUVvsS_all.csv > ~/Dropbox/LiLab/RNAseq/CSV/temp.txt\n",
    "!grep -w rho ~/Dropbox/LiLab/RNAseq/CSV/zfUVvsS_all.csv >> ~/Dropbox/LiLab/RNAseq/CSV/temp.txt\n",
    "gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/CSV/temp.txt\")\n",
    "gdf\n",
    "# \\\n",
    "\n",
    "# grep ^.*@.*fish.*$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='read counts (RNAseq)',\n",
    "    show_x_labels=False,\n",
    "    title='UV-opsin (opn1sw1)',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4,\n",
    "    show_legend=False\n",
    ")\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (1831890.775,0.2,0.8),\n",
    "    (1940418.738,1.2,1.8),\n",
    "    (1791047.718,2.2,2.8),\n",
    "    (2139963.138,3.2,3.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (253221.171,4.2,4.8),\n",
    "    (1152.782,5.2,5.8),\n",
    "    (39185.840,6.2,6.8),\n",
    "    (293059.944,7.2,7.8),\n",
    "])\n",
    "\n",
    "\n",
    "hist.render_to_file('opn1sw1.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='read counts (RNAseq)',\n",
    "    show_x_labels=False,\n",
    "    title='S-opsin (opn1sw2)',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4,\n",
    "    show_legend=False\n",
    ")\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (6178.852,0.2,0.8),\n",
    "    (6129,1.2,1.8),\n",
    "    (109067,2.2,2.8),\n",
    "    (1407,3.2,3.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (671153.171,4.2,4.8),\n",
    "    (740768.782,5.2,5.8),\n",
    "    (885780.840,6.2,6.8),\n",
    "    (847799.944,7.2,7.8),\n",
    "])\n",
    "\n",
    "\n",
    "hist.render_to_file('opn1sw2.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='read counts (RNAseq)',\n",
    "    show_x_labels=False,\n",
    "    title='rhodopsin (rho)',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4,\n",
    "    show_legend=False,\n",
    "    range=(0,800000)\n",
    ")\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (395257,0.2,0.8),\n",
    "    (106284,1.2,1.8),\n",
    "    (66731,2.2,2.8),\n",
    "    (252358,3.2,3.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (367167,4.2,4.8),\n",
    "    (93345,5.2,5.8),\n",
    "    (272102,6.2,6.8),\n",
    "    (107250,7.2,7.8),\n",
    "])\n",
    "\n",
    "\n",
    "hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='read counts (RNAseq)',\n",
    "    show_x_labels=False,\n",
    "    title='cone transducin (gnat2) vs. rod transducin (gnat1)',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4,\n",
    "    show_legend=False,\n",
    "#     range=(0,800000)\n",
    ")\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (340844,0.2,0.8),\n",
    "    (453770,1.2,1.8),\n",
    "    (434098,2.2,2.8),\n",
    "    (481402,3.2,3.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (446590,4.2,4.8),\n",
    "    (362607,5.2,5.8),\n",
    "    (318034,6.2,6.8),\n",
    "    (413598,7.2,7.8),\n",
    "])\n",
    "\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (115155,9.2,9.8),\n",
    "    (39631,10.2,10.8),\n",
    "    (46918,11.2,11.8),\n",
    "    (92691,12.2,12.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (167981,13.2,13.8),\n",
    "    (59011,14.2,14.8),\n",
    "    (97955,15.2,15.8),\n",
    "    (33217,16.2,16.8),\n",
    "])\n",
    "\n",
    "hist.render_to_file('gnats.svg')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='million reads',\n",
    "    show_x_labels=False,\n",
    "    title='',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4,\n",
    "    show_legend=False\n",
    ")\n",
    "\n",
    "hist.add('UV-Cone',[\n",
    "    (15.710882,0.2,0.8),\n",
    "    (10.873172,1.2,1.8),\n",
    "    (17.410638,2.2,2.8),\n",
    "    (15.763850,3.2,3.8),\n",
    "])\n",
    "hist.add('S-Cone',[\n",
    "    (15.390731,4.2,4.8),\n",
    "    (15.557118,5.2,5.8),\n",
    "    (15.617877,6.2,6.8),\n",
    "    (24.908813,7.2,7.8),\n",
    "])\n",
    "\n",
    "\n",
    "\n",
    "hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pygal.Histogram(\n",
    "    x_title='',\n",
    "    y_title='read counts (RNAseq)',\n",
    "    show_x_labels=False,\n",
    "    title='nrxn3',\n",
    "    style=custom_style,\n",
    "    legend_at_bottom=True,\n",
    "    legend_at_bottom_columns=4\n",
    ")\n",
    "\n",
    "hist.add('S-Cone',[\n",
    "    (2823,0.2,0.8),\n",
    "    (2810,1.2,1.8),\n",
    "    (2430,2.2,2.8),\n",
    "    (4880,3.2,3.8),\n",
    "    (2088,4.2,4.8)\n",
    "])\n",
    "hist.add('S-Cone(hib)',[\n",
    "    (1420,5.2,5.8),\n",
    "    (951,6.2,6.8),\n",
    "])\n",
    "hist.add('M-Cone',[\n",
    "    (0,7.2,7.8),\n",
    "    (1,8.2,8.8),\n",
    "    (2,9.2,9.8)\n",
    "])\n",
    "hist.add('M-Cone(hib)',[\n",
    "    (1,10.2,10.8),\n",
    "    (1,11.2,11.8),\n",
    "    (1,12.2,12.8),\n",
    "])\n",
    "\n",
    "hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
