{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED\n"
     ]
    }
   ],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import zscore\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "import os.path\n",
    "from cmcrameri import cm #colormaps\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "# 'r' : '#747474',\n",
    "# 'lr' : '#474747'\n",
    "# 'u' : '#B540B7',\n",
    "\n",
    "# 's' : '#4669F2',\n",
    "# 'esls' : '#8f9bcc',\n",
    "\n",
    "# 'm' : '#04CD22',\n",
    "# 'lslm' : '#18e236',\n",
    "# 'mslm' : '#57cb69',\n",
    "# 'eslm' : '#80bc89',\n",
    "\n",
    "\n",
    "# 'l' : '#CC2C2A',\n",
    "# 'lsll' : '#ce4340',\n",
    "# 'msll' : '#c67271',\n",
    "# 'esl' : '#d69f9e',\n",
    "\n",
    "# 'pr' : \"#E6B800\"\n",
    "# 'lslpr' : \"#cca819\"\n",
    "# 'mslpr' : \"#cfb54f\"\n",
    "# 'eslpr' : \"#dacd9a\"\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "\n",
    "def findgene_glia(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_glia(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_glia(genename):\n",
    "    gdf = findgene_glia(genename)\n",
    "    n = np.arange(1,19) # GFP+\n",
    "    n = np.append(n, 19.5 + np.arange(1,19)) # GFP-\n",
    "    h_start = 5\n",
    "    h_end = 41\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "#glia Colors\n",
    "gC = {\n",
    "    'm48' : '#c7ffec',\n",
    "    'm60' : '#94ffdb',\n",
    "    'm72' : '#57ffc7',\n",
    "    'm96' : '#2effb9',\n",
    "    'm120': '#05ffac',\n",
    "    'm192': '#00d68f',\n",
    "    \n",
    "    'c48' : '#e3e3e3',\n",
    "    'c60' : '#c9c9c9',\n",
    "    'c72' : '#ababab',\n",
    "    'c96' : '#969696',\n",
    "    'c120': '#828282',\n",
    "    'c192': '#6b6b6b',\n",
    "    \n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "gC['plt']=[\n",
    "    gC['m48'],gC['m48'],gC['m48'],\n",
    "    gC['m60'],gC['m60'],gC['m60'],\n",
    "    gC['m72'],gC['m72'],gC['m72'],\n",
    "    gC['m96'],gC['m96'],gC['m96'],\n",
    "    gC['m120'],gC['m120'],gC['m120'],\n",
    "    gC['m192'],gC['m192'],gC['m192'],\n",
    "    gC['c48'],gC['c48'],gC['c48'],\n",
    "    gC['c60'],gC['c60'],gC['c60'],\n",
    "    gC['c72'],gC['c72'],gC['c72'],\n",
    "    gC['c96'],gC['c96'],gC['c96'],\n",
    "    gC['c120'],gC['c120'],gC['c120'],\n",
    "    gC['c192'],gC['c192'],gC['c192'],\n",
    "]\n",
    "\n",
    "def findgene_rods(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_rods(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_rods(genename):\n",
    "    gdf = findgene_rods(genename)\n",
    "    n = np.arange(1,5) # GFP+\n",
    "    n = np.append(n, 4.5 + np.arange(1,5)) # GFP-\n",
    "    h_start = 7\n",
    "    h_end = 15\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#rod Colors\n",
    "rC = {\n",
    "    'r' : '#747474',\n",
    "    'c': '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "rC['plt']=[\n",
    "    rC['r'],rC['r'],rC['r'],rC['r'],\n",
    "    rC['c'],rC['c'],rC['c'],rC['c'],\n",
    "]\n",
    "\n",
    "def findgene_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_sqcones(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,8) # GFP+\n",
    "    n = np.append(n, 7.5 + np.arange(1,7)) # GFP-\n",
    "    h_start = 8\n",
    "    h_end = 21\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#squirrel Colors\n",
    "sqC = {\n",
    "    's' : '#4669F2',\n",
    "    'sHib' : '#548ced',\n",
    "    'm' : '#04CD22',\n",
    "    'mHib' : '#53e477',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "sqC['plt']=[\n",
    "    sqC['s'],sqC['s'],sqC['s'],sqC['s'],sqC['s'],\n",
    "    sqC['sHib'],sqC['sHib'],\n",
    "    sqC['m'],sqC['m'],sqC['m'],\n",
    "    sqC['mHib'],sqC['mHib'],sqC['mHib'],\n",
    "]\n",
    "\n",
    "\n",
    "def findgene_haircell(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_haircell(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_haircell(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,4) # riboTagged\n",
    "    n = np.append(n, 3.5 + np.arange(1,4)) # Whatever this is\n",
    "    h_start = 2\n",
    "    h_end = 8\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#hairCell Colors\n",
    "hcC = {\n",
    "    'h' : '#e147c0',\n",
    "    'c' : '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "hcC['plt']=[\n",
    "    hcC['h'],hcC['h'],hcC['h'],\n",
    "    hcC['c'],hcC['c'],hcC['c'],\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "# Hoang, 2020 loaders after re-analysis and re-clustering\n",
    "def findgene_zfHoang2020(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRet_HoangBlackshaw2020/\" #directory path\n",
    "    fPath = \"conesDev_Hoang.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_zfHoang2020(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRet_HoangBlackshaw2020/\" #directory path\n",
    "    fPath = \"conesDev_Hoang.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_zfHoang2020(genename): #percentExpression\n",
    "    gdf = findgene_zfHoang2020(genename)\n",
    "    n = np.arange(1,12) # pct\n",
    "#     n = np.append(n, 5.5 + np.arange(1,5)) # avg\n",
    "    h_start = 2\n",
    "    h_end = 13\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#Hoang Colors\n",
    "zfHC = {\n",
    "    'PRP' : \"#dfdac8\",\n",
    "    'eslPR' : '#dacd9a',\n",
    "    'mslPR' : '#dcc360',\n",
    "    'lslPR' : '#cca819',\n",
    "    'adPR' : '#ffd429',\n",
    "    'lslR' : '#a3a3a3',\n",
    "    'r' : '#7d7d7d',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "zfHC['plt']=[\n",
    "    zfHC['PRP'],\n",
    "    zfHC['eslPR'],zfHC['mslPR'],\n",
    "    zfHC['lslPR'],zfHC['adPR'],\n",
    "    zfHC['lslR'],\n",
    "    zfHC['r'],zfHC['u'],zfHC['s'],\n",
    "    zfHC['m'],zfHC['l']\n",
    "]\n",
    "\n",
    "# Xu, 2020 loaders after re-analysis and re-clustering\n",
    "def findgene_zfXu2020(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRetDev10x_Xu2020/ret72hpf/\" #directory path\n",
    "    fPath = \"conesXu.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_zfXu2020(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRetDev10x_Xu2020/ret72hpf/\" #directory path\n",
    "    fPath = \"conesXu.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_zfXu2020(genename): #percentExpression\n",
    "    gdf = findgene_zfXu2020(genename)\n",
    "    n = np.arange(1,7) # avg\n",
    "#     n = np.append(n, 5.5 + np.arange(1,5)) # pct\n",
    "    h_start = 2\n",
    "    h_end = 8\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#Xu Colors\n",
    "zfXC = {\n",
    "    'r' : '#7d7d7d',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'mslPR' : '#dcc360',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "zfXC['plt']=[\n",
    "    zfXC['r'],zfXC['u'],zfXC['s'],\n",
    "    zfXC['m'],zfXC['l'],zfXC['mslPR'],\n",
    "]\n",
    "\n",
    "def findgene_zfOgawa2021(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRet_Ogawa2021/\" #directory path\n",
    "    fPath = \"photoreceptors_Ogawa.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_zfOgawa2021(genename):\n",
    "    dPath = \"~/Documents/LiMolec/otherRNAseq/zfRet_Ogawa2021/\" #directory path\n",
    "    fPath = \"photoreceptors_Ogawa.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_zfOgawa2021(genename): #averageExpression\n",
    "    gdf = findgene_zfOgawa2021(genename)\n",
    "    n = np.arange(1,9) # \n",
    "    h_start = 2\n",
    "    h_end = 10\n",
    "#     h_end = 13\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "def plotBars2_zfOgawa2021(genename): #percentExpression\n",
    "    gdf = findgene_zfOgawa2021(genename)\n",
    "    n = np.arange(1,9) # pct\n",
    "#     n = np.append(n, 5.5 + np.arange(1,5)) # avg\n",
    "    h_start = 2+9\n",
    "    h_end = 9+9\n",
    "#     h_end = 13\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "\n",
    "#Hoang Colors\n",
    "zfO = {\n",
    "    'r' : '#7d7d7d',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'm4': '#cdcd04',\n",
    "    'onBC': '#ccf2ff',\n",
    "    'offBC': '#663d00',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "zfO['plt']=[\n",
    "    zfO['r'],zfO['u'],zfO['s'],\n",
    "    zfO['m'],zfO['l'],zfO['m4'],\n",
    "    zfO['onBC'],zfO['offBC']\n",
    "]\n",
    "\n",
    "print('LOADED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm\n"
     ]
    }
   ],
   "source": [
    "# %cd ~/Dropbox/Presentations/RevealReboot/resources/20191002_RNAseq/\n",
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm/\n",
    "# %cd /Users/angueyraaristjm/Desktop/ForHailey/\n",
    "# %cd ~//Documents/LiLab/Manuscripts/2021_zfConeRNAseq/resources/\n",
    "# %cd ~/Documents/LiLab/Presentations/revealjs/resources/20210608_UTH/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dark background\n",
    "plt.style.use('dark_background')\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)\n",
    "baseColor = '#ffffff' #white\n",
    "font_prop = font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# white background\n",
    "plt.style.use('default')\n",
    "params = {\"ytick.color\" : \"k\",\n",
    "          \"xtick.color\" : \"k\",\n",
    "          \"axes.labelcolor\" : \"k\",\n",
    "          \"axes.edgecolor\" : \"k\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"k\"}\n",
    "plt.rcParams.update(params)\n",
    "baseColor = '#000000' #black\n",
    "font_prop = font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [PCA Plots](#pcaPlots)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "- [Heatmap: Other Data Sets](#heatOther)\n",
    "- [Sparkline plots](#sparkline)\n",
    "- [PCA](#pca)\n",
    "- [GO clustering](https://go.princeton.edu/) \n",
    "- [Quantification](#quant)\n",
    "- [Phototransduction genes](#Phototrans)\n",
    "- [Transcription factors](#TFs)\n",
    "- [Known TFs](#knownTFs)\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.std(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## pcaPlots<a id='pcaPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatFigure_pca(plotname, figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=18)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=22)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=28)\n",
    "    \n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,4))\n",
    "    axH.ticklabel_format(style='sci',axis='x',scilimits=(0,4))\n",
    "    axH.set_title(plotname, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 18)\n",
    "    axH.xaxis.offsetText.set_fontsize(18)\n",
    "    axH.yaxis.offsetText.set_fontsize(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_RodsCones'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 21%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,50)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_UVS_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 23%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_U_S'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,100)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "# plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# US vs LM: PC1 = 13% and PC2 = 10%\n",
    "# M vs L: PC1 = 23% and PC2 = 10%\n",
    "# UV vs S: PC1 = 15% and PC2 = 13%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_bySubtype/\n",
    "#pyplot version (Sep_2019)\n",
    "plotname = 'TF_RvC_PCA'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_bySubtype/TFs_pcaData_RvC.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[7,7])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 11%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 7%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-60,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"./tempTFfigs/{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_bySubtype/\n",
    "#pyplot version (Sep_2019)\n",
    "plotname = 'TF_C_PCA'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_bySubtype/TFs_pcaData_C.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[7,7])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 7%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-50,50)\n",
    "plt.ylim(-50,50)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"./tempTFfigs/{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tSNE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'tSNE'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/tSNEData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[6,6])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('tSNE1', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('tSNE2', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_xticks([-1000,-500,0,500,1000])\n",
    "ax.set_yticks([-1000,-500,0,500,1000])\n",
    "plt.xlim(-1000,1000)\n",
    "plt.ylim(-1000,1000)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw1'\n",
    "n,h = plotBars(genename)\n",
    "h = h\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('rx')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "genename = 'nrtn'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize= [12,8])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.png\".format(genename), transparent=True, format=\"png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bar plots for opsins (manuscript)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'rho'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.ylim(0,2.0e5) # keep scaling consistent\n",
    "changeTitle('Rhodopsin', ax, pC['r'])\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "\n",
    "\n",
    "genename = 'opn1sw1'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.ylim(0,2.0e5) # keep scaling consistent\n",
    "changeTitle('UV opsin', ax, pC['u'])\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "\n",
    "genename = 'opn1sw2'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.ylim(0,2.0e5) # keep scaling consistent\n",
    "changeTitle('S opsin', ax, pC['s'])\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "\n",
    "# Summing mws and lws genes for figure\n",
    "genename = 'opn1mw1'\n",
    "n,h1 = plotBars('opn1mw1')\n",
    "n,h2 = plotBars('opn1mw2')\n",
    "n,h3 = plotBars('opn1mw3')\n",
    "n,h4 = plotBars('opn1mw4')\n",
    "h = h1+h2+h3+h4\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.ylim(0,2.0e5) # keep scaling consistent\n",
    "changeTitle('M opsins (1-4)', ax, pC['m'])\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "\n",
    "\n",
    "genename = 'opn1lw1'\n",
    "n,h1 = plotBars('opn1lw1')\n",
    "n,h2 = plotBars('opn1lw2')\n",
    "h = h1+h2\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.ylim(0,2.0e5) # keep scaling consistent\n",
    "changeTitle('L opsins (1-2)', ax, pC['l'])\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 1#2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,3.5e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,2e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gnat1','gnat2']\n",
    "# plotName = 'gnat'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "# genelist = ['grk1a','grk1b',\n",
    "#             'grk3',\n",
    "#             'grk4','grk5','grk5l',\n",
    "#             'grk6',\n",
    "#             'grk7a','grk7b']\n",
    "# plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('cacn')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['ctbp1','ctbp2a']\n",
    "plotName = 'ctbp'\n",
    "\n",
    "genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "plotName = 'cadm'\n",
    "\n",
    "genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "plotName = 'tjp'\n",
    "\n",
    "genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "plotName = 'nrxn'\n",
    "\n",
    "genelist = ['eml1','eml2','eml3']\n",
    "plotName = 'eml'\n",
    "\n",
    "genelist = ['elfn1b','elfn2a']\n",
    "plotName = 'elfn'\n",
    "\n",
    "genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "genelist = ['tbx2a','tbx2b']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "genelist = ['ncdn','ncdn']\n",
    "plotName = 'ncdn'\n",
    "\n",
    "genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "plotName = 'sema'\n",
    "\n",
    "genelist = ['cacna1aa','cacna1ab',\n",
    "            'cacna1da','cacna1db',\n",
    "            'cacna1fa','cacna1fb',\n",
    "            'cacna2d1a','cacna2d2a',\n",
    "            'cacna2d4a','cacna2d4b',\n",
    "            'cacna1sb',\n",
    "           ]\n",
    "plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "genelist = ['stxbp1a','stxbp1b',\n",
    "            'stxbp2','stxbp3','stxbp4',\n",
    "            'stxbp5a','stxbp5b','stxbp5l',\n",
    "            'stxbp6','stxbp6l']\n",
    "plotName = 'stxBP'\n",
    "\n",
    "genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "            'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "            'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "            'stx12','stx12l','stx16','stx17','stx18']\n",
    "plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "genelist = ['pcdh15a','pcdh15b','pcdh8','pcdh11','pcdh10a']\n",
    "plotName = 'pdch'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "\n",
    "genelist = ['lrit1a','lrit1b','lrit2','lrit3a','lrit3b']\n",
    "plotName = 'lrit'\n",
    "\n",
    "genelist = ['pou4f1','pou4f2','pou4f3','rbpms','grm6a','grm6b','vsx2','gfap','prkca','syt2a','rpe65a', 'opn4']\n",
    "plotName = 'controls'\n",
    "\n",
    "genelist = ['tbx2a','tbx2b','six7','opn1lw2']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "\n",
    "# genelist = ['pcdh1gc6','tex264a']\n",
    "# plotName = 'test'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "Samples = [\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "# np.around(np.divide(data.values,1e6),2)\n",
    "# opsinData = np.divide(data.values,1e6)\n",
    "opsinData = np.divide(data.values,1)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgene('rho')\n",
    "# gdf2 = findgenelist('opn1sw')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1mw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1lw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_opsins\"\n",
    "\n",
    "# gdf = findgenelist('slc24') -> NCKX transporter; slc25a5 is mitochondrial but known to express in eye\n",
    "# plotName = \"slc24\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16]) #not ephrin receptors\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "\n",
    "# gdf = findgenelist('dnmt')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "\n",
    "# gdf = findgenelist('gj')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones; lrit3 for rods\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1]) #highest ones that obscure others.\n",
    "# # gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_pcdh_partial\"\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_opn4\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "# gdf = gdf.drop([1,2])\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('sall') #sall4 mutations cause Duane syndrome; sall4 seems to be related to sall1 in zebrafish\n",
    "# plotName = \"heatMap_sall\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lrit')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "# from YiRong Peng's talk: ON SACs: slit2 and fezf1; OFF SACs: tenm3 and rnd1?; contactins; Foveal cones: vtn and cacna1f\n",
    "# gdf = findgenelist('cacna2d2b')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('nrxn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_nrxn\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('frmpd')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_frmpd\"\n",
    "\n",
    "# gdf = findgenelist('ntng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_netrins\"\n",
    "\n",
    "# gdf = findgenelist('dscam')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dscams\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dag\"\n",
    "\n",
    "# gdf = findgenelist('syne') #involved in photoreceptor migration to ONL\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_syne\"\n",
    "# gdf = findgenelist('rasgrf') #also involved in phtoreceptor migration\n",
    "\n",
    "# gdf = findgenelist('syt')\n",
    "# plotName = \"heatmap_syt\"\n",
    "\n",
    "# gdf = findgenelist('rfx')\n",
    "# plotName = \"heatmap_rfx\"\n",
    "\n",
    "# gdf = findgenelist('anks') #anks3-/- causes RP\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('spock') #https://zfin.org/action/image/view/ZDB-IMAGE-050208-655 -> GCL enriched?\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('hcls') # Laura Campello: downstream of nrl and crx in mouse\n",
    "\n",
    "# gdf = findgenelist('slc32')  #from GO analysis: slc32a1 = L-cone specific + synaptic vesicle transporter.\n",
    "# gdf = findgenelist('slc') \n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 100]\n",
    "\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "\n",
    "# from sqRNAseq + antibodies: rab32,ell2, ankrd33, pex5l, egr1, lamb1, chst3? or 7?, flna, cdh20, synj1\n",
    "# gdf = findgenelist('ell') # ell2 in sqRNAseql seems cone specific\n",
    "# gdf = findgenelist('ankrd33') #it's there, 33aa cone enriched; 33ba rods and 33bb cones.\n",
    "# gdf = findgenelist('pex5') #pek5l in sqRNAseq; pex5 in rods\n",
    "# gdf = findgenelist('egr') # egr1 looks interesting\n",
    "\n",
    "\n",
    "\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf2 = findgenelist('gnb')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('gng')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_transducin\"\n",
    "\n",
    "# gdf = findgenelist('aldh1')\n",
    "# gdf = findgenelist('foxq2')\n",
    "# gdf = findgenelist('prdm1b')\n",
    "# gdf = findgenelist('pbx1a')\n",
    "\n",
    "gdf = findgene('cdhr1a');\n",
    "gdf2 = findgene('pcdh2ab9'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('pcdh15b'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('lrit1b'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('dscamb'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('pcdh2g28'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('cadm3'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('sema3fa'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('lrit2'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('sema7a'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('sema6d'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('pcdh15a'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('pcdh10a'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('sema3fb'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('efna1b'); gdf = gdf.append(gdf2)\n",
    "gdf2 = findgene('pcdh11'); gdf = gdf.append(gdf2)\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "# data = gdf.iloc[0:,7:37].values\n",
    "# OR\n",
    "data = gdf.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values #normalizing by max\n",
    "# OR\n",
    "# data = zscore(data,axis=1, nan_policy='omit') # calculate z-score\n",
    "# data[np.isnan(data)] = np.nanmin(data) # make NaN black\n",
    "\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:37].values)\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Figures/zf/20210225_UTHApplication/resources/{0}.svg\".format('sema'), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf.symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "# foxq2 = S specific;\n",
    "\n",
    "# genelist = ['wrb','get4'] # disrupts photoreceptors synapses (Danielle, 2016)\n",
    "# genelist = ['pappaa','igf2bp2b'] # In Miller, 2018 model pappaa cleaves igf-bp to increase igf1 availability. unfortunately, this is supposed to bind mRNA not IGF itself\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1','meis1b','pbx1a']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "# genelist = ['il34','il7r']\n",
    "# genelist = ['cxxc4','cxxc1a']\n",
    "# genelist = ['mbd3b','mbd3a']\n",
    "# genelist = ['cdr2a','cdr2l']\n",
    "# genelist = ['coro6','coro2a','coro1ca']\n",
    "# genelist = ['auts2a','auts2b']\n",
    "# genelist = ['plxnb1a','plxna2']\n",
    "# genelist = ['cux1a','cux1b','cux2b']\n",
    "# genelist = ['arih1','arih2','arih1l']\n",
    "# genelist = ['crhbp','crhr2','crhr1']\n",
    "# genelist = ['chn1','chn2']\n",
    "# genelist = ['robo1','robo2','robo3','robo4'] #nothing here\n",
    "# genelist = ['ntng1a','ntng2a','ntng2b'] #low counts but interesting. Netrins are important for RGC axons going towards optic disc. ntng2b in mouse mediates rod-hc recognition!\n",
    "# genelist = ['fat1a','fat1b','yap1','scrib','cc2d2a'] #fat1 important for renal development along yap1/hippo and scribble; cc2d2a involved in cilium development\n",
    "# genelist = ['c1qbp','c1qa','c1qb'] #keep this in mind. Beth Stevens work shows that microglia engulf synapses through activation by c1q; maybe c1qbp is produced in healthy cells (inside mitochondria) to keep c1q bound and gets released upon damage.\n",
    "# genelist = ['cc2d2a','rab8'] #Bahcman 2011: cc2d2a-/- causes Joubert syndrom. Important for cilium development by association with rab8. Not expressed in adults?\n",
    "# genelist = ['per2','per3']\n",
    "# genelist = ['gpsm2','gpsm1b','gpr107','gpr108'] #\n",
    "# genelist = ['tbx2a','tbx2b','gdf6a','cnbpa','egr1', 'ndrg1a','ndrg1b'] # fate genes\n",
    "\n",
    "#Hensley, 2011 smarca4\n",
    "# genelist = ['smarca4a','guk1a','guk1b','dtnbp1a','elovl4b','id2b','ckmt2a','ckmt2b','aanat2','pbx1a'] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "# genelist = ['nme2a','nme3',''] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "\n",
    "# genelist = ['adgrl3.1','ntng2a','ntng2b'] #low counts but interesting. \n",
    "# genelist = ['cplx4a','cplx4b','cplx4c'] #low counts but interesting. \n",
    "\n",
    "# genelist = ['frmpd2', 'lrit3a'] #low counts but interesting. \n",
    "# genelist = ['si:busm1-57f23.1 ', 'mhc1uba','serinc2'] # \n",
    "\n",
    "# genelist = ['ccdc28a','ccdc25','ccdc173','sox6','prdm9','prdm1b']\n",
    "\n",
    "# genelist = ['per2','bhlhe40','nr1d2b','tefa','aanat2','aanat1'] #circadian genes. Correlate with collection time?\n",
    "\n",
    "# genelist = ['elovl4b','elovl8a'] #a cone specific gene. \n",
    "# genelist = ['pbx1a','egr1','egr3','egr4'] # fishing for new targets\n",
    "# genelist = ['foxg1b','foxq2','foxo3b','lmo4b','nfia'] # fishing for new targets\n",
    "# genelist = ['anxa13','prdx5','mpp5b','mpp6a','serpina10a'] # fishing for new targets\n",
    "# genelist = ['rbp4l','rbp3','impg1a','impg2b','ntm'] # fishing for new targets\n",
    "# genelist = ['tapt1b','tapt1a','frmpd2','wdr37','fabp7a','tle3a'] # fishing for new targets\n",
    "# genelist = ['ankrd33aa','ankrd33ab','dhx32a','dhx32b','wdr17']\n",
    "\n",
    "\n",
    "# genelist = ['slc1a8a','slc1a8b','slc1a4','slc1a7a'] # EAATs (glutamate uptake)\n",
    "# genelist = ['pmela','slc6a6b','naa35','ush2a'] # sex-associated rod genes in Sun, 2018 (all enriched in females, except pmela)\n",
    "\n",
    "# genelist = ['pcloa','pclob']\n",
    "# genelist = ['mtdha','mtdhb']\n",
    "# genelist = ['igsf9b','igsf21a','igsf21b']\n",
    "# genelist = ['rx3','lin9','igsf21b','foxq2','foxo3b','foxg1b','rxrga','smad5']\n",
    "# genelist = ['sall1a','sall1b','sall2','sall4']\n",
    "# genelist = ['rbm8a','rbm39a','rbm39b','rbmx','rbm22','rbm5']\n",
    "# genelist = ['rbm5','rbm4.1','rbm4.2']\n",
    "# genelist = ['kif5ba','kif1aa','kif1ab','kif1bp'] #kif5-/- : Charcot-MarieTooth; Campbell 2013: kif5s in all retinal layers, but only kif5ba and bb are in ONL\n",
    "# genelist = ['kif3a','kif3b','kif16ba','kif19','kif25'] #\n",
    "# genelist = ['robo2','slit1a','nbeaa','prdm1b'] #\n",
    "\n",
    "\n",
    "# genelist = ['pcdh2ab9','pcdh2g28','pcdh15a','pcdh15b','pcdh11','pcdh10a'] #interesting ones\n",
    "# plotName = 'pcdhs'\n",
    "\n",
    "# genelist = ['sema7a','sema6d','sema3fb','sema3fa','sema4ab','sema4ba'] #the good ones\n",
    "# genelist = ['sema6a','sema3aa','sema6e','sema5bb','sema4c','sema6bb','sema6dl','sema5ba'] # the inconsistent ones\n",
    "# genelist = ['sema6ba','sema5a','sema3c','sema3b','sema3e','sema3d','sema3h','sema3ab','sema4e','sema4aa','sema3gb','sema3ga','sema4bb','sema3bl','sema4ga','sema4gb'] # the negative ones\n",
    "\n",
    "# genelist = ['ncdn','flrt3'] \n",
    "# genelist = ['odf2a','odf2b']\n",
    "# genelist = ['syne1a','syne1b','syne2b','syne3']\n",
    "# genelist = ['rasgrf2a','rasgrf2b']\n",
    "# genelist = ['rasgrf2a','rasgrf2b','nmnat2','rbx1','nrf1','unc119.2','unc119b'] # involved in migration\n",
    "\n",
    "# genelist = ['syt5a','syt5b','syt1a','syt1b','syt14a']\n",
    "\n",
    "\n",
    "# genelist = ['myo19','myo16','myo7bb','myo5b','myo1cb','myo5aa','myo7aa','myo6a','myo6b']\n",
    "# genelist = ['fgfr1op','fgfr1op2','fgf1b','fgf13a','fgf11b']\n",
    "# genelist = ['prdm1b','fgfr1op2']\n",
    "# genelist = ['frmpd1a','gpsm2'] # Ulisse + Pahlberg: these proteins interact with rod transducin \n",
    "\n",
    "# genelist = ['epha7','frmpd2','pkp4','chd7','hook1']\n",
    "# genelist = ['thrb','samd7','samd1a','samd11','dscaml1'] #samd11:https://www.ncbi.nlm.nih.gov/pubmed/27734943\n",
    "# genelist = ['thrb','nrl']\n",
    "# genelist = ['opn3','opn4.1','opn4a','opn4b','opn5','opn6a','opn9'] #\n",
    "\n",
    "# genelist = ['foxq2','prdm1b','nr2f6b','pbx1a','samd7','samd1a','samd11'] # \n",
    "v"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: photoreceptors, rods, Müller Glia, microglia<a id='heatOther'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)  \n",
    "[Go to Exploring TFs](#TFs)  \n",
    "[Go to Known TFs](#knownTFs)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# geneTag = 'cry[0-9]' # this is how you can select numbers\n",
    "geneTag = 'lrrfip1'\n",
    "# geneTag = 'zgc:73359' #supposed to be pde6hb (Ogawa, Corbo, 2021)\n",
    "# geneTag = 'si:dkey-44k' # supposed to be cng1b (Ogawa, Corbo, 2021)\n",
    "gdf = findgenelist(geneTag)\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([17,3,8,7])\n",
    "gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "\n",
    "odf = findgenelist_zfOgawa2021(geneTag)\n",
    "odf = odf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "rdf = findgenelist_rods(geneTag)\n",
    "rdf = rdf.sort_values(by=[\"symbol\"])\n",
    "# rdf = rdf[rdf.baseMean > 0]\n",
    "\n",
    "mdf = findgenelist_glia(geneTag)\n",
    "mdf = mdf.sort_values(by=[\"symbol\"])\n",
    "# mdf = mdf[mdf.baseMean > 0]\n",
    "\n",
    "sqdf = findgenelist_sqcones(geneTag)\n",
    "sqdf = sqdf.sort_values(by=[\"symbol\"])\n",
    "# sqdf = sqdf[sqdf.baseMean > 0]\n",
    "\n",
    "hcdf = findgenelist_haircell(geneTag)\n",
    "hcdf = hcdf.sort_values(by=[\"symbol\"])\n",
    "# hcdf = hcdf[hcdf.baseMean > 0]\n",
    "\n",
    "# zflhdf = findgenelist_zfLarvalHoang2020(geneTag)\n",
    "# zflhdf = zflhdf.sort_values(by=[\"symbol\"])\n",
    "# zfahdf = findgenelist_zfAdultHoang2020(geneTag)\n",
    "# zfahdf = zfahdf.sort_values(by=[\"symbol\"])\n",
    "zfhdf = findgenelist_zfHoang2020(geneTag)\n",
    "zfhdf = zfhdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "zfxdf = findgenelist_zfXu2020(geneTag)\n",
    "zfxdf = zfxdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "# plotName = 'heatMap_vamp'\n",
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "## plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "genenameso = odf['symbol'].values\n",
    "samplenameso = [] #skipping names to replace in intaglio\n",
    "datao = odf.iloc[0:,2:10].values\n",
    "figo, axo = plt.subplots(figsize=[8,8])\n",
    "imo, cbaro = heatmap_zfOgawa2021(datao, genenameso, samplenameso, ax=axo)\n",
    "### plt.savefig(\"{0}_Ogawa.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "# Hoang reanalyzed 10x dataset and Xu reanalyzed 3dpf dataset\n",
    "# split heatmaps\n",
    "genenameszflh = zfhdf['symbol'].values\n",
    "samplenameszflh = [] #skipping names to replace in intaglio\n",
    "datazflh = zfhdf.iloc[0:,2:8] #avg. expression\n",
    "# datazflh = zfhdf.iloc[0:,2+12:8+12] # % expression\n",
    "genenameszfah = zfhdf['symbol'].values\n",
    "samplenameszfah = [] #skipping names to replace in intaglio\n",
    "datazfah = zfhdf.iloc[0:,8:13] # avg expression\n",
    "# datazfah = zfhdf.iloc[0:,8+12:13+12] # % expression\n",
    "\n",
    "genenameszfx = zfxdf['symbol'].values\n",
    "samplenameszfx = [] #skipping names to replace in intaglio\n",
    "datazfx = zfxdf.iloc[0:,2:8]\n",
    "\n",
    "figr, (axzflh,axzfah,axzfx) = plt.subplots(1,3,figsize=[20,36])\n",
    "imzflh, cbarzflh = heatmap_zflarvaHoang2020(datazflh, genenameszflh, samplenameszflh, ax=axzflh)\n",
    "imzfah, cbarzfah = heatmap_zfadultHoang2020(datazfah, genenameszfah, samplenameszfah, ax=axzfah)\n",
    "imzfx, cbarzfx = heatmap_zfXu2020(datazfx, genenameszfx, samplenameszfx, ax=axzfx)\n",
    "### plt.savefig(\"{0}_10x.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "#single heatmap\n",
    "# genenameszfh = zfhdf['symbol'].values\n",
    "# samplenameszfh = [] #skipping names to replace in intaglio\n",
    "# datazfh = zfhdf.iloc[0:,2:13] #avg. expression\n",
    "# # datazfh = zfhdf.iloc[0:,2+11:13+11] #% expression\n",
    "# figzfh, axzfh = plt.subplots(figsize=[16,16])\n",
    "# imzfh, cbarzfh = heatmap_zfHoang2020(datazfh, genenameszfh, samplenameszfh, ax=axzfh)\n",
    "# plt.show()\n",
    "\n",
    "genenamessq = sqdf['symbol'].values\n",
    "samplenamessq = [] #skipping names to replace in intaglio\n",
    "datasq = sqdf.iloc[0:,8:21]\n",
    "figsq, axsq = plt.subplots(figsize=[16,16])\n",
    "imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq)\n",
    "### plt.savefig(\"{0}_sq.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "genenamesr = rdf['symbol'].values\n",
    "samplenamesr = [] #skipping names to replace in intaglio\n",
    "datar = rdf.iloc[0:,7:16]\n",
    "genenamesm = mdf['symbol'].values\n",
    "samplenamesm = [] #skipping names to replace in intaglio\n",
    "datam = mdf.iloc[0:,5:41]\n",
    "genenameshc = hcdf['symbol'].values\n",
    "samplenameshc = [] #skipping names to replace in intaglio\n",
    "datahc = hcdf.iloc[0:,2:8]\n",
    "\n",
    "figr, (axr,axm, axhc) = plt.subplots(1,3,figsize=[24,16])\n",
    "# figr = plt.figure(figsize=[16,2],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(1,28)\n",
    "# axr = figr.add_subplot(fig_grid[0,0:8])\n",
    "# axm = figr.add_subplot(fig_grid[0,9:])\n",
    "imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr)\n",
    "imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm)\n",
    "imhc, cbarhc = heatmap_haircell(datahc, genenameshc, samplenameshc, ax=axhc)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:15: UserWarning: FixedFormatter should only be used together with FixedLocator\n",
      "  from ipykernel import kernelapp as app\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhUAAAFiCAYAAABBKyqaAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAtTElEQVR4nO3deVhV9b7H8c9GHFFDRM0hB5zQ6pw6euzmkFOTmWZ5rBywSa95moebDadralmdboPlKU2yVJQ07apJ9WSFaUI5NnmUNFBM0AIcEJHxe//wgiFsRPaCzYb363m+T7XXb6313Svb+9Pav7WWS5IJAADAQ37ebgAAAFQPhAoAAOAIQgUAAHAEoQIAADiCUAEAABxBqAAAAI7w93YDAGq2yy67TGFhYerYsaNOnjyp5557Tlu2bPF2WwDKySiKov5Y7777rpmZ9e/fv0L388gjj1heXp79Uc+ePb3+/imKKl/x8wcArwgJCdGLL76ojIwMDRkyRHXq1FHLli21fft2de/eXfv27dP69etVr169Sunnmmuu0dGjR2Vm6tixY6XsE6huCBUAvGL48OGqVauWXn/9dX366afKycnRwYMHlZeXp2uvvVZt27ZVv3791L179wrv5a677tJHH32kxo0bV/i+gOqMORUAvKLgbMCmTZuKLYuMjNS1116rvXv3avv27RXax4wZM/SPf/xDWVlZyszMVP369St0f0B1RqgA4BUBAQGSpGPHjhVblpycrKuvvrpC91+7dm3Nnz9f48aNU0ZGhm688Ua9/fbbat++fYXuF6jOCBUAvMLlckmSzMwr+7///vs1btw4paWlaejQofrmm2+80gdQnTCnAkCN9MYbb2jVqlXq378/gQJwCGcqANRI2dnZGjFihLfbAKoVzlQAKJN27drJzDRv3jxJ0s0336yNGzfqyJEjMjOFhIRIkhISErR7925JUo8ePbRq1SolJSXJzHTnnXfKzGRmuv322yVJ69atK3zttttuK7Kv6OjoYn1ER0fLzNSiRQtJ0i233KKoqCgdPHhQmZmZ2rt3rxYuXKgrrriiQo9Hly5d9Morr2jr1q1KSUlRVlaWEhMT9fbbb6tt27YVum+gquJMBYBz9sILL2jKlCk6duyYdu3apdzcXMXHxxcZM2zYMC1fvlwul0s7duzQvn371LBhQ61bt06SFBoaqvPPP1/fffedjhw5Ikk6ePBgmXto1KiRFi9erMGDBysnJ0e7du1SfHy8OnTooLCwMIWFhWnx4sWaMGGCTp486dRblyQ99dRTmjZtmmrVqqWjR49q9+7d8vf3V6dOnTRx4kTddNNN6tOnj+Li4hzdL+ALvH4HLoqiqlaVdEfNdu3amZnZd999Z2Zmc+fOtfr16xdbNyEhwZKSkiw5OdliY2OtdevWZd7HmfuKjo4utiw6OtrMzLZs2WI5OTn27LPPWpMmTYqMGTx4sP38889mZrZy5coyv++EhAQzM+vYsWOp41566SWLi4uzG264wWrVqlX4ekBAgL399ttmZvbFF194/d8jRXmhvN4ARVFVrEoLFWZmc+bMcbtuwRfz9u3brV69eue0jzP3VVqoyMvLs5EjR7rdfnBwsO3du9fMzMaNG1em913WUCGpxEAlyWrXrm2JiYmWl5dnzZs39/q/S4qqzGJOBYBzkpubq//+7/8+67gZM2Y4/rPDH0VERGjFihVul6ekpOiRRx6RpMK/OikzM7PE13NycrRt2zb5+flxzwvUOMypAHBOEhIS9Ntvv511XEVfplkwYbQ0q1atUnp6ui655BI1a9ZMv//+u6M9XHDBBbr22mvVvXt3tWjRQg0bNlRAQEDhrcW5OydqGkIFgHNy4MCBMo1LSkqq0D5++OGHs47Jzc3Vnj17dOmll6pt27aOhYrAwEC9+eabuuWWW+TnxwlfoAD/NQCoEZy6c6e/v7/WrVun0aNHa8eOHbrtttvUtm1b+fv7y+VyyeVy6b333nNkX4Cv4UwFAJ908cUXa+PGjaWOKbjMU5ISExMd2e/o0aP15z//WZs2bVLfvn2Vk5NTbExwcLAj+wJ8DWcqAPikiRMnnnXM8OHD1ahRo8IbVDnh0ksvlXTqSaolBQrp1D04gJqIUAHAJ4WFhemmm25yuzw4OFgvv/yyJBX+1Qn+/qdO8DZr1qzE5VdeeWXh2RGgpiFUAPBJcXFxWrp0qWbMmKEmTZoUWTZ48GDFxMSoffv2WrlypSIjIx3b75dffilJevDBBzVkyJAiy0aOHKkPPvig8E6atWvXdmy/gC8gVADwSWPGjNGmTZv0j3/8Q4cOHdL333+vmJgYJSUl6fPPP1fnzp21ZMkSjR492tH9rly5UosXL1aDBg308ccfKz4+Xhs3btS+ffu0fPlyLVu2TAsWLJAkBQQEOLpvoKrz+VDRtWtXZWVladeuXapTp47bcWPHjlVMTIxSU1OVlpam2NhYjR07thI7BeCkw4cPq2/fvrr99tv11VdfqXXr1vrLX/6inJwcRUZGatCgQRo7dmyF3IDr9ttv12OPPaZt27YpODhY3bp10/79+3Xbbbdp0qRJhZeunn/++Y7vG6jqvH5bT0/q008/NTOzwYMHux0zZ84cMzPbuXOnzZ4929544w3buXPnWW83TFFU1auC23S3a9fO671QFFWsvN5AuWvEiBFmZrZ48WK3Y8LCwszMbNGiRebv71/4ur+/vy1atMjMzEaPHu3190JRVNmKUEFRVbq83kC5qm7duvbLL7/Y4cOHrUWLFm7H7dmzx5KSkkp8sFHdunXtwIEDFhcX5/X3Q1FU2YpQQVFVt3x2TsVjjz2mkJAQPfXUUzp06FCJY3r16qWOHTsqPDy8xN9Vs7KyNH/+fHXp0kU9e/as6JYBAKjWfDJUXHDBBXr88ce1efNmvfXWW27HXX755ZJOXwJWkujoaElS7969nW0SAIAaxidDxcsvv6y6devq7rvvLvV+/iEhIZJUeM14SXbv3l1kLAAAKB+fCxUDBgzQqFGjlJCQoNmzZ+vIkSM6cuSI1q5dq6FDhxYZ27RpU0lSamqq2+0V3Lq3YCwAACgfnwsV06dPlyR16tRJjRs31hdffKEdO3Zo4MCBWrNmjWbOnFk4NigoSPn5+crOzna7vczMTOXn5ysoKKjE5VOnTpWZnbUAAKjpfOoppX379lW/fv2UkZGhSZMmafHixYXL/vrXv2rJkiV64okn9O2332rVqlWSJD+/0nOTn5/fWccAqDoGDhzo7RYAuOFT36ZhYWGSpCeffLJIoJCkzZs3a/To0crLy9OUKVMkSWlpaZKkBg0auN1mwW10C8YCAIDy8alQcemllyozM9PtFR9btmzR119/rV69eqlevXqFcykCAwPdbrNgmbt5F9OmTZPL5SqxAADAaT4VKgIDA3Xo0CHl5OS4HbN3717VqlVLTZo0UXx8vCSpS5cubsd37txZkgrHAgCA8vGpUJGamqqmTZuWOgeiRYsWysvL0+HDhxUTEyNJGjRokNvxBb/PFowFAADl5/Xbepa1nn/+eTMzGzVqVInL27dvbydOnLCYmJjC10q7TXe9evUsKSmp3Lfp/iNvHxuKoiiKqgLl9QbKXK1bt7b09HRLTk62Hj16FFnWokUL27Bhg5mZDR06tPD1ggeKRUREOP5AMUIFRVEURRUprzdwTjV06FA7fvy4ZWdn27Jly+zZZ5+1efPmWWpqqpmZPfnkk8XW+eOjz19//XXHHn1OqKAoiqKoIuX1Bs65OnbsaPPmzbO9e/daVlaW/fbbb7Zq1Srr37+/23XGjRtnsbGxlpaWZmlpaRYbG2vjxo3zqA9CBUVRFEWdLtf//w3K4Y930uQSUwBATedTV38AAICqi1ABAAAcQagAAACOIFQAAABHECoAAIAjCBUAAMARhAoAAOAIQgUAAHAEoQIAADiCUAEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAShAgAAOIJQAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwBKECAAA4glABAAAcQagAAACOIFQAAABHECoAAIAjCBUAAMARhAoAAOAIQgUAAHAEoQIAADiCUAEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAShAgAAOIJQAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwBKECAAA4glABAAAcQagAAACOIFQAAABHECoAAIAjCBUAAMARhAoAAOAIQgUAAHAEoQIAADiCUAEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAShAgAAOIJQAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwRLUJFc8//7zMTDk5OW7HjB07VjExMUpNTVVaWppiY2M1duzYSuwSAIDqyyXJvN2Epzp16qSffvpJfn5+crlcql27drExc+bM0aRJk7Rr1y598cUXMjNdeeWVCg0N1dy5c3X33Xef837NTh86l8vl0XsAAKA6MF+vNWvW2LZt2+yrr76ynJycYsvDwsLMzGzRokXm7+9f+Lq/v78tWrTIzMxGjx59zvv9I28fA4qiKIqqAuX1BjyqoUOHmpnZFVdcYdHR0SWGij179lhSUpLVq1ev2LK6devagQMHLC4u7pz3TaigKIqiqNPl03MqateurVdffVUffPCB1q9fX+KYXr16qWPHjgoPD9fJkyeLLc/KytL8+fPVpUsX9ezZs6JbBgCg2vLpUPHII4/oggsu0GOPPeZ2zOWXXy5J+vLLL92OiY6OliT17t3b2QYBAKhBfDZUtGrVSk899ZReeeUV7d271+24kJAQSVJcXJzbMbt37y4yFgAAnDufDRX/8z//o/T0dM2cObPUcU2bNpUkpaamuh2TkpJSZOwfTZ06VWZWYgEAgNP8vd1AefTr10+jR4/WHXfcoYyMjFLHBgUFKT8/X9nZ2W7HZGZmKj8/X0FBQU63CgBAjeFzZyr8/Pz0+uuva8uWLXrvvffKvM7Zlp9tDAAAKJ3PfZPefffduuSSS/Tggw+WaXxaWpokqUGDBm7HBAQEFBn7R9OmTZPL5SqxAADAaT4VKurUqaMZM2Zo6dKl2rhxY5nWKZhLERgY6HZMwbLS5l0AAIDS+VSoqF+/voKCgnTLLbeUOHFywIAB8vf3L/znBx54QPHx8ZKkLl26uN1u586dJalwLAAAOHc+NVEzOztb77//vtvlgwYNUnBwsJYtWybp1GWkBWcfBg0apHXr1pW43sCBAyVJMTExzjYMAEANUi0eKFYgOjpaffv2LfZAsT179qhBgwYKCQkpdlfNevXqKT4+Xunp6erates57Y8HigEAcJpP/fxRXtOmTVPLli0VHh4uf//TJ2f8/f01b948tWzZUs8884z3GgQAoBrwqZ8/ymvRokXq06ePJk2apB49emjt2rVyuVxFHn0eGRnp7TYBAPB5Xn+qmVPl7imlBTVu3DiLjY21tLQ0S0tLs9jYWBs3bly598dTSimKoijqdFWrORWVjTkVAACcViPmVAAAgIpHqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcIQjd9SsXbu2Bg4cqNDQUDVu3LjUezaYmZ599lkndgsAAKoQj29+ddNNN+nNN99UcHDwqQ2e5SZQZlbk+Ru+jJtfAQBwmkff7n379tXSpUvl5+enn376SWvWrFFSUpLS09Od6g8AAPgIj0LFY489JpfLpUceeUSvvfaaQy0BAABf5NHPH4cOHVJubq5at27tYEu+g58/AAA4zaOrPxo2bKj4+HinegEAAD7Mo1Dxyy+/qHnz5k71AgAAfJhHoWLhwoXq2LGjBgwY4FA7AADAV3k0p6J27dqKjo5Wt27d9PDDDysyMlLZ2dkOtle1MacCAIDTPL5PRUBAgObOnatbb71VR44c0ffff6+kpCS34cLMNGHCBE92WWUQKgAAOM3jUHHvvffqxRdfVL169co0nptfAQBQPXn07X7jjTdq1qxZys/PV1RUlD766CNufgUAQA3lUai47777ZGa69dZbtWLFCqd6AgAAPsijnz+OHDmitLQ0hYSEONiS7+DnDwAATvPoktKcnBwdOHDAqV4AAIAP8yhUfPPNN+ratSv/lw4AADwLFS+//LKCgoL05JNPOtUPAADwYeZJPfTQQ5aVlWVz58619u3be7QtX6s/8nYvFEVRFOXt8mii5hdffCFJuvDCCxUcHCxJ2rdvnw4dOqTMzMwS1zEzXXnlleXdZZXCRE0AAE7zKFTk5eWd8zrc/AoAgOrJo2/3P//5zzp+/Ljy8/Od6gcAAPgoj2/TXZNxpgIAgNM8uvrDz+/cV7/nnns82SUAAKiiPAoVy5YtO6dg8fjjj2vWrFme7BIAAFRh5b50JC8vzyIjI83lcp117IwZMywvL89yc3O9fsmLU8UlpRRFURRVpMq/clJSkuXm5tqiRYtKHffaa69Zbm6u5eTk2IQJE7z9hh0rQgVFURRFFanyr9y5c2fbv3+/5ebm2rvvvlvimHnz5lleXp5lZmba3/72N2+/WUeLUEFRFEVRRcqzDYSEhNi+ffssNzfX5s2bV/i6n5+fLVmyxPLy8uzYsWN25ZVXevuNOl6ECoqiKIoqUp5vpH379hYfH2+5ubn25ptvWp06dWz16tWWl5dnKSkpdtlll3n7TVZIESooiqIo6nQ5dp+KCy64QNHR0Wrfvr2Sk5PVqlUrJSUl6ZprrtG///1vJ3ZR5XCfCgAATnP05letWrVSdHS0OnXqpD179uiqq65SYmKiU5uvcggVAACcVmqo6Nev3zlvsFWrVvrnP/+pRx99VAcPHixxzIYNG855u1URoQIAgNNKDRV5eXlFvjjLvFGXq9T1eKAYAADVT6nf7gsXLixXqAAAVH/+W0PKPDa3R3wFdoKqggeKeYAzFQBqMkIFzuTRsz8AAAAKOD65oWnTpmrbtq3MTImJiUpLS3N6FwCAClDWMw+cdYA7jp2pGDNmjLZu3apDhw5p8+bN2rJli3777Td9++23uvnmm53aDQAAqKI8nlNRt25dRUZGavjw4XK5XDp+/Lj27NkjSerUqZMaNmwoM9PKlSs1ZswYZWdnO9F3lcCcCgDVybmeqWBOBc7k8ZmK8PBw3XDDDdq/f79GjRqloKAg9ejRQz169FBQUJBGjRql/fv3a8SIEQoPD3eiZwAAUAV5dKaiX79+WrdunXbv3q0+ffooNTW1xHHBwcGKiYlRSEiIBg0apPXr15d3l1UKZyoAVCecqYCnPDpTMWHCBJmZJk+e7DZQSFJKSoomT54sl8uliRMnerJLAABQRXl0piIhIUF16tRR69atyzT+wIEDys7OVocOHcq7yyqFMxUAqhPOVMBTHp2paNGihRISEso8fu/evWrRooUnuwQAAFWUR6Hi6NGjatiwYZnHBwQE6Pjx457sEgAAVFEehYpt27YpNDRU55133lnHBgYGqlu3btq+fbsnuwQAAFWUR6Fi4cKFql27th5++OGzjn344YdVq1YtRUREeLJLAABQRXl886vPP/9c/fr10y233KKVK1eWOObGG2/U0qVL9e2336pfv36e7K5KYaImgOqEiZrwlEeholmzZmrcuLHef/99XXLJJVq0aJHCw8P1ww8/SJIuvvhiTZw4UWFhYYqLi9OYMWN0+PDhs253//795W2pUhEqAFQnhAp4yqNQkZeXV/jF6nK5inzJFtlJKcvOZGaqXbt2eVuqVIQKANUJoQKecvuU0oYNG571So3169eXOSwAAIDqzW2omDt3rq677jqFhobq0KFDJY4ZP368Tp48qd9//73CGgQAAL7B7dUfTZs2VaNGjZSSkuJ25YSEBEVGRlZIYwAAwLe4PVNx8OBBSVJISIh2795d4hiXy8VcAgCOuXrysTKN++ytxhXcCYDycBsq1qxZo7CwMK1YsUIPPPCANm/ezN0wAeAM5zq5EajO3IaK5cuXa/Xq1Ro+fLjWrl3rdgP9+/dXbm5umXfoS1d3AACAsnMbKiTppptu0vjx4zVixAh17NhRAQEBRX7uaNeunbKzswt/KgEAADVXqaHCzLRgwQItWLCgxOV5eXmKjY3V4MGDK6Q5AAA/scB3ePTsDwAAgAKlnqk4mwULFmjnzp1O9QIAAHyYR6HizjvvdKoPAADg4/j5AwAAOIJQAQAAHEGoAAAAjiBUAAAAR/hkqGjZsqVmz56thIQEZWVlKSUlRR9//LGuu+46t+uMHTtWMTExSk1NVVpammJjYzV27NhK7BoAgOrN50LFn/70J33//fe65557tH//fr3xxhuKiorSZZddpqioKD333HPF1pkzZ44iIiLUpEkTRUZGavHixQoMDFRERITmzJnjhXcBAED149ElpZXNz89PS5cuVbNmzXTXXXdp/vz5hcuCg4P12Wef6cknn9Qnn3yir7/+WpIUFhamSZMmKSIiQnfccUfhc0r8/f317rvvatKkSfrqq694hDsAAB7yqTMVgwYNUmhoqD788MMigUKSUlJSdN9990k6FSQKTJ06VcnJyZo4cWKRB5/l5uZqwoQJSkpK0jPPPFMp/QMAUJ35VKjYvHmz+vTpo0cffbTE5Vu3bpV06kFnktSrVy917NhR4eHhOnnyZLHxWVlZmj9/vrp06aKePXtWXOMAANQAPhUqjh49qpiYGCUkJJS4vHnz5pKk9PR0SdLll18uSfryyy/dbjM6OlqS1Lt3bydbBQCgxvGpUHE2BVdzbNiwQZIUEnLqyX5xcXFu19m9e3eRsQAAoHx8aqJmaTp37qwnnnhCaWlphfMtmjZtKklKTU11u15KSkqRsWeaOnUqcy4AOKasjzGXeJQ5fE+1CBWBgYFavXq1GjVqpEmTJun48eOSpKCgIOXn5ys7O9vtupmZmcrPz1dQUFBltQsAQLXk8z9/1K9fX1FRUQoNDdXChQv19ttvF1nu51f6W/Tz8zvrGAAAcHY+/W3q7++vFStWqHfv3vr444911113FVmelpYmSWrQoIHbbQQEBBQZe6Zp06bJ5XKVWAAA4DSfDRUul0sLFy7UkCFDtG7dOo0cObLIfSik03MpAgMD3W6nYFlp8y4AAMDZ+eycitmzZ2v06NGKiYnR9ddfX+J9KOLjT01y6tKli5KSkkrcTufOnYuMBVDc1ZOPlXnsZ281rsBOAFRlPnmmYvr06fr73/+urVu3asiQIcrIyChxXExMjKRTd+J0Z+DAgUXGAgCA8vG5UHHffffp6aef1o8//qirr75ax465/z+ozZs365dfftGECRNUr169Ysvr1aunu+66Sz///LO2bNlSkW0DAFDt+VSoGDt2rF577TXFxcXpqquucju58o+mTZumli1bKjw8XP7+p3/t8ff317x589SyZUvuQwEAgAN8ak7FjBkz5Ofnpw0bNujuu+92O+7IkSOaNWuWJGnRokXq06ePJk2apB49emjt2rVyuVy68sorFRoaqrlz5/KEUgAAHOBToaLgMs4JEyaUOm7v3r2FoUKS7r77bn399de65557NG7cOEmnbt0dFhamiIiIimsYAIAaxKdCRYcOHcq9bkREBAECAIAK5FOhAgBQNmV9xgjPF4GTfGqiJgAAqLoIFQAAwBH8/AHUQGW9QyZ3xwRwLjhTAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwBPepAABUGm4fXr1xpgIAADiCUAEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAShAgAAOIJQAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwBKECAAA4glABAAAcQagAAACOIFQAAABHECoAAIAj/L3dAAB44urJx8o07rO3GldwJwA4UwEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAShAgAAOIJQAQAAHEGoAAAAjiBUAAAAR/DsD8BDUQM+KdO4oeuGVHAnAOBdnKkAAACO4EwFAABetuWi7mUe2/Onf1dgJ57hTAUAAHAEoQIAADiCUAEAABxBqAAAAI4gVAAAAEcQKgAAgCMIFQAAwBGECgAA4AhCBQAAcAR31AS8gOeFAKiOOFMBAAAcQagAAACOIFQAAABHECoAAIAjCBUAAMARhAoAAOAIQgUAAHAEoQIAADiCUAEAABxBqAAAAI4gVAAAAEfw7A8ANcrVk4+VeexnbzWuwE6A6oczFQAAwBE1KlQEBwfrpZdeUlxcnE6cOKF9+/bp/fff14UXXujt1gAA8Hk1JlS0adNG33zzjR599FHt2bNHs2fP1vr16zVs2DBt2rRJgwYN8naLAAD4tBozp+K9995Tu3btNHz4cH300UeFr0+dOlUbNmzQkiVL1LVrVx09etSLXcLbogZ8UqZxQ9cNqeBOAMD31IgzFVdccYUGDx6s2bNnFwkUkhQfH6/JkyerRYsWuueee7zUIQAAvq9GnKm49dZbJUlvvPFGictXr16t/fv3a8yYMZo5c2ZltgYAQLlsuah7mcb1/OnfFdzJaTXiTMXll1+uffv2KT4+3u2YdevWqVu3bjrvvPMqsTMAAKqPGhEqQkJCFBcXV+qY3bt3y8/PTx06dKikrgAAqF6q/c8f/v7+aty4sVJTU0sdl5KSIklq2rRpkdenTp2qZ5555qz7MbNy9wjfYzr3f9+VtY6j3izH/suzTmXsw9vvpTybOtd1KmMflbUOH6luVaVD6XK5ivxztT9TERQUJEnKzMwsdVxGRkaR8QAA4NxU+1BRoFatWqUu9/ev9idtAACoUNX+mzQtLU2S1KBBg1LHBQQEFBlfYNq0aZo2bVrFNOcD/vizzpmnuVAyjtm545hVPI5xxeMY14AzFbm5uTp27JgCAwNLHVew/GxzLwAAQMmqfaiQTt3gqkuXLqWO6dy5s/Lz85WQkFBJXQEAUL3UiFARExOjdu3aKSQkxO2YAQMGaOfOndymGwCAcqoRoeL999+XJN1///0lLr/hhht0wQUXaMmSJZXZFgAA1Y7VhPr8888tJyfHhg0bVuT1kJAQO3DggB08eNDOO+88r/dZ1eqPvN2LrxTHjGNWFYtjzDGujKr2V38UuP3227Vu3TqtXr1aUVFR2rFjh1q1aqUbb7xRLpdLw4YN46ePEpTlxl8oimN27jhmFY9jXPE4xpJLp9JFjRAcHKwpU6bohhtuUJs2bZSSkqKYmBjNmDFDO3bs8HZ7AAD4tBoVKgAAQMWpERM1AQBAxSNUwHHt2rWTmWnq1KnebgUAUIkIFdVAdHS0zKxI5ebmKi0tTRs2bNADDzygOnXqeLtNn2VmWrt27VnH9e/fX2amp556SpL0n//5nzIzPfvss2Xe1w8//KCTJ0+qSZMm5e7XVwQFBWn69Onavn27MjIylJGRoV27dumNN95Q+/btvd2eTyr4LMjKylLz5s3LvN6OHTsKPztw7qKjo5WTk+PtNqoEQkU1kZ+fr2eeeaawXnzxRX3yyScKDQ3Va6+9ps2bNys4ONjbbdYoy5YtU1ZWlsaOHVum8ZdeeqkuvvhiRUVF6fDhwxXcnXdddNFF2rFjh55++mllZ2frnXfe0VtvvaXU1FTde++9+ve//62RI0d6u02fVadOHd1xxx1lGtuvXz917969gjtCTeL161opzyo6OtpycnJKXFa/fn1bsmSJmZm99957ldJPu3btzMxs6tSpXj82TpSZ2dq1a886rn///mZm9tRTTxW+9uGHH5qZWb9+/c66/muvvWZmZjfccIPX33NFlr+/v+3atctyc3Nt1KhRxZaPGTPGDh06ZIMHD/Z6r75W0dHRlpuba0ePHrXdu3eXaZ1FixaZmdnBgwdr9P0VPD3u7j6Da1pxpqKay8zM1IQJE5SRkaGRI0ee9RHwcFZERIQkady4caWO8/f31+jRo5WSkqKoqKjKaM1rrrrqKnXt2lXh4eH64IMPii1fsmSJOnTooC+++MIL3fk+M9OSJUvUqVMnDR48uNSxTZo00ciRI7Vlyxb9+OOPldQhqjNCRQ1w4sQJ7dq1Sw0bNuQnkEq2Zs0aHT58WKNGjSp1Xst1112n5s2bKzIyUrm5uZXYYeW78MILJUnbtm1zO+bEiROV1U61NG/ePEmn5vWUZvz48apfv37heMBThIoaolmzZpL4sK5s2dnZWr58uZo0aaLrr7/e7bjbbrtNkrRw4cLKas1r8vLyJEkdO3b0cifV17Zt27R161aNGDGi8L/9kkycOFHHjx/nuUdwDKGiBujfv7/atm2rrVu3Kj09vciyYcOGac2aNfrtt9908uRJ7d27V++8885ZJ2717dtXy5YtU1JSkk6ePKmEhAS99dZbatu2rdt1GjdurOnTp2vXrl06ceKEDhw4oPDwcJ1//vmOvM+q6mw/gQQFBWno0KHauXOntmzZUpmtecXmzZslSZMnT1avXr283E31NW/ePNWpU0e33357icv79u2rCy+8UEuXLtXx48crtzlUa16f2EF5ViVNEqpdu7Z16tTJHnvsMTt69Kjl5ubaVVddVbi8Vq1ahRO0fv31V5s7d64999xztnTpUsvIyLDs7GybNGlSift7/vnnzcwsNTXV3nvvPXvuueds4cKFlpqaaunp6XbbbbcVm6hZt25d27hxo+Xl5dlHH31kzz//vH300UdmZpaYmGiBgYFeP47uypOJmgW1d+9eO3nypDVp0qTYsnvuucfMzB5//HGvv9fKqk8//dTMzLKzs23OnDnWs2dPr/dUHeqPnwUNGza09PR0i4uLK3HswoULzczssssuM0m2du1aJmo6cNwp7zdAeVjR0dFWmv379xe7ouCll14yM7M5c+ZY3bp1iyxr06aNbdu2zfLy8uyaa64psqzgC3D16tXWuHHjIssaN25s77//vmVlZRULFWPGjCnxC/ehhx6y5cuXl/hlW1XKiVAxc+ZMM7MSg9qmTZssNzfX2rRp4/X3WlnVsGFDmzNnjuXm5hb+Od21a5c98sgjVfrPQlWvM7/cwsPDzcxs4MCBRcYFBgbaiRMn7Icffih8jVDh3HGv4eX1BigPKzo62vLy8mzq1Kk2depUmz59umVlZdnRo0etd+/e5ufnV2R827ZtLTs723744QerVatWidts1aqVHT9+vMiHTkBAgKWmplpycrI1atSoxPX8/Pxs48aNxULF448/bmZW7NHzvlBOhIpu3bqZmdmGDRtKfL0s26+O1aVLF5s5c6YlJCQUhouMjAybPHmy13vzxTrzy61Xr15mZhYZGVlk3P33329mZvfdd1/ha4QK5457DS+vN0B5WCX9gX799dfNzOzvf/97sfEFHyi33HJLqdt95513zMwsNDTUJNnw4cPNzGzKlCmlrnf99dcXCxWDBg0yM7OdO3fa0KFDrXbt2l4/bmUtJ0KFpMKzP+3bty987YUXXjAzs7CwMK+/T2/X5ZdfbitXriwMF+PHj/d6T75WJX0WfPfdd3by5Elr2rRp4Ws//vijnThxosjPjoQKZ497TS0malZT06ZN09GjRzVt2jSdd955RZZ16dJFUumX9EnS1q1bJUldu3Y9p/ViY2OLvfbll19qypQpCgkJKbzMcu3atbr//vvVsGHDsr0pLzl27Jjq1q171nH16tWTdOreICWJiIiQn59f4R02XS6Xxo0bp+PHj2vFihXONeyjYmNjNWLECD399NOSpOnTp3u5o+ph3rx5qlu3buEVRr1799ZFF12kFStW6MiRI95tDtUOoaKaSk1N1QsvvKDg4ODCD+kzWRnv83/muLKud6Z//vOf6ty5sx544AF99tln6tWrl2bNmqXNmzdX6WddHDhwQK1atTrruDZt2kiSkpOTS1y+ZMkS5eXlFV4FctVVV6l169b68MMPudT3D5599lklJiaqXbt2xQIxzl1ERIROnDihiRMnSpImTZokSdybAhWCUFGNvfrqq0pMTNS9995b5J4AP//8syTpL3/5S6nrFywvGF/W9f7jP/7D7bLExES9/vrruummm9S8eXO9+OKLCg0N1ZQpU87+hrxk165d6tChQ2FocOeKK66QdPoMz5kOHjyoL7/8UqGhoerZs2fh/zkuWLDA2YargaSkJEmnzubAM0ePHtUHH3yg0NBQDR8+XH/7298UFxen9evXe7s1VFNe/w2G8qxK+z0vLCzMzMw+/PDDwtfat29v2dnZ9t133xWbxFlQLVu2tPT0dPvxxx8LXwsICLDDhw9bUlKSNWzYsMT1/Pz8bMOGDcXmVEinnkNy5vgGDRqYmVlUVJTXj6O7Gjt2rJmZzZo1y+2YDh06FJtNX1KNHz/ezMzmz59vGRkZlpiYaC6Xy+vvsbIrLCzM7SThoKAgS09Pt59//tnrffpaufss6NOnj5mZpaWlmZnZo48+WmwMcyqcP+41tLzeAOVhlfYH2uVy2bZt28zMbMCAAYWvv/zyy2ZmtmDBAmvQoEGRdbp06WLfffed5eXl2ZAhQ4ose+CBB8zM7JNPPrFmzZoVWRYcHGwffvihnThxolioeOKJJywxMdE6depUZJ2bb77ZzMz+9a9/ef04uis/P7/CY/jggw8WW96hQwfbuXOn5eXl2XXXXVfqtgICAuz48eOFkxFnzpzp9fdX2VXwZy8yMtICAgKKLAsKCrJ169YxUbOcVdpnwU8//WRmZllZWcX+25UIFRV13Gta+QvVmpnpv/7rv/T555/rlVdeUY8ePWRmmjJlis4//3yNHz9eAwcOVFRUlNLS0hQSEqLhw4erTp06uu+++/TJJ58U2d6sWbPUrl07PfTQQ9q5c6dWr16tpKQktWnTRtdff73q1Kmj8ePHF3tQ1Pfff68WLVpo+/bt+t///V/9+uuv6tatm4YOHark5GS9+OKLlXlYzkl+fr6GDx+utWvX6tVXX9XEiRP1+eef6/jx4+revbuuvfZa+fv769FHH9XHH39c6rYyMjK0evVqjR49WlLNuC33mV544QVdccUVuvXWW3XNNddozZo1OnDggNq0aaNhw4apUaNGeuaZZ2rksalI4eHhevXVV7Vq1Sr9/vvv3m6n2vHz89PUqVPdLj9y5IhmzZpViR15j9eTDeVZlSUlf/zxx2ZmdueddxZ5fdiwYRYVFWW///67ZWVl2b59++ydd96xCy+8sNTt9e/f35YvX27Jycl28uRJS0hIsH/961/Wpk0ba9GiRYk/f/zpT3+ypUuXWnJysmVmZlp8fLzNmTPHQkJCvH4My1INGjSwKVOm2JYtWyw1NdUyMzNt7969tmjRIuvVq1eZtzNkyBAzM9u0aZPX35O3yuVy2fjx423NmjWWnJxsWVlZ9uuvv1pERIT99a9/9Xp/vlqlfRY0adLEMjMzi9xZ94/FmQrPjvvZJCQkeL3PyijX//8NAACAR7j6AwAAOIJQAQAAHEGoAAAAjiBUAAAARxAqAACAIwgVAADAEYQKAADgCEIFAABwBKECAAA4glABAAAcQagAAACOIFQAAABHECoAAIAj/g+xg6UL3Cu5PQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x360 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# now plot a single gene\n",
    "genename = 'lrrfip1a'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "# fH, ax = plt.subplots(figsize= [10,6.6666])\n",
    "# fH, ax = plt.subplots(figsize= [5,3.3333])\n",
    "fH, ax = plt.subplots(figsize= [8,5]) # for paper\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "ax.set_ylim([0,60]);\n",
    "ax.set_yticklabels(['{:,.0f}'.format(x) for x in ax.get_yticks()])\n",
    "dPath = \"/Users/angueyraaristjm/Desktop/ForLaura/\"\n",
    "# dPath = \"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/\"\n",
    "plt.savefig(dPath + genename + \".svg\", transparent=True, format=\"svg\")\n",
    "# plt.savefig(dPath + genename + \".png\", transparent=False, format=\"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sparkline\n",
    "genename = 'auts2b'\n",
    "scalar= 3;\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25*scalar,0.5*scalar])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=14, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=8, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "dPath = \"/Users/angueyraaristjm/Documents/LiLab/Applications/2021_FacultyApps/2021_06_CIBR/resources/\"\n",
    "plt.savefig(dPath + genename + \".svg\", transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "8/(12/10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename='nr2f6b'\n",
    "# plot developmental course\n",
    "def plotDevCourse(genename):\n",
    "    gdf = findgene_zfHoang2020(genename)\n",
    "    n = np.arange(0,5) #PRP, ePR, mPR, lPR, aPR, lR\n",
    "    a = 0; #for avg. expression\n",
    "    a = 12; #for pct. expression\n",
    "    h_start = 2 + a\n",
    "    h_end = 7 + a\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def formatFigureDevCourse(genename, figH, axH, plotH):\n",
    "    [font_path, fontTicks, fontLabels, fontTitle] = defaultFonts();\n",
    "    axH.set_xticks([0,1,2,3,4])\n",
    "    axH.set_xticklabels(['PRP','ePR','mPR','lPR','aPR']);\n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,2))\n",
    "#     axH.set_ylabel('norm. counts', fontproperties=fontLabels)\n",
    "    axH.set_ylabel('avg. expression', fontproperties=fontLabels)\n",
    "    axH.set_title(genename, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "\n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 24)\n",
    "    axH.yaxis.offsetText.set_fontsize(24)\n",
    "\n",
    "n,h = plotDevCourse(genename)\n",
    "fH, ax = plt.subplots(figsize= [12,8])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=zfHC['plt'])\n",
    "formatFigureDevCourse(genename, fH, ax, pH)\n",
    "\n",
    "dPath = \"/Users/angueyraaristjm/Desktop/ForLaura/\"\n",
    "# plt.savefig(dPath + genename + \"_dev.svg\", transparent=True, format=\"svg\")\n",
    "plt.savefig(dPath + genename + \"_dev.png\", transparent=True, format=\"png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=12\n",
    "gdf = findgene_zfHoang2020(genename)\n",
    "gdf.iloc[0,2+a:7+a]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genenames = gdf['symbol'].values\n",
    "# samplenames = [] #skipping names to replace in intaglio\n",
    "# data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# genenamesr = rdf['symbol'].values\n",
    "# samplenamesr = [] #skipping names to replace in intaglio\n",
    "# datar = rdf.iloc[0:,7:16]\n",
    "# genenamesm = mdf['symbol'].values\n",
    "# samplenamesm = [] #skipping names to replace in intaglio\n",
    "# datam = mdf.iloc[0:,5:41]\n",
    "\n",
    "# genenamessq = sqdf['symbol'].values\n",
    "# samplenamessq = [] #skipping names to replace in intaglio\n",
    "# datasq = sqdf.iloc[0:,8:21]\n",
    "\n",
    "# figr = plt.figure(figsize=[8,24],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(3,2)\n",
    "# ax = figr.add_subplot(fig_grid[0,:])\n",
    "# axr = figr.add_subplot(fig_grid[1,0])\n",
    "# axm = figr.add_subplot(fig_grid[1,1])\n",
    "# axsq = figr.add_subplot(fig_grid[2,:])\n",
    "# # axm = plt.subplot(2,2,2)\n",
    "# im, cbar = heatmap(data, genenames, samplenames, ax=ax,cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr, cmap=\"inferno\", cbarlabel=\"cpm\")\n",
    "# imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm, cmap=\"inferno\", cbarlabel=\"tmm\")\n",
    "# imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq, cmap=\"inferno\", cbarlabel=\"counts\")\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Saving heatmaps for Hoang adult photoreceptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneTag = 'xbp1'\n",
    "gdf = findgenelist(geneTag)\n",
    "\n",
    "zfahdf = findgenelist_zfAdultHoang2020(geneTag)\n",
    "zfahdf = zfahdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# zfahdf = zfahdf.drop([2])\n",
    "zfahdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMapHoang_' + geneTag\n",
    "\n",
    "genenameszfah = zfahdf['symbol'].values\n",
    "samplenameszfah = [] #skipping names to replace in intaglio\n",
    "datazfah = zfahdf.iloc[0:,2+8:9+8]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[8,18])\n",
    "imzfah, cbarzfah = heatmap2_zfadultHoang2020(datazfah, genenameszfah, samplenameszfah, ax=ax)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Saving heatmaps for Ogawa adult photoreceptors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneTag = 'xbp1'\n",
    "gdf = findgenelist(geneTag)\n",
    "\n",
    "odf = findgenelist_zfOgawa2021(geneTag)\n",
    "odf = odf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# zfahdf = zfahdf.drop([2])\n",
    "odf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMapOgawa_' + geneTag\n",
    "\n",
    "genenameso = odf['symbol'].values\n",
    "samplenameso = [] #skipping names to replace in intaglio\n",
    "datao = odf.iloc[0:,2:10].values\n",
    "figo, axo = plt.subplots(figsize=[8,8])\n",
    "imo, cbaro = heatmap_zfOgawa2021(datao, genenameso, samplenameso, ax=axo)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## sparkline plot for markdown file<a id='sparkline'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'xbp1'\n",
    "scalar= 3;\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25*scalar,.25*scalar])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'{0}fpkm'.format(np.around(np.mean(h),decimals=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genename = 'kcnk1'\n",
    "# n,h = plotBars_sqcones(genename)\n",
    "# fH, ax = plt.subplots(figsize=[2.2,.25])\n",
    "# ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt')\n",
    "# pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=sqC['plt'])\n",
    "# ax.text(.92, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6)\n",
    "# ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=4)\n",
    "# ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}sq.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## PCA <a id='pca'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is supposed to run only once. Creates sparklines and a markdown file with entries for these genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/\"\n",
    "# fname = \"fromPython_PCgenes_RvC_fromPython.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/\"\n",
    "# fname = \"fromPython_PCgenes_USvLM.md\"\n",
    "\n",
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/\"\n",
    "fname = \"fromPython_PCgenes_LvM.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/\"\n",
    "# fname = \"fromPython_PCgenes_UvS.md\"\n",
    "\n",
    "\n",
    "\n",
    "pcadir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/PCA_genes/\"\n",
    "%cd {pcadir}\n",
    "pcr = pd.read_csv(deseqdir + \"pcaWeights.csv\")\n",
    "pcr.sort_values(by=['PC1'])\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparkline plots\n",
    "genename = 'spock3'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# nTop = 100;\n",
    "# print(\"Started...\")\n",
    "# for i in range(nTop):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Up plots created...\")\n",
    "\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Down plots created...\")\n",
    "        \n",
    "# fid = open(fname, 'w+t') \n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in rods \\n\\n\\n\")\n",
    "# for i in range(nTop):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(i+1) + \" ![R](./resources/R.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "\n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in cones \\n\\n\\n\")\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(-i) + \" ![Cs](./resources/Cs.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "# fid.close()\n",
    "\n",
    "\n",
    "# print(\"MD file created. \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(plt.subplots_adjust)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# histogram of PC weights\n",
    "pcr.sort_values(by=['PC1'])\n",
    "a = pcr['PC1']\n",
    "# a = np.divide(a,pcr['PC1'].sum())\n",
    "a = a[a != 0] #there are too many nil-values\n",
    "fH, ax = plt.subplots(figsize=[12,6])\n",
    "n, h, patches = ax.hist(np.multiply(a,100), bins=100, log=True) #use bins='fd' for optimal binning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Quantification<a id='quant'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### DEGs by PCA groups\n",
    "> noteshere\n",
    "\n",
    "``` bash\n",
    "# code to add symbol to first column header\n",
    "echo -e \"symbol$(cat LvsM_shrinkFC.csv)\" > LvsM_shrinkFC.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.2 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.58 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\")\n",
    "gdf = gdf[~gdf['symbol'].str.contains('mt-', na=False)] # removing mitochondrial genes for figure\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# thresholdCol = 'log2FoldChange'\n",
    "thresholdCol = 'baseMean'\n",
    "# thresholdCol = 'padj'\n",
    "# thresholdCol = 'pvalue'\n",
    "\n",
    "\n",
    "sortCol = 'log2FoldChange'\n",
    "\n",
    "topG = 25\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/RvsC_rawFC.csv\")\n",
    "Rup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Rup = Rup.drop(['stat'], axis = 1)\n",
    "Rup = Rup.set_index('symbol')\n",
    "Rup = Rup.reset_index()\n",
    "Rup = Rup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Rup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Rup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Rup = pd.concat([Rup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Rup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Rup = Rup[~nanrows]\n",
    "Rup = Rup.sort_values(by=sortCol, ascending = False)\n",
    "\n",
    "\n",
    "Cup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<-fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Cup = Cup.drop(['stat'], axis = 1)\n",
    "Cup = Cup.set_index('symbol')\n",
    "Cup = Cup.reset_index()\n",
    "Cup = Cup.iloc[0:topG+9,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Cup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Cup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Cup = pd.concat([Cup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Cup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Cup = Cup[~nanrows]\n",
    "Cup = Cup.sort_values(by=sortCol, ascending = True)\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/USvsML_rawFC.csv\")\n",
    "MLup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<-fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "MLup = MLup.drop(['stat'], axis = 1)\n",
    "MLup = MLup.set_index('symbol')\n",
    "MLup = MLup.reset_index()\n",
    "MLup = MLup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(MLup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=MLup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "MLup = pd.concat([MLup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(MLup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "MLup = MLup[~nanrows]\n",
    "MLup = MLup.sort_values(by=sortCol, ascending = True)\n",
    "\n",
    "USup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "USup = USup.drop(['stat'], axis = 1)\n",
    "USup = USup.set_index('symbol')\n",
    "USup = USup.reset_index()\n",
    "USup = USup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(USup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=USup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "USup = pd.concat([USup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(USup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "USup = USup[~nanrows]\n",
    "USup = USup.sort_values(by=sortCol, ascending = False)\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/UvsS_rawFC.csv\")\n",
    "Uup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Uup = Uup.drop(['stat'], axis = 1)\n",
    "Uup = Uup.set_index('symbol')\n",
    "Uup = Uup.reset_index()\n",
    "Uup = Uup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Uup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Uup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Uup = pd.concat([Uup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Uup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Uup = Uup[~nanrows]\n",
    "Uup = Uup.sort_values(by=sortCol, ascending = False)\n",
    "\n",
    "Sup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<-fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Sup = Sup.drop(['stat'], axis = 1)\n",
    "Sup = Sup.set_index('symbol')\n",
    "Sup = Sup.reset_index()\n",
    "Sup = Sup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Sup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Sup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Sup = pd.concat([Sup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Sup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Sup = Sup[~nanrows]\n",
    "Sup = Sup.sort_values(by=sortCol, ascending = True)\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/LvsM_rawFC.csv\")\n",
    "Mup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Mup = Mup.drop(['stat'], axis = 1)\n",
    "Mup = Mup.set_index('symbol')\n",
    "Mup = Mup.reset_index()\n",
    "Mup = Mup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Mup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Mup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Mup = pd.concat([Mup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Mup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Mup = Mup[~nanrows]\n",
    "Mup = Mup.sort_values(by=sortCol, ascending = False)\n",
    "\n",
    "Lup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<-fcLim)].sort_values(by=thresholdCol, ascending=False)\n",
    "Lup = Lup.drop(['stat'], axis = 1)\n",
    "Lup = Lup.set_index('symbol')\n",
    "Lup = Lup.reset_index()\n",
    "Lup = Lup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Lup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Lup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Lup = pd.concat([Lup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Lup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Lup = Lup[~nanrows]\n",
    "Lup = Lup.sort_values(by=sortCol, ascending = True)\n",
    "\n",
    "\n",
    "print('Imported all data frames')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Cup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # # trying to sort by baseMean but only for relevant comparison\n",
    "# DS = pd.read_csv(\"00_bySubtype/UvsS_rawFC.csv\")\n",
    "# Uup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)] #.sort_values(by=thresholdCol, ascending=False)\n",
    "# Uup = Uup.drop(['stat'], axis = 1)\n",
    "# Uup = Uup.set_index('symbol')\n",
    "# Uup = Uup.reset_index()\n",
    "\n",
    "# tempdf = gdf[gdf[\"symbol\"].isin(Uup[\"symbol\"])]\n",
    "# tempdf = tempdf.set_index('symbol')\n",
    "# tempdf = tempdf.loc[~tempdf.index.duplicated()] #removing some of the duplicated genes\n",
    "# tempdf = tempdf.reindex(index=Uup['symbol'])\n",
    "# tempdf = tempdf.reset_index()\n",
    "\n",
    "# Uup = pd.concat([Uup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "# nanrows = np.isnan(Uup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "# Uup = Uup[~nanrows]\n",
    "# Uup['sortingColumn']=Uup.iloc[0:,13:18].sum(axis=1)\n",
    "# Uup = Uup.sort_values(by='sortingColumn', ascending = False)\n",
    "# Uup = Uup.iloc[0:topG,:]\n",
    "# # Uup = Uup.sort_values(by=sortCol, ascending = False)\n",
    "\n",
    "# Sup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<-fcLim)] #.sort_values(by=thresholdCol, ascending=False)\n",
    "# Sup = Sup.drop(['stat'], axis = 1)\n",
    "# Sup = Sup.set_index('symbol')\n",
    "# Sup = Sup.reset_index()\n",
    "\n",
    "\n",
    "# tempdf = gdf[gdf[\"symbol\"].isin(Sup[\"symbol\"])]\n",
    "# tempdf = tempdf.set_index('symbol')\n",
    "# tempdf = tempdf.loc[~tempdf.index.duplicated()] #removing some of the duplicated genes\n",
    "# tempdf = tempdf.reindex(index=Sup['symbol'])\n",
    "# tempdf = tempdf.reset_index()\n",
    "\n",
    "# Sup = pd.concat([Sup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "# nanrows = np.isnan(Sup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "# Sup = Sup[~nanrows]\n",
    "# Sup['sortingColumn']=Sup.iloc[0:,18:24].sum(axis=1)\n",
    "# Sup = Sup.sort_values(by='sortingColumn', ascending = False)\n",
    "# Sup = Sup.iloc[0:topG,:]\n",
    "# # Sup = Sup.sort_values(by=sortCol, ascending = True)\n",
    "\n",
    "# Sup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using row normalization (and greys color map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'top' + str(topG) + \"RvC\" + \"_norm\"\n",
    "samplenames = [] #skipping names\n",
    "genenames = np.concatenate([Rup['symbol'].values,\n",
    "                            Cup['symbol'].values,])\n",
    "\n",
    "#normalized data to max\n",
    "data = np.concatenate([Rup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,\n",
    "                       Cup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"fpkm (normalized)\")\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LMvSU\" + \"_norm\"\n",
    "samplenames = [] #skipping names\n",
    "genenames = np.concatenate([MLup['symbol'].values,\n",
    "                            USup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([MLup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,\n",
    "                      USup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"fpkm (normalized)\")\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LvM\" + \"_norm\"\n",
    "samplenames = [] #skipping names\n",
    "genenames = np.concatenate([Lup['symbol'].values,\n",
    "                            Mup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Lup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,\n",
    "                       Mup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"fpkm (normalized)\")\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'top' + str(topG) + \"SvU\" + \"_norm\"\n",
    "samplenames = [] #skipping names\n",
    "genenames = np.concatenate([Sup['symbol'].values,\n",
    "                            Uup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Sup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,\n",
    "                       Uup.iloc[0:25,7:37].apply(lambda x: x/x.max(), axis=1).values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"fpkm (normalized)\")\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## using z-score mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'top' + str(topG) + \"RvC\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Rup['symbol'].values,\n",
    "                            Cup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Rup.iloc[0:25,7:37].values,\n",
    "                       Cup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LMvSU\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([MLup['symbol'].values,\n",
    "                            USup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([MLup.iloc[0:25,7:37].values,\n",
    "                      USup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LvM\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Lup['symbol'].values,\n",
    "                            Mup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Lup.iloc[0:25,7:37].values,\n",
    "                       Mup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'top' + str(topG) + \"SvU\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Sup['symbol'].values,\n",
    "                            Uup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Sup.iloc[0:25,7:37].values,\n",
    "                       Uup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using vst for heatmaps \n",
    "\n",
    "> does make it look better because it mitigates variance between samples...\n",
    "> but is it the right thing to do?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vst = pd.read_csv(\"00_bySubtype/RvsC_vstData.csv\")\n",
    "vst.iloc[0:,1:31]\n",
    "vst[vst[\"symbol\"].isin([\"rho\",\"rhol\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sortCol = 'log2FoldChange'\n",
    "sortCol = 'baseMean'\n",
    "# sortCol = 'padj'\n",
    "topG = 25\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/RvsC_rawFC.csv\")\n",
    "Rup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Rup = Rup.drop(['stat'], axis = 1)\n",
    "Rup = Rup.set_index('symbol')\n",
    "Rup = Rup.reset_index()\n",
    "Rup = Rup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Rup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Rup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Rup = pd.concat([Rup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Rup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Rup = Rup[~nanrows]\n",
    "\n",
    "Cup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Cup = Cup.drop(['stat'], axis = 1)\n",
    "Cup = Cup.set_index('symbol')\n",
    "Cup = Cup.reset_index()\n",
    "Cup = Cup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Cup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Cup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Cup = pd.concat([Cup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Cup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Cup = Cup[~nanrows]\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/USvsML_rawFC.csv\")\n",
    "MLup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "MLup = MLup.drop(['stat'], axis = 1)\n",
    "MLup = MLup.set_index('symbol')\n",
    "MLup = MLup.reset_index()\n",
    "MLup = MLup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(MLup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=MLup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "MLup = pd.concat([MLup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(MLup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "MLup = MLup[~nanrows]\n",
    "\n",
    "USup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "USup = USup.drop(['stat'], axis = 1)\n",
    "USup = USup.set_index('symbol')\n",
    "USup = USup.reset_index()\n",
    "USup = USup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(USup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=USup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "USup = pd.concat([USup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(USup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "USup = USup[~nanrows]\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/UvsS_rawFC.csv\")\n",
    "Uup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Uup = Uup.drop(['stat'], axis = 1)\n",
    "Uup = Uup.set_index('symbol')\n",
    "Uup = Uup.reset_index()\n",
    "Uup = Uup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Uup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Uup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Uup = pd.concat([Uup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Uup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Uup = Uup[~nanrows]\n",
    "\n",
    "Sup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Sup = Sup.drop(['stat'], axis = 1)\n",
    "Sup = Sup.set_index('symbol')\n",
    "Sup = Sup.reset_index()\n",
    "Sup = Sup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Sup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Sup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Sup = pd.concat([Sup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Sup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Sup = Sup[~nanrows]\n",
    "\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/LvsM_rawFC.csv\")\n",
    "Mup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Mup = Mup.drop(['stat'], axis = 1)\n",
    "Mup = Mup.set_index('symbol')\n",
    "Mup = Mup.reset_index()\n",
    "Mup = Mup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Mup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Mup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Mup = pd.concat([Mup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Mup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Mup = Mup[~nanrows]\n",
    "\n",
    "Lup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Lup = Lup.drop(['stat'], axis = 1)\n",
    "Lup = Lup.set_index('symbol')\n",
    "Lup = Lup.reset_index()\n",
    "Lup = Lup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = vst[vst[\"symbol\"].isin(Lup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Lup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Lup = pd.concat([Lup, tempdf.iloc[0:,1:31]], axis=1)\n",
    "nanrows = np.isnan(Lup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Lup = Lup[~nanrows]\n",
    "\n",
    "# Lup\n",
    "\n",
    "print('Imported all data frames')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'top' + str(topG) + \"RvC\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Rup['symbol'].values,\n",
    "                            Cup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Rup.iloc[0:25,7:37].values,\n",
    "                       Cup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LMvSU\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([MLup['symbol'].values,\n",
    "                            USup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([MLup.iloc[0:25,7:37].values,\n",
    "                      USup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'top' + str(topG) + \"LvM\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Lup['symbol'].values,\n",
    "                            Mup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Lup.iloc[0:25,7:37].values,\n",
    "                       Mup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'top' + str(topG) + \"SvU\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Sup['symbol'].values,\n",
    "                            Uup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Sup.iloc[0:25,7:37].values,\n",
    "                       Uup.iloc[0:25,7:37].values,])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Exploring Phototransduction genes<a id='Phototrans'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)  \n",
    "[Go to Known TFs](#knownTFs)  \n",
    "[Go to RNAseq explorer](#heatOther)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For manual collection dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "tflist = pd.read_csv(\"00_bySubtype/phototransGenes.csv\") # get list of all PTG\n",
    "# tflist = pd.read_csv(\"00_bySubtype/phototransGenes_RvC.csv\") # get list of all PTG reordered by rods vs. cones\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract PTG data only\n",
    "# tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove PTGs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract PTG data only\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of phototransduction genes\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "tempdf = tfdf.copy()\n",
    "# fixing duplication of rgs9bp\n",
    "tempdf.iloc[2,7:]=tempdf.iloc[2,7:]+tempdf.iloc[46,7:]\n",
    "tempdf.iloc[2,3] = np.mean(tempdf.iloc[2,7:])\n",
    "tempdf.iloc[2,1]=tempdf.iloc[46,1]\n",
    "# tempdf.reindex()\n",
    "# tempdf = tempdf.drop_duplicates()\n",
    "tempdf.iloc[46,0]='duplicate'\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.drop(['duplicate'])\n",
    "tempdf = tempdf.reindex(tflist['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "tempdf\n",
    "\n",
    "\n",
    "\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMap_phototrans'\n",
    "# plotName = 'heatMap_phototrans_RvC'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tempdf['symbol'].values\n",
    "\n",
    "data = tempdf.iloc[0:,7:].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### replicate heatmap for Sun's rod data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "rdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRods_Sun2018/zfRods.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/phototransGenes.csv\") # get list of all PTG\n",
    "rdf = rdf[rdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract PTG data only\n",
    "# tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove PTGs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of phototransduction genes\".format(np.size(rdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "temprdf = rdf.copy()\n",
    "# fixing duplication of rgs9bp\n",
    "temprdf.iloc[40,7:]=temprdf.iloc[40,7:]+temprdf.iloc[41,7:]\n",
    "temprdf.iloc[40,3] = np.mean(temprdf.iloc[40,7:])\n",
    "temprdf.iloc[40,1]=temprdf.iloc[41,1]\n",
    "temprdf.iloc[41,0]='duplicate'\n",
    "temprdf = temprdf.set_index('symbol')\n",
    "temprdf = temprdf.drop(['duplicate'])\n",
    "temprdf = temprdf.reset_index()\n",
    "# fixing duplication of gnb1b\n",
    "temprdf.iloc[9,7:]=temprdf.iloc[9,7:]+temprdf.iloc[10,7:]\n",
    "temprdf.iloc[9,3] = np.mean(temprdf.iloc[9,7:])\n",
    "temprdf.iloc[9,1]=temprdf.iloc[10,1]\n",
    "temprdf.iloc[10,0]='duplicate'\n",
    "temprdf = temprdf.set_index('symbol')\n",
    "temprdf = temprdf.drop(['duplicate'])\n",
    "temprdf = temprdf.reindex(tflist['symbol'])\n",
    "temprdf = temprdf.reset_index()\n",
    "temprdf\n",
    "\n",
    "# tflist['symbol']\n",
    "\n",
    "\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMap_phototrans_SunRods'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tempdf['symbol'].values\n",
    "\n",
    "data = temprdf.iloc[0:,7:].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_rods(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = tempdf[['symbol','log2FoldChange']].copy()\n",
    "# temp.set_index('symbol')\n",
    "temp2 = temprdf[['symbol','log2FoldChange']].copy()\n",
    "# temp2.set_index('symbol')\n",
    "temp = temp.merge(temp2, left_on='symbol', right_on='symbol')\n",
    "temp = temp.fillna(0)\n",
    "# temp.head(50)\n",
    "\n",
    "plt.scatter(temp['log2FoldChange_x'], temp['log2FoldChange_y'])\n",
    "# df1.merge(df2, left_on='lkey', right_on='rkey')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### replicate heatmap for Hoang's 10x data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "hdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRet_HoangBlackshaw2020/conesDev_Hoang.csv\") # get data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/phototransGenes.csv\") # get list of all PTG\n",
    "hdf = hdf[hdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract PTG data only\n",
    "# tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove PTGs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of phototransduction genes\".format(np.size(rdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "temphdf = hdf.copy()\n",
    "# temphdf.set_index('symbol').index.duplicated()\n",
    "temphdf = temphdf.set_index('symbol')\n",
    "temphdf = temphdf.reindex(tflist['symbol'])\n",
    "temphdf = temphdf.reset_index()\n",
    "temphdf\n",
    "\n",
    "# tflist['symbol']\n",
    "\n",
    "\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMap_phototrans_HoangAdultPhoto'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tempdf['symbol'].values\n",
    "\n",
    "data = temphdf.iloc[0:,8:13].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_zfadultHoang2020(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### replicate heatmap for Ogawa's FACS+10x data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "odf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRet_Ogawa2021/photoreceptors_Ogawa.csv\") # get data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/phototransGenes.csv\") # get list of all PTG\n",
    "odf = odf[odf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract PTG data only\n",
    "# tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove PTGs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of phototransduction genes\".format(np.size(odf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "tempodf = odf.copy()\n",
    "# tempodf.set_index('symbol').index.duplicated()\n",
    "tempodf = tempodf.set_index('symbol')\n",
    "tempodf = tempodf.reindex(tflist['symbol'])\n",
    "tempodf = tempodf.reset_index()\n",
    "tempodf\n",
    "\n",
    "# tflist['symbol']\n",
    "\n",
    "\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'heatMap_phototrans_Ogawa'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tempdf['symbol'].values\n",
    "\n",
    "data = tempodf.iloc[0:,2:10].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_zfOgawa2021(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "angueyraR = tempdf[tempdf['symbol'] == 'rho'].iloc[:,7:13].values\n",
    "angueyraC = tempdf[tempdf['symbol'] == 'rho'].iloc[:,14:37].values\n",
    "\n",
    "sunR = temprdf[temprdf['symbol'] == 'rho'].iloc[:,7:11].values\n",
    "sunC = temprdf[temprdf['symbol'] == 'rho'].iloc[:,12:].values\n",
    "\n",
    "hoangR = temprdf[temprdf['symbol'] == 'rho'].iloc[:,8].values\n",
    "hoangC = temprdf[temprdf['symbol'] == 'rho'].iloc[:,9:13].values\n",
    "\n",
    "ogawaR = tempodf[tempodf['symbol'] == 'rho'].iloc[:,2].values\n",
    "ogawaC = tempodf[tempodf['symbol'] == 'rho'].iloc[:,3:10].values\n",
    "\n",
    "# np.mean(angueyraR)/np.mean(sunR) # relative depth of FACS rods, crappy marker\n",
    "\n",
    "# plotting relative values\n",
    "plotName = 'rhoContamination'\n",
    "fH, axH = plt.subplots(figsize= [8,8])\n",
    "\n",
    "barW = 0.8; # bar width\n",
    "barD = barW + .1;\n",
    "# barP = -np.size(geneList)+1; # position in x-axis\n",
    "barStep = 1; # distance between photoreceptor groups\n",
    "barPos = []; #array to save bar positions\n",
    "\n",
    "i = 0;\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(angueyraC)/np.mean(angueyraR), width=barW, color='#000000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(sunC)/np.mean(sunR), width=barW, color='#000000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(hoangC)/np.mean(hoangR), width=barW, color='#000000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(ogawaC)/np.mean(ogawaR), width=barW, color='#000000');\n",
    "\n",
    "fontLabels = formatFigure_General('', fH, axH, pH)\n",
    "    \n",
    "axH.set_xticks([1,2,3,4]);\n",
    "axH.set_xticklabels(['Angueyra et al., 2021','Sun et al., 2017', 'Hoang et al., 2020', 'Ogawa and Corbo, 2021'],rotation=45, ha='right')\n",
    "axH.set_ylim([0,0.8]); # this was 400\n",
    "\n",
    "axH.set_ylabel('Relative reads for $\\itrho$\\n in non-rod samples', fontproperties=fontLabels)\n",
    "\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting abolsute values\n",
    "plotname = ''\n",
    "fH, axH = plt.subplots(figsize= [12,8])\n",
    "\n",
    "barW = 0.25; # bar width\n",
    "barD = barW + .1;\n",
    "# barP = -np.size(geneList)+1; # position in x-axis\n",
    "barStep = 1; # distance between photoreceptor groups\n",
    "barPos = []; #array to save bar positions\n",
    "\n",
    "i = 0;\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(angueyraR)/np.mean(angueyraR), width=barW, color='#000000');\n",
    "pH = plt.bar([barStep*i+barD], np.mean(angueyraC)/np.mean(angueyraR), width=barW, color='#FF0000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(sunR)/np.mean(sunR), width=barW, color='#000000');\n",
    "pH = plt.bar([barStep*i+barD], np.mean(sunC)/np.mean(sunR), width=barW, color='#FF0000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(hoangR)/np.mean(hoangR), width=barW, color='#000000');\n",
    "pH = plt.bar([barStep*i+barD], np.mean(hoangC)/np.mean(hoangR), width=barW, color='#FF0000');\n",
    "i = i+1;\n",
    "pH = plt.bar([barStep*i], np.mean(ogawaR)/np.mean(ogawaR), width=barW, color='#000000');\n",
    "pH = plt.bar([barStep*i+barD], np.mean(ogawaC)/np.mean(ogawaR), width=barW, color='#FF0000');"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### make plot with average number of reads for each study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### repeat this for a few genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneList = pd.DataFrame({'symbol':['tbx2a','tbx2b','foxq2']})\n",
    "\n",
    "gdf = pd.read_csv(\"/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "gdf = gdf[gdf[\"symbol\"].isin(geneList[\"symbol\"])] # use as logical to extract PTG data only\n",
    "gdf = gdf.set_index('symbol')\n",
    "gdf = gdf.reindex(geneList['symbol'])\n",
    "gdf = gdf.reset_index()\n",
    "\n",
    "\n",
    "rdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRods_Sun2018/zfRods.csv\") # get fpkm data\n",
    "rdf = rdf[rdf[\"symbol\"].isin(geneList[\"symbol\"])] # use as logical to extract PTG data only\n",
    "rdf = rdf.set_index('symbol')\n",
    "rdf = rdf.reindex(geneList['symbol'])\n",
    "rdf = rdf.reset_index()\n",
    "\n",
    "hdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRet_HoangBlackshaw2020/conesDev_Hoang.csv\") # get data\n",
    "hdf = hdf[hdf[\"symbol\"].isin(geneList[\"symbol\"])] # use as logical to extract PTG data only\n",
    "hdf = hdf.set_index('symbol')\n",
    "hdf = hdf.reindex(geneList['symbol'])\n",
    "hdf = hdf.reset_index()\n",
    "\n",
    "odf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRet_Ogawa2021/photoreceptors_Ogawa.csv\") # get data\n",
    "odf = odf[odf[\"symbol\"].isin(geneList[\"symbol\"])] # use as logical to extract PTG data only\n",
    "odf = odf.set_index('symbol')\n",
    "odf = odf.reindex(geneList['symbol'])\n",
    "odf = odf.reset_index()\n",
    "\n",
    "gdf\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samplenames = []\n",
    "genenames = gdf['symbol'].values\n",
    "\n",
    "plotName = 'heatMap_F0tfs'\n",
    "data = gdf.iloc[0:,7:].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "fig, ax = plt.subplots(figsize=[16,4])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'heatMap_F0tfs_SunRods'\n",
    "data = rdf.iloc[0:,7:].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "fig, ax = plt.subplots(figsize=[16,4])\n",
    "im, cbar = heatmap_rods(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'heatMap_F0tfs_HoangAdultPhoto'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = hdf['symbol'].values\n",
    "data = hdf.iloc[0:,8:13].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "fig, ax = plt.subplots(figsize=[16,4])\n",
    "im, cbar = heatmap_zfadultHoang2020(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'heatMap_F0tfs_Ogawa'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = odf['symbol'].values\n",
    "data = odf.iloc[0:,2:10].apply(lambda x: x/x.max(), axis=1).values\n",
    "# data = tempdf.iloc[0:,7:].apply(lambda x: zscore(x), axis=1).values\n",
    "data = np.nan_to_num(data,0)\n",
    "fig, ax = plt.subplots(figsize=[16,4])\n",
    "im, cbar = heatmap_zfOgawa2021(data, genenames, samplenames, ax=ax, cbarlabel=\"norm.\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "angueyraR = tempdf[tempdf['symbol'] == 'rho'].iloc[:,7:13].values\n",
    "angueyraC = tempdf[tempdf['symbol'] == 'rho'].iloc[:,14:37].values\n",
    "\n",
    "sunR = temprdf[temprdf['symbol'] == 'rho'].iloc[:,7:11].values\n",
    "sunC = temprdf[temprdf['symbol'] == 'rho'].iloc[:,12:].values\n",
    "\n",
    "hoangR = temprdf[temprdf['symbol'] == 'rho'].iloc[:,8].values\n",
    "hoangC = temprdf[temprdf['symbol'] == 'rho'].iloc[:,9:13].values\n",
    "\n",
    "ogawaR = temprdf[temprdf['symbol'] == 'rho'].iloc[:,2].values\n",
    "ogawaC = temprdf[temprdf['symbol'] == 'rho'].iloc[:,3:10].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Exploring TFs<a id='TFs'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)  \n",
    "[Go to Known TFs](#knownTFs)  \n",
    "[Go to RNAseq explorer](#heatOther)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All TFs: ranked by avg. expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "topG = 100\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "\n",
    "plotName = 'TF_top' + str(topG) + 'ByExpression'\n",
    "fH, axH = plt.subplots(figsize=[40,20])\n",
    "\n",
    "# x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "x_pos = np.arange(0,topG,1)\n",
    "\n",
    "pH = plt.bar(x_pos, topExp[\"baseMean\"], width=.8, log=1, color=\"#000000\");\n",
    "\n",
    "knownTFs = ['neurod1','crx','rx1','nr2e3','samd7','rx2','otx5','foxq2','six6b','six7','nfia','meis1b','egr1','thrb','rorb']\n",
    "mask = np.isin(topExp[\"symbol\"],knownTFs)\n",
    "plt.bar(x_pos[mask], topExp[mask][\"baseMean\"], color = '#34A5DA')\n",
    "   \n",
    "plt.xticks(x_pos, topExp[\"symbol\"], rotation=90)\n",
    "\n",
    "formatFigure_General(\"\", fH, axH, pH)\n",
    "plt.ylabel(\"fpkm\")\n",
    "plt.xlabel(\"gene\")\n",
    "plt.xlim([-1,topG+1])\n",
    "plt.ylim([10,1330])\n",
    "# plt.ylim(bottom = 0, top=topG+1)\n",
    "# axH.xaxis.set_tick_params(rotation=66)\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "# plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "\n",
    "\n",
    "plt.show()\n",
    "# topExp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# heatmap after applying initial filters\n",
    "# sortCol = 'log2FoldChange'\n",
    "# sortCol = 'baseMean'\n",
    "sortCol = 'padj'\n",
    "# sortCol = 'pvalue'\n",
    "# topG = 25\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/RvsC_rawFC.csv\") # load DeSEQ2 analysis: rods vs. cones\n",
    "DS = DS[DS[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "Rup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Rup = Rup.drop(['stat'], axis = 1)\n",
    "Rup = Rup.set_index('symbol')\n",
    "Rup = Rup.reset_index()\n",
    "Rup = Rup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Rup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Rup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Rup = pd.concat([Rup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "\n",
    "Cup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']<fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "Cup = Cup.drop(['stat'], axis = 1)\n",
    "Cup = Cup.set_index('symbol')\n",
    "Cup = Cup.reset_index()\n",
    "Cup = Cup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Cup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Cup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Cup = pd.concat([Cup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "\n",
    "plotName = 'TF_RvC_top' + str(topG) + \"DEGnorm\"\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Rup['symbol'].values,\n",
    "                            Cup['symbol'].values,])\n",
    "\n",
    "data = np.concatenate([Rup.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values,\n",
    "                       Cup.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values,])\n",
    "\n",
    "# data = np.concatenate([Rup.iloc[0:,7:37].values,\n",
    "#                        Cup.iloc[0:,7:37].values,])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading lists of TFs and producing heatmap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rods vs Cones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.58 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "tflist = pd.read_csv(\"00_bySubtype/manual_TFs_updatedOct2021.csv\") # get list of all TFs\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/TF_RvC_top20PC1.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "print(\"\\nNon-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# keep sorting of loaded list (useful for PCA)\n",
    "tfdf = tfdf.set_index('symbol')\n",
    "tfdf = tfdf.reindex(index=tflist['symbol'])\n",
    "tfdf = tfdf.reset_index()\n",
    "\n",
    "topG = tfdf.shape[0]\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "# topExp = tfdf.head(topG)\n",
    "\n",
    "nG_rods = len(tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]<fcLim) ])\n",
    "nG_cones = len(tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]>fcLim) ])\n",
    "print(\"Using baseMean>0, padj>{0}, log2FC>{1}\".format(pLim,fcLim))\n",
    "print(\"\\tNumber of rod-enriched genes: {0}\".format(nG_rods))\n",
    "print(\"\\tNumber of cone-enriched genes: {0}\".format(nG_cones))\n",
    "print(\"\\tTotal: {0}\".format(nG_rods+nG_cones))\n",
    "\n",
    "# plotname = ''\n",
    "# fH, axH = plt.subplots(figsize=[10,20])\n",
    "\n",
    "# # x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "# x_pos = np.arange(topG,0,-1)\n",
    "\n",
    "# pH = plt.barh(x_pos, topExp[\"baseMean\"], height=.8, log=1);\n",
    "   \n",
    "# plt.title(\"top expressed TFs\")\n",
    "# plt.yticks(x_pos, topExp[\"symbol\"])\n",
    "\n",
    "# formatFigure_General(\"top expressed TFs\", fH, axH, pH)\n",
    "# plt.xlabel(\"avgFPKM\")\n",
    "# plt.ylabel(\"gene\")\n",
    "# # plt.xlim(right=250)\n",
    "# plt.ylim(bottom = 0, top=topG+1)\n",
    "# # axH.xaxis.set_tick_params(rotation=66)\n",
    "\n",
    "# plt.show()\n",
    "# topExp\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tfdf['symbol'].values\n",
    "\n",
    "# data = tfdf.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values #normalized data to max\n",
    "data = tfdf.iloc[0:,7:37].values\n",
    "\n",
    "plotName = 'TF_RvC_top20PCA'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format('TF_R_manual_updated'), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### (UV+S) _vs._ (M+L) Exploring with code; this is incomplete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.58 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "\n",
    "# exploring\n",
    "# tflist = pd.read_csv(\"00_bySubtype/tempTFs.csv\") # get list of all TFs\n",
    "# tflist = pd.DataFrame({\"symbol\":[\"nfe2l1b\",\"ahr1b\",\"sox6\",\"fgf1b\",\"churc1\",\"eloca\",\"taf13\",\"eaf1\",\"six7\",\"smc3\",\"neurod1\",]})\n",
    "# gdf = pd.read_csv(\"00_USvsLM/USvsLM_raw.csv\") # \n",
    "# gdf = pd.read_csv(\"00_bySubtype/USvsML_rawFC.csv\") #\n",
    "\n",
    "# R vs. C\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "#(US vs. ML)\n",
    "# gdf = pd.read_csv(\"00_bySubtype/TFs_US_rawFC.csv\") # This is the one that ended up working best for paper; unsure how I found stat1a, taf13, churc1 and others as LM enriched. Maybe ignoring padj?\n",
    "# L vs. M\n",
    "# gdf = pd.read_csv(\"00_bySubtype/LvsM_rawFC.csv\") # This is ok, but missing thrb.\n",
    "# UV vs. S\n",
    "# gdf = pd.read_csv(\"00_bySubtype/UvsS_rawFC.csv\") # This is ok, but missing thrb.\n",
    "\n",
    "\n",
    "\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# gdf = pd.read_csv(\"00_LvsM/LvsM_raw.csv\") # \n",
    "# gdf = pd.read_csv(\"00_bySubtype/LvsUSM.csv\") # \n",
    "# gdf = pd.read_csv(\"00_bySubtype/MvsUSL.csv\") # \n",
    "\n",
    "\n",
    "#custom list\n",
    "# tflist = pd.read_csv(\"00_bySubtype/temp.csv\") # get list of all TFs\n",
    "# gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "\n",
    "# gdf = pd.read_csv(\"00_bySubtype/TFs_Groups_rawFC.csv\") # get fpkm data\n",
    "\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/TF_RvC_top20PC1.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.dropna(subset=[\"baseMean\"])\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "print(\"\\nNon-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# keep sorting of loaded list (useful for PCA)\n",
    "tfdf = tfdf.set_index('symbol')\n",
    "tfdf = tfdf.reindex(index=tflist['symbol'])\n",
    "tfdf = tfdf.reset_index()\n",
    "\n",
    "topG = tfdf.shape[0]\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "# topExp = tfdf.head(topG)\n",
    "\n",
    "nG_us = len(tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]>fcLim) ])\n",
    "nG_ml = len(tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]<fcLim) ])\n",
    "print(\"Using baseMean>0, padj<{0}, log2FC>{1}\".format(pLim,fcLim))\n",
    "print(\"\\tNumber of US-enriched genes: {0}\".format(nG_us))\n",
    "print(\"\\tNumber of ML-enriched genes: {0}\".format(nG_ml))\n",
    "print(\"\\tTotal: {0}\".format(nG_us+nG_ml))\n",
    "\n",
    "# plotname = ''\n",
    "# fH, axH = plt.subplots(figsize=[10,20])\n",
    "\n",
    "# # x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "# x_pos = np.arange(topG,0,-1)\n",
    "\n",
    "# pH = plt.barh(x_pos, topExp[\"baseMean\"], height=.8, log=1);\n",
    "   \n",
    "# plt.title(\"top expressed TFs\")\n",
    "# plt.yticks(x_pos, topExp[\"symbol\"])\n",
    "\n",
    "# formatFigure_General(\"top expressed TFs\", fH, axH, pH)\n",
    "# plt.xlabel(\"avgFPKM\")\n",
    "# plt.ylabel(\"gene\")\n",
    "# # plt.xlim(right=250)\n",
    "# plt.ylim(bottom = 0, top=topG+1)\n",
    "# # axH.xaxis.set_tick_params(rotation=66)\n",
    "\n",
    "# plt.show()\n",
    "# topExp\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)\n",
    "\n",
    "tfdf.head(20)\n",
    "# tfdf[tfdf[\"symbol\"]=='thrb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempdf = tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]<-fcLim) ].sort_values(by=\"log2FoldChange\", ascending=False)#.to_csv()\n",
    "tempdf = tfdf[tfdf[\"symbol\"]=='prdm1b']\n",
    "tempdf\n",
    "# tempdf.to_csv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tempdf[\"symbol\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now I need to subtract genes that are specific to a single subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# G_ml = tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]>fcLim) ]\n",
    "# G_ml = tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]<0) ]\n",
    "# G_ml = G_ml.sort_values(by=\"log2FoldChange\", ascending=True)\n",
    "# G_ml\n",
    "# plotName = ''\n",
    "# # plotName = 'heatMap_vamp'\n",
    "# samplenames = [] #skipping names to replace in intaglio\n",
    "# genenames = G_ml['symbol'].values\n",
    "\n",
    "# data = G_ml.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values\n",
    "\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=[16,36])\n",
    "# im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# # texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# # plt.savefig(\"./00_bySubtype/tempTFfigs/{0}_vst.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tfdf[ (tfdf[\"baseMean\"]>0) & (tfdf[\"padj\"]<pLim) & (tfdf[\"log2FoldChange\"]<fcLim) ]\n",
    "# G_ml[\"symbol\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = [\n",
    "#   'nr2f6b', 'six7', 'arntl1b', 'maff', 'neurod1', 'per3', 'crebzf',\n",
    "#        'sall1a', 'tgif1', 'rorca', 'homeza', 'si:ch211-198a12.6', 'roraa',\n",
    "#        'nr2f2', 'ell2', 'supt16h', 'egr1', 'etv5b', 'sox6', 'rx1',\n",
    "#        'fosab', 'klf11b', 'six6b', 'malt1', 'ppargc1a', 'hnf1a', 'mef2ca',\n",
    "#        'hic1l', 'sall2', 'mybbp1a', 'thraa', 'si:dkey-111k8.3',\n",
    "#        'lrrfip1a', 'taf6', 'ell', 'si:dkey-15h8.16', 'nkx2.2a',\n",
    "#        'si:ch73-120g24.4', 'mef2ab', 'rnf25', 'sox5', 'si:ch73-109d9.2',\n",
    "#        'si:ch73-138e16.2', 'ncoa1', 'znf1009', 'zgc:110249', 'myf6',\n",
    "#        'mef2cb', 'si:dkey-18a10.3', 'nr1i2', 'tbx21', 'si:dkey-219c3.2',\n",
    "#        'zgc:113336', 'znf1152', 'znf1130', 'tbx2b', 'znf740b',\n",
    "#        'si:ch73-299h12.9', 'zgc:175284', 'si:dkey-20i20.3', 'hif1an',\n",
    "#        'elk3', 'zgc:174275', 'mlx', 'taf5', 'si:ch211-162i8.5',\n",
    "#        'si:ch211-245n8.4', 'nr2c2', 'atf3', 'bach1a', 'six6a', 'znf1119',\n",
    "#        'znf1160', 'si:dkey-20i20.7', 'si:ch73-299h12.8', 'znf970',\n",
    "#        'si:ch211-255f4.11', 'zic6', 'gbx2', 'taf2', 'aff2', 'znf1122',\n",
    "#        'si:dkey-172k15.3', 'foxp3a', 'e2f3', 'nr4a3', 'pprc1',\n",
    "#        'zgc:113220', 'egr3', 'egr4', 'rx3', 'pxna', 'si:cabz01054394.6',\n",
    "#        'znf1128', 'znf1168', 'paf1', 'znf1181', 'zgc:163077', 'ahr2',\n",
    "#        'si:dkey-253d23.9', 'si:ch211-79k12.2', 'samd12', 'gtf3c4'\n",
    "    'cngb1a','cngb1b'\n",
    "           ]\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L-specific DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/TF_L_top20PC1.csv\") # get list of some TFs\n",
    "tflist = pd.read_csv(\"00_bySubtype/conesubtypeTFs_revisedOct2021.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "# print(\"\\n non-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# keep sorting of loaded list (useful for PCA)\n",
    "tfdf = tfdf.set_index('symbol')\n",
    "tfdf = tfdf.reindex(index=tflist['symbol'])\n",
    "tfdf = tfdf.reset_index()\n",
    "\n",
    "topG = tfdf.shape[0]\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "# topExp = tfdf.head(topG)\n",
    "\n",
    "# plotname = ''\n",
    "# fH, axH = plt.subplots(figsize=[10,20])\n",
    "\n",
    "# # x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "# x_pos = np.arange(topG,0,-1)\n",
    "\n",
    "# pH = plt.barh(x_pos, topExp[\"baseMean\"], height=.8, log=1);\n",
    "   \n",
    "# plt.title(\"top expressed TFs\")\n",
    "# plt.yticks(x_pos, topExp[\"symbol\"])\n",
    "\n",
    "# formatFigure_General(\"top expressed TFs\", fH, axH, pH)\n",
    "# plt.xlabel(\"avgFPKM\")\n",
    "# plt.ylabel(\"gene\")\n",
    "# # plt.xlim(right=250)\n",
    "# plt.ylim(bottom = 0, top=topG+1)\n",
    "# # axH.xaxis.set_tick_params(rotation=66)\n",
    "\n",
    "# plt.show()\n",
    "\n",
    "\n",
    "# topExp\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'TF_C_manual'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tfdf['symbol'].values\n",
    "\n",
    "# data = tfdf.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values #normalized data to max\n",
    "data = tfdf.iloc[0:,7:37].values\n",
    "\n",
    "# plotName = 'TF_L_top20PCA'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "# plt.savefig(\"/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/{0}.svg\".format('TF_C_manual_updated'), transparent=True, format=\"svg\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### M-specific DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "tflist = pd.read_csv(\"00_bySubtype/TF_M_top20PC2.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "# print(\"\\n non-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# keep sorting of loaded list (useful for PCA)\n",
    "tfdf = tfdf.set_index('symbol')\n",
    "tfdf = tfdf.reindex(index=tflist['symbol'])\n",
    "tfdf = tfdf.reset_index()\n",
    "\n",
    "topG = tfdf.shape[0]\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "# topExp = tfdf.head(topG)\n",
    "\n",
    "plotname = ''\n",
    "fH, axH = plt.subplots(figsize=[10,20])\n",
    "\n",
    "# x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "x_pos = np.arange(topG,0,-1)\n",
    "\n",
    "pH = plt.barh(x_pos, topExp[\"baseMean\"], height=.8, log=1);\n",
    "   \n",
    "plt.title(\"top expressed TFs\")\n",
    "plt.yticks(x_pos, topExp[\"symbol\"])\n",
    "\n",
    "formatFigure_General(\"top expressed TFs\", fH, axH, pH)\n",
    "plt.xlabel(\"avgFPKM\")\n",
    "plt.ylabel(\"gene\")\n",
    "# plt.xlim(right=250)\n",
    "plt.ylim(bottom = 0, top=topG+1)\n",
    "# axH.xaxis.set_tick_params(rotation=66)\n",
    "\n",
    "plt.show()\n",
    "# topExp\n",
    "# tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tfdf['symbol'].values\n",
    "\n",
    "# data = tfdf.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values #normalized data to max\n",
    "data = tfdf.iloc[0:,7:37].values\n",
    "\n",
    "plotName = 'TF_M_top20PCA'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## UV and S specific DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "tflist = pd.read_csv(\"00_bySubtype/TF_US_top20PCs.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "# print(\"\\n non-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# keep sorting of loaded list (useful for PCA)\n",
    "tfdf = tfdf.set_index('symbol')\n",
    "tfdf = tfdf.reindex(index=tflist['symbol'])\n",
    "tfdf = tfdf.reset_index()\n",
    "\n",
    "topG = tfdf.shape[0]\n",
    "topExp = tfdf.sort_values(by=\"baseMean\", ascending=False).head(topG)\n",
    "# topExp = tfdf.head(topG)\n",
    "\n",
    "# plotname = ''\n",
    "# fH, axH = plt.subplots(figsize=[10,20])\n",
    "\n",
    "# # x_pos = [i for i, _ in enumerate( topExp[\"symbol\"])]\n",
    "# x_pos = np.arange(topG,0,-1)\n",
    "\n",
    "# pH = plt.barh(x_pos, topExp[\"baseMean\"], height=.8, log=1);\n",
    "   \n",
    "# plt.title(\"top expressed TFs\")\n",
    "# plt.yticks(x_pos, topExp[\"symbol\"])\n",
    "\n",
    "# formatFigure_General(\"top expressed TFs\", fH, axH, pH)\n",
    "# plt.xlabel(\"avgFPKM\")\n",
    "# plt.ylabel(\"gene\")\n",
    "# # plt.xlim(right=250)\n",
    "# plt.ylim(bottom = 0, top=topG+1)\n",
    "# # axH.xaxis.set_tick_params(rotation=66)\n",
    "\n",
    "# plt.show()\n",
    "# # topExp\n",
    "# # tfdf.sort_values(by=\"baseMean\", ascending=False).head(100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = tfdf['symbol'].values\n",
    "\n",
    "data = tfdf.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values #normalized data to max\n",
    "# data = tfdf.iloc[0:,7:37].values\n",
    "\n",
    "plotName = 'TF_US_top20PCA'\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Known TFs<a id='knownTFs'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)  \n",
    "[Go to Exploring TFs](#TFs)  \n",
    "[Go to RNAseq explorer](#heatOther)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.25 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\") # get fpkm data\n",
    "# tflist = pd.read_csv(\"00_bySubtype/zfin_TFs.csv\") # get list of all TFs\n",
    "# tflist = pd.read_csv(\"00_bySubtype/manual_TFs.csv\") # get list of some TFs\n",
    "tflist = pd.read_csv(\"00_bySubtype/manual_TFs_ii.csv\") # get list of some TFs\n",
    "tfdf = gdf[gdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "tfdf = tfdf[tfdf[\"baseMean\"]>0] # remove TFs without reads\n",
    "# tfdf = tfdf.sort_values(by=\"baseMean\", ascending=False)\n",
    "print(\"\\n non-zero expression detected in {0}/{1} of transcription factors\".format(np.size(tfdf[\"symbol\"]),np.size(tflist[\"symbol\"])))\n",
    "\n",
    "# vst = pd.read_csv(\"00_bySubtype/RvsC_vstData.csv\")\n",
    "# vst.iloc[0:,1:31]\n",
    "# vst[vst[\"symbol\"].isin([\"rho\",\"rhol\"])]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sortCol = 'log2FoldChange'\n",
    "# sortCol = 'baseMean'\n",
    "# sortCol = 'padj'\n",
    "\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/RvsC_rawFC.csv\") # load DeSEQ2 analysis: rods vs. cones\n",
    "DS = DS[DS[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "Rup = DS[(DS['baseMean']>0)]\n",
    "Rup = Rup.drop(['stat'], axis = 1)\n",
    "Rup = Rup.set_index('symbol')\n",
    "# this is to keep loaded csv order\n",
    "Rup = Rup.reindex(index=tflist[\"symbol\"])\n",
    "# this is is manual_TFs was loaded\n",
    "# Rup = Rup.reindex(index=[\"crx\",\"otx2a\",\"otx2b\",\"otx5\",\"pax6a\",\"pax6b\",\"onecut1\",\"atoh7\",\"lhx4\",\"pou3f1\",\"neurog1\",\"rx3\",\"nrl\",\"nr2e3\",\"neurod1\",\"ascl1a\",\"ascl1b\",\"samd7\",\"nr1d1\",\"nr2e1\",\"nr2f1a\",\"nr2f1b\",\"nr2f6a\",\"nr2f6b\",\"rorb\",\"gtf2ird1\",\"thrb\",\"rxrga\",\"rxrgb\",\"gdf6a\",\"gdf6b\",\"six6a\",\"six6b\",\"six7\",\"prdm1a\",\"prdm1b\",\"vsx2\",\"tbx2a\",\"tbx2b\",\"foxq2\",\"skor1a\"])\n",
    "Rup = Rup.reset_index()\n",
    "\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Rup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Rup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Rup = pd.concat([Rup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "Rup = Rup[Rup[\"baseMean\"]>0] # remove TFs without reads\n",
    "\n",
    "# Rup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'TF_RvC_manual'\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Rup['symbol'].values])\n",
    "\n",
    "data = Rup.iloc[0:,7:37].values\n",
    "data = np.concatenate([Rup.iloc[0:,7:37].apply(lambda x: x/x.max(), axis=1).values])\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.savefig(\"./00_bySubtype/tempTFfigs/{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRetDev10x_Xu2020/ret72hpf/conesXu.csv\")\n",
    "tfxf = xdf[xdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "\n",
    "tfxf = tfxf.set_index('symbol')\n",
    "tfxf = tfxf.reindex(index=tflist['symbol'])\n",
    "tfxf = tfxf.reset_index()\n",
    "tfxf = tfxf.fillna(value=0)\n",
    "\n",
    "\n",
    "genenameszfx = tfxf['symbol'].values\n",
    "samplenameszfx = [] #skipping names to replace in intaglio\n",
    "# datazfx = tfxf.iloc[0:,2:8]\n",
    "datazfx = np.concatenate([tfxf.iloc[0:,2:8].apply(lambda x: x/x.max(), axis=1).values])\n",
    "\n",
    "figr, axzfx = plt.subplots(figsize=[16,36])\n",
    "imzfx, cbarzfx = heatmap_zfXu2020(datazfx, genenameszfx, samplenameszfx, ax=axzfx)\n",
    "# plt.savefig(\"./00_bySubtype/tempTFfigs/Xu2020_{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hdf = pd.read_csv(\"~/Documents/LiMolec/otherRNAseq/zfRet_HoangBlackshaw2020/conesDev_Hoang.csv\")\n",
    "tfhf = hdf[hdf[\"symbol\"].isin(tflist[\"symbol\"])] # use as logical to extract TF data only\n",
    "    \n",
    "tfhf = tfhf.set_index('symbol')\n",
    "tfhf = tfhf.reindex(index=tflist['symbol'])\n",
    "tfhf = tfhf.reset_index()\n",
    "tfhf = tfhf.fillna(value=0)\n",
    "\n",
    "\n",
    "genenameszfx = tfhf['symbol'].values\n",
    "samplenameszfx = [] #skipping names to replace in intaglio\n",
    "# datazfx = tfxf.iloc[0:,2:8]\n",
    "datazflh = np.concatenate([tfhf.iloc[0:,2:8].apply(lambda x: x/x.max(), axis=1).values])\n",
    "datazfah = np.concatenate([tfhf.iloc[0:,8:13].apply(lambda x: x/x.max(), axis=1).values])\n",
    "\n",
    "figr, (axzflh,axzfah) = plt.subplots(1,2,figsize=[20,36])\n",
    "imzflh, cbarzflh = heatmap_zflarvaHoang2020(datazflh, genenameszfx, samplenameszfx, ax=axzflh)\n",
    "imzfah, cbarzfah = heatmap_zfadultHoang2020(datazfah, genenameszfx, samplenameszfx, ax=axzfah)\n",
    "\n",
    "plt.savefig(\"./00_bySubtype/tempTFfigs/Hoang2021_{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Subtype-specific DEGs\n",
    "> Not very pretty; lot of DEG contamination with sparse genes  \n",
    "> lfcshrink gets rid of some of them but not all and obliterates interesting ones (tbx2 and pcdh10a for example)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\n",
    "\n",
    "pLim = 0.1 # false-positive rate < 10% given p-value < 0.01\n",
    "fcLim = 0.58 # FC > 1.5 (in log2 = 0.58)\n",
    "pctLim = 30 # percent of relevant cells expression this gene\n",
    "gdf = pd.read_csv(\"00_rodsVcones/rodVCones_fpkm.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sortCol = 'log2FoldChange'\n",
    "sortCol = 'baseMean'\n",
    "topG = 25\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/LvsUSM.csv\")\n",
    "# DSU[DSU['p_val_adj']<0.05].sort_values(by='avg_logFC', ascending=False).head(25)\n",
    "Lup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "# Lup = Lup.drop(['stat'], axis = 1)\n",
    "Lup = Lup.set_index('symbol')\n",
    "Lup = Lup.reset_index()\n",
    "Lup = Lup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Lup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Lup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Lup = pd.concat([Lup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Lup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Lup = Lup[~nanrows]\n",
    "\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/MvsUSL.csv\")\n",
    "# DSU[DSU['p_val_adj']<0.05].sort_values(by='avg_logFC', ascending=False).head(25)\n",
    "Mup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "# # Mup = Mup.drop(['stat'], axis = 1)\n",
    "Mup = Mup.set_index('symbol')\n",
    "Mup = Mup.reset_index()\n",
    "Mup = Mup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Mup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Mup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Mup = pd.concat([Mup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Mup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Mup = Mup[~nanrows]\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/SvsUML.csv\")\n",
    "# DSU[DSU['p_val_adj']<0.05].sort_values(by='avg_logFC', ascending=False).head(25)\n",
    "Sup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "# # Sup = Sup.drop(['stat'], axis = 1)\n",
    "Sup = Sup.set_index('symbol')\n",
    "Sup = Sup.reset_index()\n",
    "Sup = Sup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Sup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Sup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Sup = pd.concat([Sup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Sup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Sup = Sup[~nanrows]\n",
    "\n",
    "\n",
    "DS = pd.read_csv(\"00_bySubtype/UvsSML.csv\")\n",
    "# DSU[DSU['p_val_adj']<0.05].sort_values(by='avg_logFC', ascending=False).head(25)\n",
    "Uup = DS[(DS['padj']<pLim) & (DS['log2FoldChange']>fcLim)].sort_values(by=sortCol, ascending=False)\n",
    "# # Uup = Uup.drop(['stat'], axis = 1)\n",
    "Uup = Uup.set_index('symbol')\n",
    "Uup = Uup.reset_index()\n",
    "Uup = Uup.iloc[0:topG,:]\n",
    "\n",
    "tempdf = gdf[gdf[\"symbol\"].isin(Uup[\"symbol\"])]\n",
    "tempdf = tempdf.set_index('symbol')\n",
    "tempdf = tempdf.reindex(index=Uup['symbol'])\n",
    "tempdf = tempdf.reset_index()\n",
    "\n",
    "Uup = pd.concat([Uup, tempdf.iloc[0:,7:37]], axis=1)\n",
    "nanrows = np.isnan(Uup.iloc[0:,7:37].sum(axis=1, skipna=False))\n",
    "Uup = Uup[~nanrows]\n",
    "\n",
    "Uup\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'top' + str(topG)\n",
    "# plotName = 'heatMap_vamp'\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "genenames = np.concatenate([Uup['symbol'].values,Sup['symbol'].values,Mup['symbol'].values,Lup['symbol'].values])\n",
    "data = np.concatenate([Uup.iloc[0:25,7:37].values, Sup.iloc[0:25,7:37].values,Mup.iloc[0:25,7:37].values,Lup.iloc[0:25,7:37].values])\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[16,36])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "# plotName = 'heatMap_vamp'\n",
    "genenames = Lup['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = Lup.iloc[0:25,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "#### plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "plotName = 'test'\n",
    "# plotName = 'heatMap_vamp'\n",
    "genenames = Mup['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = Mup.iloc[0:25,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "#### plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'test'\n",
    "# plotName = 'heatMap_vamp'\n",
    "genenames = Sup['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = Sup.iloc[0:25,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "#### plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n",
    "\n",
    "plotName = 'test'\n",
    "# plotName = 'heatMap_vamp'\n",
    "genenames = Uup['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = Uup.iloc[0:25,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap_z(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "#### plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bar plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw2','olfm2b','plxnb1a', 'arr3b','si:ch73-221f6.1', 'zgc:109934', 'erfl3', 'unc119.2', 'fkbp1aa'] # \n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## misc. things<a id='misc'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding genenames to pcaWeights and using opportunity to sort by PC1 value and remove NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2\"\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_rodsVcones/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_USvsLM/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_LvsM/pcaWeights_Rout.csv\")\n",
    "pcr = pd.read_csv(deseqdir + \"/00_UvsS/pcaWeights_Rout.csv\")\n",
    "# pcr.head\n",
    "gdf = pd.read_csv(deseqdir + \"/00_rodsVcones/rodVCones_fpkm.csv\")\n",
    "pcr = pcr.set_index('symbol')\n",
    "pcr = pcr.reindex(index=gdf['symbol'])\n",
    "pcr = pcr.reset_index()\n",
    "pcr.insert(1,'genename',gdf['genename'],True)\n",
    "pcr.dropna(subset=['PC1'], inplace=True)\n",
    "pcr.sort_values(by=['PC1'], inplace=True)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_rodsVcones/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_USvsLM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_LvsM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_UvsS/pcaWeights.csv\", index=False)\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(gdf.to_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trying to draw counts for single gene by wedges in a circle \n",
    "> Prob not worth it, difficult to interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "\n",
    "n = 30;\n",
    "r = 0;\n",
    "wdg_theta = 360/30;\n",
    "\n",
    "patches = []\n",
    "circles = []\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,0:6]), wdg_theta * (0), (wdg_theta * (6)), edgecolor=\"none\", facecolor = pC['r'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,7:11]), wdg_theta * (6), (wdg_theta * (11)), edgecolor=\"none\", facecolor = pC['u'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,12:17]), wdg_theta * (11), (wdg_theta * (17)), edgecolor=\"none\", facecolor = pC['s'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,18:24]), wdg_theta * (17), (wdg_theta * (24)), edgecolor=\"none\", facecolor = pC['m'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,25:30]), wdg_theta * (24), (wdg_theta * (30)), edgecolor=\"none\", facecolor = pC['l'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "for i in range(n):\n",
    "    p = matplotlib.patches.Wedge(0, data[r,i], wdg_theta * (i), (wdg_theta * (i+1)) -2, edgecolor=\"none\", facecolor = pC['plt'][i], alpha = 0.4)\n",
    "    patches.append(p)\n",
    "\n",
    "for i in [0.001,.2,.4,.6,.8,1]:\n",
    "    c = matplotlib.patches.Circle((0,0), np.max(data[r,:])*i, edgecolor = '#000000', facecolor='none', alpha=0.59)\n",
    "    circles.append(c)\n",
    "\n",
    "plt.xlim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "plt.ylim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "\n",
    "\n",
    "circleColl = matplotlib.collections.PatchCollection(circles, match_original=True);\n",
    "ax.add_collection(circleColl)\n",
    "\n",
    "patchColl = matplotlib.collections.PatchCollection(patches, match_original=True);\n",
    "ax.add_collection(patchColl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[r,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[4,i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(0,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From 2019, Whitaker ,..., Swaroop\n",
    "> genes involved in rod synapse size and localization\n",
    "\n",
    "> [link to biorXiv](https://www.biorxiv.org/content/biorxiv/early/2019/08/30/753012.full.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('dpf') # dpf3 -> not this one but dpf2 in all photoreceptors\n",
    "# gdf = findgenelist('epb') # epb4.1l2 -> not this one but some interesting patterns in this family\n",
    "# gdf = findgenelist('grtp') # grtp1 -> this one is interesting!\n",
    "# gdf = findgenelist('kcnj') # kncj14 -> squirrel too but not zebrafish\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('llgl') # llgl2 -> nope\n",
    "# gdf = findgenelist('rab') # rab28 -> nope; this family is too big\n",
    "# gdf = gdf[gdf.baseMean > 2]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('rom1') #rom1 -> rod-specific; what about outer segments?\n",
    "\n",
    "# gdf = findgenelist('arl') # arl2bp -> nope but arl3l2 in all photoreceptors, and arl3l1 downregulated in uv and s. Both confirmed eye specific (https://zfin.org/ZDB-GENE-040426-1678; https://zfin.org/ZDB-GENE-040426-1649)\n",
    "# These genes are supponsed to interact with unc119c (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and act as their GDF\n",
    "# gdf = gdf[gdf.baseMean > 5]\n",
    "# # gdf = gdf[gdf.baseMean < 300]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('unc119') # unc119c is photoreceptor specific (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and CoIPs with arl3l2\n",
    "\n",
    "\n",
    "gdf = findgenelist('ppar') #ppargc2\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "\n",
    "### https://www.sciencedirect.com/science/article/pii/S2211124718303218\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "# gdf = findgenelist('sdc') #sdcbp interacts with lrit 1\n",
    "# gdf = findgenelist('grip') #grip1 interacts with lrit1\n",
    "\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gdf = gdf[gdf.baseMean > .45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
