{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [NAPARI](https://napari.org/) testing\n",
    "## and adding [cellpose](https://cellpose-napari.readthedocs.io/en/latest/index.html) (installer [here](https://pypi.org/project/cellpose-napari/0.1.3/))\n",
    "\n",
    "To upgrade run in a cell:\n",
    "```python\n",
    "!pip install cellpose-napari==0.1.3\n",
    "!pip install cellpose --upgrade\n",
    "```\n",
    "\n",
    "```bash\n",
    "napari -w cellpose-napari\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic usage:\n",
    "```python\n",
    "viewer = napari.view_image(data.astronaut(), rgb=True)\n",
    "nbscreenshot(viewer)\n",
    "viewer.close()\n",
    "```\n",
    "\n",
    "example notebooks [here](https://github.com/sofroniewn/napari-training-course/blob/master/lessons/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## to install\n",
    "# !pip install napari[all]\n",
    "# !pip install QT\n",
    "# !pip install napari[pyqt5]\n",
    "# !pip install cellpose-napari==0.1.4\n",
    "# !pip install napari-nikon-nd2\n",
    "# !pip install magicgui\n",
    "# !pip install napari-pyclesperanto-assistant\n",
    "# !pip install napari-clusters-plotter #hdbscan failed\n",
    "# !pip install napari-plot-profile\n",
    "# !pip install napari-brightness-contrast\n",
    "# !pip install napari-curtain\n",
    "# !pip install napari-3d-ortho-viewer\n",
    "# !pip install napari-manual-split-and-merge-labels\n",
    "# !pip install napari-crop\n",
    "# !pip install napari-stl-exporter\n",
    "\n",
    "## to upgrade:\n",
    "# !pip install cellpose-napari==0.1.4\n",
    "# !pip install cellpose-napari --upgrade\n",
    "# !pip install cellpose --upgrade\n",
    "# !pip install napari --upgrade\n",
    "# !pip install mxnet-mkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import napari\n",
    "import cellpose_napari\n",
    "import cellpose\n",
    "from cellpose import models\n",
    "from napari.utils import nbscreenshot\n",
    "from tifffile import imread\n",
    "import numpy as np\n",
    "from scipy import ndimage\n",
    "from scipy.stats import mannwhitneyu\n",
    "import napari_nikon_nd2\n",
    "import os \n",
    "from magicgui import magicgui\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "from plotParams import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotStyle = 'dark';\n",
    "# plotStyle = 'light';\n",
    "\n",
    "if plotStyle=='dark':\n",
    "    # dark background\n",
    "    params = {\"ytick.color\" : \"w\",\n",
    "              \"xtick.color\" : \"w\",\n",
    "              \"axes.labelcolor\" : \"w\",\n",
    "              \"axes.edgecolor\" : \"w\",\n",
    "             \"axes.linewidth\" : 3,\n",
    "             \"xtick.major.width\" : 3,\n",
    "             \"ytick.major.width\" : 3,\n",
    "             \"xtick.major.size\" : 8,\n",
    "             \"ytick.major.size\" : 8,\n",
    "             \"text.color\" : \"w\"}\n",
    "    plt.rcParams.update(params)\n",
    "elif plotStyle=='light':\n",
    "    # white background\n",
    "    params = {\"ytick.color\" : \"k\",\n",
    "              \"xtick.color\" : \"k\",\n",
    "              \"axes.labelcolor\" : \"k\",\n",
    "              \"axes.edgecolor\" : \"k\",\n",
    "             \"axes.linewidth\" : 3,\n",
    "             \"xtick.major.width\" : 3,\n",
    "             \"ytick.major.width\" : 3,\n",
    "             \"xtick.major.size\" : 8,\n",
    "             \"ytick.major.size\" : 8,\n",
    "             \"text.color\" : \"k\"}\n",
    "    plt.rcParams.update(params)\n",
    "font_prop = font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc')\n",
    "\n",
    "\n",
    "matplotlib.rcParams['axes.prop_cycle'] = matplotlib.cycler(color=batlow)\n",
    "\n",
    "# viewer = napari.Viewer()\n",
    "\n",
    "print('Opened viewer; plot style is ' + plotStyle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'></a>\n",
    "***\n",
    "- [Extraction of z planes for analysis](#zExtract)\n",
    "- [Segmentation with cellpose](#cellSeg)\n",
    "- [Manual correction of segmentation](#manualCuration)\n",
    "- [Create thumbnails _WIP_](#thumbnails)\n",
    "- [Quantification](#quantification)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "***\n",
    "## Extract layers from z-stacks<a id='zExtract'>âˆ®</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting list of files\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/zf/20220107_tbx2bF3s_xOG_s1C\"\n",
    "!ls $dPath/*[002,004].nd2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2bF3s/xOGs1C/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# F3[tbx2b]\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/zf/20220107_tbx2bF3s_xOG_s1C/\"\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L01_002'; gene = 'wt'; zlims = [58,61];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L02_002'; gene = 'het'; zlims = [45,48];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L03_002'; gene = 'het'; zlims = [43,47];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L04_002'; gene = 'ko'; zlims = [40,47]; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L05_002'; gene = 'het'; zlims = [64,69];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L06_002'; gene = 'het'; zlims = [38,42]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L06_004'; gene = 'het'; zlims = [43,46]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L08_002'; gene = 'het'; zlims = [63,66]; # DAPI is dim\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L09_002'; gene = 'ko'; zlims = [28,31];\n",
    "# ### fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L09_004'; gene = 'ko'; zlims = [33,39]; # side-view\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L10_002'; gene = 'het'; zlims = [18,25]; # a little curved\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L11_002'; gene = 'het'; zlims = [30,40]; # very curved\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L12_002'; gene = 'wt'; zlims = [18,22];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L13_002'; gene = 'ko'; zlims = [30,34];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L13_004'; gene = 'ko'; zlims = [34,45]; #GFP is dim and variable\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L14_002'; gene = 'het'; zlims = [33,38]; # ONH in FOV\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L15_002'; gene = 'wt'; zlims = [30,34]; # flat and good for DAPI\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L16_002'; gene = 'het'; zlims = [15,18]; # mCherry+ and flat and good for DAPI **EXAMPLE**\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L17_002'; gene = 'wt'; zlims = [28,30]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L18_002'; gene = 'wt'; zlims = [36,40];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L19_002'; gene = 'ko'; zlims = [30,35]; # not lor!!!!\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L20_002'; gene = 'wt'; zlims = [33,40];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L21_002'; gene = 'het'; zlims = [30,35]; # ONH in FOV\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L22_002'; gene = 'het'; zlims = [32,38];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L23_002'; gene = 'wt'; zlims = [69,71];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L24_002'; gene = 'wt'; zlims = [54,57];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L25_002'; gene = 'ko'; zlims = [43,60]; # not lor!!!!\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L26_002'; gene = 'ko'; zlims = [48,52]; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L26_004'; gene = 'ko'; zlims = [38,41]; # lor and flat **EXAMPLE**\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L27_002'; gene = 'het'; zlims = [38,41]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L28_002'; gene = 'het'; zlims = [18,23]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L29_002'; gene = 'wt'; zlims = [40,45];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L30_002'; gene = 'het'; zlims = [57,59]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L31_002'; gene = 'wt'; zlims = [72,74]; # flat and good for DAPI\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L32_002'; gene = 'het'; zlims = [40,43];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L33_002'; gene = 'het'; zlims = [27,29]; # ONH in FOV\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L34_002'; gene = 'wt'; zlims = [23,25];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L35_002'; gene = 'ko'; zlims = [56,58]; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L35_004'; gene = 'ko'; zlims = [27,29];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L36_002'; gene = 'ko'; zlims = [17,19]; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L36_004'; gene = 'ko'; zlims = [45,47];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L37_002'; gene = 'het'; zlims = [53,55];\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L38_002'; gene = 'het'; zlims = [36,40]; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L39_002'; gene = 'ko'; zlims = [53,74]; # Very very curved; will need manual correction\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L40_002'; gene = 'wt'; zlims = [55,59]; # mCherry+; ONH in FOV\n",
    "\n",
    "\n",
    "# create subdirectory for analysis\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "if (os.path.isdir(dOut)==False):\n",
    "    os.mkdir(dOut)\n",
    "    print('Created output directory')\n",
    "\n",
    "print('Canvas clear')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load whole stack to figure out best planes for cellpose\n",
    "viewer.open((dPath+fPath + \".nd2\"))\n",
    "viewer.layers.remove(viewer.layers[len(viewer.layers)-1]) # remove transmitted detector image\n",
    "viewer.layers.select_next();\n",
    "if len(viewer.layers)==2: #no DAPI\n",
    "    nChannels = 2\n",
    "    viewer.layers[0].colormap = 'green'\n",
    "    viewer.layers[0].name = 'R'\n",
    "    viewer.layers[1].colormap = 'magenta'\n",
    "    viewer.layers[1].name = 'U'\n",
    "elif len(viewer.layers)==3: #DAPI\n",
    "    nChannels = 3\n",
    "    viewer.layers[0].colormap = 'gray'\n",
    "    viewer.layers[0].name = 'N'\n",
    "    viewer.layers[0].opacity = 0.33\n",
    "    viewer.layers[1].colormap = 'green'\n",
    "    viewer.layers[1].name = 'R'\n",
    "    viewer.layers[2].colormap = 'magenta'\n",
    "    viewer.layers[2].name = 'U'\n",
    "\n",
    "print('Loaded: ' + fPath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make mips (and remove any previous ones)\n",
    "# # code for eyes without immuno\n",
    "zlims = [55,59];\n",
    "if len(viewer.layers)>3:\n",
    "    for l in viewer.layers[3:]:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "for l in [viewer.layers[0],viewer.layers[1],viewer.layers[2]]:\n",
    "    l.visible = False\n",
    "#     viewer.layers.remove(l.name + '_mip')\n",
    "    viewer.add_image(l.data[zlims[0]:zlims[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")\n",
    "    \n",
    "    \n",
    "# # for FPs ------- # for OSs # code for eyes with 1D4 immuno\n",
    "# zlims = [28,30]; zlims2 = [14,18];\n",
    "# if len(viewer.layers)>3:\n",
    "#     for l in viewer.layers[3:]:\n",
    "#         viewer.layers.remove(l)\n",
    "\n",
    "# for l in [viewer.layers[0],viewer.layers[1]]:\n",
    "#     l.visible = False\n",
    "# #     viewer.layers.remove(l.name + '_mip')\n",
    "#     viewer.add_image(l.data[zlims[0]:zlims[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")\n",
    "    \n",
    "# for l in [viewer.layers[2]]:\n",
    "#     l.visible = False\n",
    "# #     viewer.layers.remove(l.name + '_mip')\n",
    "#     viewer.add_image(l.data[zlims2[0]:zlims2[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# when things look good, save in folder for batch cellpose analysis\n",
    "l = viewer.layers['R_mip']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers['U_mip']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers['N_mip']; l.save(dOut + l.name + '.tiff')\n",
    "print('MIP layers saved for ' + fPath)\n",
    "# l = viewer.layers['LWS_mip']; l.save(dOut + l.name + '.tiff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Go back and run next one &uarr;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# after running cellpose, rename cellpose images and save before manual curation:\n",
    "baseName = 'C'\n",
    "lname = baseName + '_mip'\n",
    "viewer.layers[lname + '_cp_masks_000'].name = baseName + '_seg'\n",
    "viewer.layers[lname + '_cp_outlines_000'].name = baseName + '_outlines'\n",
    "viewer.layers[lname + '_cp_cellprob_000'].name = baseName + '_prob'\n",
    "viewer.layers[lname + '_cp_flows_000'].name = baseName + '_flows'\n",
    "\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_outlines']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_prob']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_flows']; l.save(dOut + l.name + '.tiff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Segmentation using cellpose<a id='cellSeg'>âˆ®</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run cellpose in napari GUI and identify good parameters.\n",
    "Usually for:\n",
    "1. xOPS:GFP: diameter = 20; flow_threshold = 0.4 and mask_threshold = 0.0\n",
    "1. sws1:mCherry: diameter = 40; flow_threshold = 0.4 and mask_threshold = 0.0\n",
    "1. sws2:mCherry:\n",
    "1. mws2:GFP:\n",
    "1. thrb:tdTomato:\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2bF3s/xOGs1C/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# F3[tbx2b]\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/zf/20220107_tbx2bF3s_xOG_s1C/\"\n",
    "fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L01_002'; gene = 'wt'; zlims = [58,61];\n",
    "\n",
    "# redefine analysis subdirectory\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "print('Viewer cleared...')\n",
    "\n",
    "# load mips\n",
    "viewer.open(dOut + 'R' + \"_mip.tiff\", plugin='builtins', colormap = 'green', blending='additive');\n",
    "viewer.open(dOut + 'U' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "viewer.open(dOut + 'N' + \"_mip.tiff\", plugin='builtins', colormap = 'gray', opacity= 0.33, blending='additive');\n",
    "\n",
    "print('Loaded mips')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cellpose: batch analysis\n",
    "#### Define parameters and files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# define analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2bF3s/xOGs1C/\"\n",
    "\n",
    "# collect file paths\n",
    "\n",
    "gene = 'tbx2b'\n",
    "fPaths = [\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L01_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L02_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L03_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L04_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L05_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L06_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L06_004',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L08_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L09_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L10_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L11_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L12_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L13_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L13_004',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L14_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L15_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L16_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L17_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L18_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L19_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L20_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L21_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L22_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L23_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L24_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L25_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L26_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L26_004',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L27_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L28_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L29_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L30_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L31_002',\n",
    "#    '20220107_5dpf_xOG_s1C_tbx2bF3_L32_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L33_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L34_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L35_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L35_004',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L36_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L36_004',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L37_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L38_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L39_002',\n",
    "    '20220107_5dpf_xOG_s1C_tbx2bF3_L40_002',\n",
    "         ]\n",
    "\n",
    "# define cellpose params\n",
    "cpParams = {\n",
    "    'model' : 'cyto2', # default is 'cyto' or 'cyto2'\n",
    "    'net_avg' : True,\n",
    "    'channels' : [0,0], #single channel without nucleus info\n",
    "    'diameterR' : 20,\n",
    "    'diameterU' : 40,\n",
    "    'flow_threshold' : 0.4,\n",
    "    'mask_threshold' : 0.0\n",
    "}\n",
    "\n",
    "# define model to use (e.g. 'cyto2')\n",
    "model = models.Cellpose(gpu=False, model_type=cpParams['model'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Starting analysis:')\n",
    "for fPath in fPaths:\n",
    "    print('\\n' + fPath)\n",
    "    print('\\t Rods')\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/' + 'R_mip.tiff'\n",
    "    img = cellpose.io.imread(dOut)\n",
    "#     imgplot = plt.imshow(img)\n",
    "    masks, flows, styles, diams = model.eval(img, \n",
    "                                             diameter=cpParams['diameterR'], channels=cpParams['channels'],\n",
    "                                             do_3D=False, net_avg = cpParams['net_avg'], interp = True,\n",
    "                                             flow_threshold = cpParams['flow_threshold'], mask_threshold = cpParams['mask_threshold'])\n",
    "    cellpose.io.masks_flows_to_seg(img, masks, flows, diams, dOut, cpParams['channels']) # save results\n",
    "    \n",
    "#     print('\\t UV cones')\n",
    "#     dOut = dAnalysis + gene + '/' + fPath + '/' + 'U_mip.tiff'\n",
    "#     img = cellpose.io.imread(dOut)\n",
    "# #     imgplot = plt.imshow(img)\n",
    "#     masks, flows, styles, diams = model.eval(img, \n",
    "#                                              diameter=cpParams['diameterU'], channels=cpParams['channels'],\n",
    "#                                              do_3D=False, net_avg = cpParams['net_avg'], interp = True,\n",
    "#                                              flow_threshold = cpParams['flow_threshold'], mask_threshold = cpParams['mask_threshold'])\n",
    "#     cellpose.io.masks_flows_to_seg(img, masks, flows, diams, dOut, cpParams['channels']) # save results\n",
    "    \n",
    "print('Finished cellpose batch analysis')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Manual correction of cellpose segmentation<a id='manualCuration'>âˆ®</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "# for l in viewer.layers:\n",
    "#         viewer.layers.remove(l)\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2bF3s/xOGs1C/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# F3[tbx2b]\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/zf/20220107_tbx2bF3s_xOG_s1C/\"\n",
    "# ALL THESE DONE FOR RODS ONLY\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L01_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L02_002'; gene = 'het';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L03_002'; gene = 'het';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L04_002'; gene = 'ko'; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L05_002'; gene = 'het'; # ok DAPI\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L06_002'; gene = 'het'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L06_004'; gene = 'het'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L08_002'; gene = 'het'; # mCherry+; DAPI is dim\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L09_002'; gene = 'ko'; # lor; DAPI is ok\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L10_002'; gene = 'het'; # a little curved\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L11_002'; gene = 'het'; # very curved\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L12_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L13_002'; gene = 'ko'; # lor; DAPI is ok\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L13_004'; gene = 'ko'; # lor; GFP is dim and variable\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L14_002'; gene = 'het'; # ONH in FOV; consider rejecting from analysis\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L15_002'; gene = 'wt'; # flat and good for DAPI\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L16_002'; gene = 'het'; # mCherry+ and flat and good for DAPI **EXAMPLE**\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L17_002'; gene = 'wt'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L18_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L19_002'; gene = 'ko'; # not lor!!!!\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L20_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L21_002'; gene = 'het'; # ONH in FOV\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L22_002'; gene = 'het';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L23_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L24_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L25_002'; gene = 'ko'; # not lor!!!!\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L26_002'; gene = 'ko'; # lor\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L26_004'; gene = 'ko'; # lor and flat **EXAMPLE**\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L27_002'; gene = 'het'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L28_002'; gene = 'het'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L29_002'; gene = 'wt';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L30_002'; gene = 'het'; # mCherry+\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L31_002'; gene = 'wt'; # flat and good for DAPI\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L32_002'; gene = 'het';\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L33_002'; gene = 'het'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L34_002'; gene = 'wt'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L35_002'; gene = 'ko'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L35_004'; gene = 'ko'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L36_002'; gene = 'ko'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L36_004'; gene = 'ko'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L37_002'; gene = 'het'; \n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L38_002'; gene = 'het'; mCherry+\n",
    "#### #fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L39_002'; gene = 'ko'; # too curved; unreliable segmentation. Exclude from analysis\n",
    "# fPath = '20220107_5dpf_xOG_s1C_tbx2bF3_L40_002'; gene = 'wt'; \n",
    "\n",
    "\n",
    "# define subdirectory for analysis\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "print('Viewer cleared...')\n",
    "\n",
    "# clear key binds\n",
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    ...\n",
    "@viewer.bind_key('b', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    ...\n",
    "@viewer.bind_key('Shift-x', overwrite=True)\n",
    "def removeLabel(viewer):\n",
    "    ...\n",
    "\n",
    "# load mips\n",
    "viewer.open(dOut + 'R' + \"_mip.tiff\", plugin='builtins', colormap = 'gray', blending='additive');\n",
    "viewer.open(dOut + 'U' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "viewer.open(dOut + 'N' + \"_mip.tiff\", plugin='builtins', colormap = 'bop blue', blending='additive');\n",
    "\n",
    "# load segmentation\n",
    "segData = np.load(dOut + 'R' + \"_mip\" + \"_seg.npy\", allow_pickle=True).item()\n",
    "viewer.add_labels(segData['masks'], name='R_seg',blending='additive');\n",
    "viewer.layers['R_seg'].preserve_labels = True;\n",
    "\n",
    "# segData = np.load(dOut + 'U' + \"_mip\" + \"_seg.npy\", allow_pickle=True).item()\n",
    "# viewer.add_labels(segData['masks'], name='U_seg',blending='additive');\n",
    "# viewer.layers['U_seg'].visible = False\n",
    "# viewer.layers['U_seg'].preserve_labels = True;\n",
    "\n",
    "# viewer.layers['R_seg'].contour = 5\n",
    "# viewer.layers['U_seg'].contour = 5\n",
    "\n",
    "print('Loaded  ' + fPath + ' !\\n Now fix it and save it!')\n",
    "\n",
    "#define useful keyboard shortcuts\n",
    "\n",
    "@viewer.bind_key('Shift-x', overwrite=True)\n",
    "def removeLabel(viewer):\n",
    "    lname = 'R_seg'\n",
    "    tempd = viewer.layers[lname].data\n",
    "    tempd[tempd == viewer.layers[lname].selected_label]=0\n",
    "    viewer.layers[lname].data = tempd\n",
    "    print('Cut!')\n",
    "\n",
    "print('removeLabel on R_seg (\"Shift-X\")')\n",
    "\n",
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'R_seg'\n",
    "    if (viewer.layers[lname].preserve_labels == True):\n",
    "        viewer.layers[lname].preserve_labels = False\n",
    "    elif (viewer.layers[lname].preserve_labels == False):\n",
    "        viewer.layers[lname].preserve_labels = True\n",
    "print('toggle R_seg preserve_labels (\"k\")')\n",
    "        \n",
    "@viewer.bind_key('b', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'U_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "print('toggle U_mip visibility (\"B\")')\n",
    "\n",
    "@viewer.bind_key('n', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'R_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "print('toggle R_mip visibility (\"N\")')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load z-stack if needed\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201030_CRtbx2ab/\"\n",
    "# viewer.open((dPath+fPath + \".nd2\"))\n",
    "# viewer.layers.remove(viewer.layers[fPath+' [2]']) # remove transmitted detector image\n",
    "# viewer.layers.remove(viewer.layers[fPath+' [3]']) # remove immuno layer\n",
    "# viewer.layers[fPath].colormap = 'magenta'\n",
    "# viewer.layers[fPath].name = 'Mz'\n",
    "# viewer.layers[fPath+' [1]'].colormap = 'green'\n",
    "# viewer.layers[fPath+' [1]'].visible = False\n",
    "# viewer.layers[fPath+' [1]'].name = 'Lz'\n",
    "# viewer.layers.select_next();\n",
    "# print('Loaded!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### resave curated segmentation after napari-ing around (only doing R_mip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseName = 'R'\n",
    "lname = baseName + '_seg'\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated.tiff')\n",
    "print('Done with ' + lname + ' for ' + fPath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Go back and run next one &uarr;](#manualCuration)\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manual merging of labels\n",
    "mergeLabels = [167,168]\n",
    "# lname = 'M_seg'\n",
    "lname = 'L_seg'\n",
    "if viewer.layers[lname].visible == False:\n",
    "    print('Layer not active')\n",
    "else:\n",
    "    tempd = viewer.layers[lname].data\n",
    "    tempd[tempd == mergeLabels[1]]=mergeLabels[0]\n",
    "    viewer.layers[lname].data = tempd\n",
    "    print('Merged in ' + lname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually delete label\n",
    "selLabel = 49\n",
    "lname = 'L_seg'\n",
    "tempd = viewer.layers[lname].data\n",
    "tempd[tempd == selLabel]=0\n",
    "viewer.layers[lname].data = tempd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### resave curated segmentation after napari-ing around"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# baseName = 'R'\n",
    "# lname = baseName + '_seg'\n",
    "# l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated_incomplete.tiff')\n",
    "# # l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated.tiff')\n",
    "# viewer.layers['L_mip'].visible = True\n",
    "# viewer.layers['L_seg'].visible = True\n",
    "# viewer.layers['M_seg'].contour = 0\n",
    "# viewer.layers['M_mip'].visible = False\n",
    "# viewer.layers['M_seg'].visible = False\n",
    "# print('Done with ' + lname)\n",
    "\n",
    "# @viewer.bind_key('Shift-x', overwrite=True)\n",
    "# def removeLabel(viewer):\n",
    "# #     lname = 'M_seg'mvvvv\n",
    "#     lname = 'L_seg'\n",
    "#     tempd = viewer.layers[lname].data\n",
    "#     tempd[tempd == viewer.layers[lname].selected_label]=0\n",
    "#     viewer.layers[lname].data = tempd\n",
    "#     print('Cut!')\n",
    "\n",
    "# print('removeLabel on L_seg (\"Shift-X\")')\n",
    "\n",
    "# @viewer.bind_key('k', overwrite=True)\n",
    "# def toggle_sel(viewer):\n",
    "#     lname = 'L_seg'\n",
    "#     if (viewer.layers[lname].preserve_labels == True):\n",
    "#         viewer.layers[lname].preserve_labels = False\n",
    "#     elif (viewer.layers[lname].preserve_labels == False):\n",
    "#         viewer.layers[lname].preserve_labels = True\n",
    "# print('toggle L_seg preserve_labels (\"k\")')\n",
    "\n",
    "# @viewer.bind_key('b', overwrite=True)\n",
    "# def toggle_sel(viewer):\n",
    "#     lname = 'M_mip'\n",
    "#     if (viewer.layers[lname].visible == True):\n",
    "#         viewer.layers[lname].visible = False\n",
    "#     elif (viewer.layers[lname].visible == False):\n",
    "#         viewer.layers[lname].visible = True\n",
    "# print('toggle L_mip visibility (\"B\")')\n",
    "\n",
    "        \n",
    "# @viewer.bind_key('n', overwrite=True)\n",
    "# def toggle_sel(viewer):\n",
    "#     lname = 'L_mip'\n",
    "#     if (viewer.layers[lname].visible == True):\n",
    "#         viewer.layers[lname].visible = False\n",
    "#     elif (viewer.layers[lname].visible == False):\n",
    "#         viewer.layers[lname].visible = True\n",
    "\n",
    "# print('toggle L_mip visibility (\"N\")')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reload curated segmentations after saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all except mips\n",
    "if len(viewer.layers)>2:\n",
    "    for l in viewer.layers[2:]:\n",
    "        viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer.open(dOut + 'M' + \"_seg_curated_incomplete.tiff\", name='M_seg', plugin='builtins');\n",
    "viewer.open(dOut + 'R' + \"_seg_curated.tiff\", name='R_seg', plugin='builtins');\n",
    "viewer.layers['R_seg'].contour = 5\n",
    "\n",
    "print(fPath)\n",
    "nR = len(np.unique(viewer.layers['R_seg'].data))-1\n",
    "print(\"Image has {0} Rods\".format(nR))\n",
    "# nR = len(np.unique(viewer.layers['L_seg'].data))-1\n",
    "# print(\"Image has {0} L cones\".format(nR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all lauers\n",
    "if len(viewer.layers)>0:\n",
    "    for l in viewer.layers[0:]:\n",
    "        viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## To Do: Batch export of png with mips and contours of curated segmentation<a id='thumbnails'>âˆ®</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This could do the trick:\n",
    "```python\n",
    "def blended_img(viewer):\n",
    "    import napari\n",
    "    import numpy as np\n",
    "    \n",
    "    blended = np.zeros(viewer.layers[0].data.shape + (4,))\n",
    "    for layer in viewer.layers:\n",
    "        # normalize data by clims\n",
    "        normalized_data = (layer.data - layer.contrast_limits[0]) / (\n",
    "        layer.contrast_limits[1] - layer.contrast_limits[0]\n",
    "    )\n",
    "        colormapped_data = layer.colormap.map(normalized_data.flatten())\n",
    "        colormapped_data = colormapped_data.reshape(normalized_data.shape + (4,))\n",
    "\n",
    "        blended = blended + colormapped_data\n",
    "    \n",
    "    blended[..., 3] = 1 # set alpha channel to 1\n",
    "\n",
    "    return blended\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Quantification<a id='quantification'></a>: This needs updated (Feb 2022)\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Compile counts by looping through all files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def cellCounter(viewer,dAnalysis,gene,fPath,photoLabel):\n",
    "    import os.path\n",
    "    # clear viewer\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "    viewer.layers.select_all()\n",
    "    viewer.layers.remove_selected()\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    # load segmentation\n",
    "    segPath = dOut + photoLabel + \"_seg_curated.tiff\"\n",
    "    if os.path.isfile(segPath):\n",
    "        viewer.open(segPath, name='Seg', plugin='builtins', blending='additive');\n",
    "        nCells = np.unique(viewer.layers['Seg'].data).shape[0]\n",
    "    else:\n",
    "        nCells = float(\"nan\")\n",
    "    print(fPath + ': nR = ' + str(nCells)) \n",
    "#     np.savez(dOut + 'quantificationGFP.npz',\n",
    "#              lcones=lcones,\n",
    "#              lRFP=lRFP,\n",
    "#              lGFP=lGFP,\n",
    "#              lRFPsd=lRFPsd,\n",
    "#              lGFPsd=lGFPsd,\n",
    "#              lRtile=lRtile,\n",
    "#              lGtile=lGtile)\n",
    "#     results = zip(['lcones','lRFP','lGFP','lRFPsd','lGFPsd','lRtile','lGtile'],\n",
    "#                  [lcones,lRFP,lGFP,lRFPsd,lGFPsd,lRtile,lGtile])\n",
    "    \n",
    "    return nCells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# because it's simple, just going to create csv manually\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRskor1aF0s/xOGaCT/\"\n",
    "\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\n",
    "    \"20211129_5dpf_xOG_aCtdT_wt_L07_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_wt_L07_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_wt_L09_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_wt_L09_004\",\n",
    "         ]\n",
    "for fPath in wt_fPaths:\n",
    "    cellCounter(viewer,dAnalysis,gene,fPath,'R')\n",
    "    \n",
    "    \n",
    "gene = 'skor1a'\n",
    "skor1a_fPaths = [\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L07_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L07_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L08_002\", #from fragment analysis: this is wt\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L08_004\", #from fragment analysis: this is wt\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L09_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L09_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L10_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L11_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L12_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L12_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L13_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L13_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L14_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L15_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L16_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L17_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L17_004\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L18_002\",\n",
    "    \"20211129_5dpf_xOG_aCtdT_skor1aF0_L18_004\",\n",
    "         ]\n",
    "for fPath in skor1a_fPaths:\n",
    "    cellCounter(viewer,dAnalysis,gene,fPath,'R')\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcMillie(viewer,lname):\n",
    "    outHist,outBins = np.histogram(viewer.layers[lname].data.flatten(), bins = 1000)\n",
    "    return outHist,outBins\n",
    "\n",
    "def GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=10):\n",
    "    # clear viewer\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "    viewer.layers.select_all()\n",
    "    viewer.layers.remove_selected()\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    print('Quantifying GFP in L cones of ' + fPath + '(' + gene + ')\\n')\n",
    "    # load mips\n",
    "    viewer.open(dOut + 'M' + \"_mip.tiff\", plugin='builtins', colormap = 'green', blending='additive');\n",
    "    viewer.open(dOut + 'L' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "    # viewer.open(dOut + 'M' + \"_seg_curated_incomplete.tiff\", name='M_seg', plugin='builtins', blending='additive');\n",
    "    viewer.open(dOut + 'L' + \"_seg_curated.tiff\", name='L_seg', plugin='builtins', blending='additive');\n",
    "    # viewer.layers['M_seg'].preserve_labels = True;\n",
    "    viewer.layers['L_seg'].preserve_labels = True;\n",
    "    viewer.layers['L_seg'].contour = 5\n",
    "    # erode L cone segmentation\n",
    "    viewer.layers['L_seg'].data = ndimage.grey_erosion(viewer.layers['L_seg'].data, size=(erosionValue,erosionValue))\n",
    "    # divide images into 1000 bins\n",
    "    mH, mB = calcMillie(viewer,'M_mip')\n",
    "    lH, lB = calcMillie(viewer,'L_mip')\n",
    "\n",
    "    lcones = np.unique(viewer.layers['L_seg'].data)\n",
    "    lcones = lcones[lcones>0]\n",
    "    lRFP = np.empty(np.shape(lcones))\n",
    "    lGFP = np.empty(np.shape(lcones))\n",
    "    lRFPsd = np.empty(np.shape(lcones))\n",
    "    lGFPsd = np.empty(np.shape(lcones))\n",
    "    lRtile = np.empty(np.shape(lcones))\n",
    "    lGtile = np.empty(np.shape(lcones))\n",
    "    \n",
    "    i=0;\n",
    "    for l in lcones:\n",
    "        lRFP[i] = np.mean(viewer.layers['L_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lGFP[i] = np.mean(viewer.layers['M_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lRFPsd[i] = np.std(viewer.layers['L_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lGFPsd[i] = np.std(viewer.layers['M_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lRtile[i] = len(lB[lB<lRFP[i]])\n",
    "        lGtile[i] = len(mB[mB<lGFP[i]])\n",
    "        i=i+1;\n",
    "        \n",
    "    np.savez(dOut + 'quantificationGFP.npz',\n",
    "             lcones=lcones,\n",
    "             lRFP=lRFP,\n",
    "             lGFP=lGFP,\n",
    "             lRFPsd=lRFPsd,\n",
    "             lGFPsd=lGFPsd,\n",
    "             lRtile=lRtile,\n",
    "             lGtile=lGtile)\n",
    "    results = zip(['lcones','lRFP','lGFP','lRFPsd','lGFPsd','lRtile','lGtile'],\n",
    "                 [lcones,lRFP,lGFP,lRFPsd,lGFPsd,lRtile,lGtile])\n",
    "    \n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "erodeV = 10; #12;\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "for fPath in wt_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "    \n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "for fPath in tbx2a_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                 \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "\n",
    "for fPath in tbx2b_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "print('Finished analysis')\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for l in viewer.layers:\n",
    "    print(l)\n",
    "    viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all results\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "                \n",
    "                \n",
    "rName = 'lGtile'\n",
    "\n",
    "wtResults = []\n",
    "i=1;\n",
    "for fPath in wt_fPaths:\n",
    "    gene = 'wt'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    wtResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "aResults = []\n",
    "i=1;\n",
    "for fPath in tbx2a_fPaths:\n",
    "    gene = 'tbx2a'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    aResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "bResults = []\n",
    "i=1;\n",
    "for fPath in tbx2b_fPaths:\n",
    "    gene = 'tbx2b'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    bResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "\n",
    "# combine across images\n",
    "wtResults = np.concatenate(wtResults).ravel()\n",
    "aResults = np.concatenate(aResults).ravel()\n",
    "bResults = np.concatenate(bResults).ravel()\n",
    "\n",
    "# plot histogram\n",
    "pdfV = True\n",
    "# wtH, wtB = np.histogram(wtResults, bins = 50, density=pdfV)\n",
    "# aH, aB = np.histogram(aResults, bins = 50, density=pdfV)\n",
    "# bH, bB = np.histogram(bResults, bins = 50, density=pdfV)\n",
    "\n",
    "# plot histogram but first calculate a fixed bin number for all genotypes\n",
    "# allBins = np.histogram_bin_edges(np.concatenate([wtResults,aResults,bResults]).ravel(),bins='fd');\n",
    "allBins = np.arange(0,1010,10);\n",
    "allBinsC = (allBins[:-1] + allBins[1:]) / 2 # Centers\n",
    "wtH, wtB = np.histogram(wtResults, bins = allBins, density=pdfV)\n",
    "aH, aB = np.histogram(aResults, bins = allBins, density=pdfV)\n",
    "bH, bB = np.histogram(bResults, bins = allBins, density=pdfV)\n",
    "\n",
    "# # this can be used for cumulative distribution then normalized\n",
    "# wtH = np.cumsum(wtH)/np.max(np.cumsum(wtH));\n",
    "# aH = np.cumsum(aH)/np.max(np.cumsum(aH));\n",
    "# bH = np.cumsum(bH)/np.max(np.cumsum(bH));\n",
    "\n",
    "wt95CI = np.min(allBinsC[(np.cumsum(wtH))>.095])/1000;\n",
    "wGBright = np.sum(wtH[allBinsC>wt95CI*1000]);\n",
    "aGBright = np.sum(aH[allBinsC>wt95CI*1000]);\n",
    "bGBright = np.sum(bH[allBinsC>wt95CI*1000]);\n",
    "\n",
    "print('GFP threshold is {0}'.format(wt95CI))\n",
    "\n",
    "baseColor = 'k'\n",
    "if (plotStyle=='light'):\n",
    "    baseColor = 'k';\n",
    "\n",
    "# fH, (axH1,axH2,axH3) = plt.subplots(nrows=3, ncols=1, figsize=(10, 20))\n",
    "fH, (axH1,axH2,axH3) = plt.subplots(nrows=1, ncols=3, figsize=(30, 8))\n",
    "axH1.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH2.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH3.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH2.stairs(aH,aB/1000, fill=True, color = zfC['lws1'], alpha = 0.8)\n",
    "axH3.stairs(bH,bB/1000, fill=True, color = zfC['sws2'], alpha = 0.8)\n",
    "\n",
    "\n",
    "axH1.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "axH2.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "axH3.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "\n",
    "\n",
    "# axH.set_title('Distribution of GFP inside L cones')\n",
    "for axH in (axH1,axH2,axH3):\n",
    "    formatFigureMain(fH, axH, _)\n",
    "    axH.set_xlabel('Normalized GFP values')\n",
    "    if pdfV:\n",
    "        axH.set_ylabel('Probability density')\n",
    "    else:\n",
    "        axH.set_ylabel('Counts')\n",
    "## normalized\n",
    "# axH.stairs(wtH/np.max(wtH),wtB, fill=True, color = 'k', alpha = 0.5)\n",
    "# axH.stairs(aH/np.max(aH),aB, fill=True, color = 'r', alpha = 0.5)\n",
    "# axH.stairs(bH/np.max(bH),bB, fill=True, color = 'b', alpha = 0.5)\n",
    "\n",
    "dPath = '/Users/angueyraaristjm/Documents/LiLab/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Presentations/revealjs/resources/20211008_UCLA/'\n",
    "plt.savefig(dPath + \"CRtbx2F0_LM_hist.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "print('# cells of bright GFP cells: wt={0}, tbx2a={1}, tbx2b={2}'.format(wGBright*1000,aGBright*1000,bGBright*1000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# fH, (axH1,axH2,axH3) = plt.subplots(nrows=3, ncols=1, figsize=(10, 20))\n",
    "fH, axH = plt.subplots(nrows=1, ncols=1, figsize=(9, 6))\n",
    "\n",
    "\n",
    "# plot histogram\n",
    "pdfV = True\n",
    "# wtH, wtB = np.histogram(wtResults, bins = 50, density=pdfV)\n",
    "# aH, aB = np.histogram(aResults, bins = 50, density=pdfV)\n",
    "# bH, bB = np.histogram(bResults, bins = 50, density=pdfV)\n",
    "\n",
    "# plot histogram but first calculate a fixed bin number for all genotypes\n",
    "# allBins = np.histogram_bin_edges(np.concatenate([wtResults,aResults,bResults]).ravel(),bins='fd');\n",
    "allBins = np.arange(0,1010,10);\n",
    "allBinsC = (allBins[:-1] + allBins[1:]) / 2 # Centers\n",
    "wtH, wtB = np.histogram(wtResults, bins = allBins, density=pdfV)\n",
    "aH, aB = np.histogram(aResults, bins = allBins, density=pdfV)\n",
    "bH, bB = np.histogram(bResults, bins = allBins, density=pdfV)\n",
    "\n",
    "# # this can be used for cumulative distribution then normalized\n",
    "# wtH = np.cumsum(wtH)/np.max(np.cumsum(wtH));\n",
    "# aH = np.cumsum(aH)/np.max(np.cumsum(aH));\n",
    "# bH = np.cumsum(bH)/np.max(np.cumsum(bH));\n",
    "\n",
    "wt95CI = np.min(allBinsC[(np.cumsum(wtH))>.095])/1000;\n",
    "wGBright = np.sum(wtH[allBinsC>wt95CI*1000]);\n",
    "aGBright = np.sum(aH[allBinsC>wt95CI*1000]);\n",
    "bGBright = np.sum(bH[allBinsC>wt95CI*1000]);\n",
    "\n",
    "\n",
    "axH.stairs(wtH,wtB/1000, fill=True, color = 'k', alpha = .5)\n",
    "axH.stairs(aH,aB/1000, fill=True, color = zfC['lws1'], alpha = .5)\n",
    "axH.stairs(bH,bB/1000, fill=True, color = zfC['sws2'], alpha = .5)\n",
    "\n",
    "\n",
    "axH.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "\n",
    "formatFigureMain(fH, axH, _)\n",
    "# axH.set_title('Distribution of GFP inside L cones')\n",
    "axH.set_xlabel('Normalized GFP values')\n",
    "if pdfV:\n",
    "    axH.set_ylabel('Probability density')\n",
    "else:\n",
    "    axH.set_ylabel('Counts')\n",
    "## normalized\n",
    "# axH.stairs(wtH/np.max(wtH),wtB, fill=True, color = 'k', alpha = 0.5)\n",
    "# axH.stairs(aH/np.max(aH),aB, fill=True, color = 'r', alpha = 0.5)\n",
    "# axH.stairs(bH/np.max(bH),bB, fill=True, color = 'b', alpha = 0.5)\n",
    "\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "dPath = '/Users/angueyraaristjm/Documents/LiLab/Conferences/2021_RCS/resources/'\n",
    "# plt.savefig(dPath + \"CRtbx2F0_LM_hist.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "print('# cells of bright GFP cells: wt={0}, tbx2a={1}, tbx2b={2}'.format(wGBright*1000,aGBright*1000,bGBright*1000))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### After establishing threshold based on all data, count number of cells exceeding threshold for each eye"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all results\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "                \n",
    "rName = 'lGtile'\n",
    "\n",
    "wtResults = []\n",
    "i=1;\n",
    "for fPath in wt_fPaths:\n",
    "    gene = 'wt'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     wtResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    wtResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "    \n",
    "aResults = []\n",
    "i=1;\n",
    "for fPath in tbx2a_fPaths:\n",
    "    gene = 'tbx2a'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     aResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    aResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "bResults = []\n",
    "i=1;\n",
    "for fPath in tbx2b_fPaths:\n",
    "    gene = 'tbx2b'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     bResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    bResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "geneList = ['wt','tbx2a','tbx2b']\n",
    "\n",
    "\n",
    "plotname = ''\n",
    "fH, axH = plt.subplots(figsize= [6,8])\n",
    "\n",
    "barW = np.size(geneList)+2; # bar width\n",
    "barD = np.size(geneList)+1; # bar distance whitin 1 photoreceptor subtype\n",
    "barP = -np.size(geneList)+1; # position in x-axis\n",
    "barStep = 0.9; # distance between photoreceptor groups\n",
    "barPos = []; #array to save bar positions\n",
    "\n",
    "\n",
    "jitter = estimateJitter(wtResults)/400\n",
    "pH = plt.bar([barStep+1/barD], np.mean(wtResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+1/barD], np.mean(wtResults), yerr = [[0.025],[np.std(wtResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(wtResults))*[barStep+1/barD]+jitter, wtResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['wt'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+1/barD))\n",
    "\n",
    "\n",
    "\n",
    "jitter = estimateJitter(aResults)/400\n",
    "pH = plt.bar([barStep+2/barD], np.mean(aResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+2/barD], np.mean(aResults), yerr = [[0.025],[np.std(aResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(aResults))*[barStep+2/barD]+jitter, aResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['tbx2a'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+2/barD))\n",
    "\n",
    "jitter = estimateJitter(bResults)/400\n",
    "pH = plt.bar([barStep+3/barD], np.mean(bResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+3/barD], np.mean(bResults), yerr = [[0.025],[np.std(bResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(bResults))*[barStep+3/barD]+jitter, bResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['tbx2b'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+3/barD))\n",
    "\n",
    "formatFigureMain(fH, axH, pH)\n",
    "\n",
    "# axH.set_ylabel('GFP+ L cones per 64 x 64 $\\mu$m$^2$', fontproperties=formatFigureMain(fH, axH, pH))\n",
    "# axH.set_ylim([0,180]);\n",
    "# stath = 170\n",
    "\n",
    "axH.set_ylabel('Fraction of GFP+ L cones', fontproperties=formatFigureMain(fH, axH, pH))\n",
    "axH.set_ylim([0,1]);\n",
    "stath = 0.85\n",
    "\n",
    "geneList = ['control','F$\\emptyset$[tbx2a]','F$\\emptyset$[tbx2b]']\n",
    "\n",
    "axH.xaxis.set_tick_params(rotation=45)\n",
    "axH.set_xticks(np.sort(barPos));\n",
    "axH.set_xticklabels(geneList);\n",
    "\n",
    "hH = plt.hlines(stath, barPos[0], barPos[1], colors='#000000',linewidth=4)\n",
    "plt.text((barPos[0]+barPos[1])/2, stath, '*', font_properties=font_prop, fontsize=36, ha='center')# plt.savefig(savePath + \"Counts_tbx2_RU.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "\n",
    "dPath = '/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Presentations/revealjs/resources/20211008_UCLA/'\n",
    "plt.savefig(dPath + \"CRtbx2F0_GFPLcones.svg\", transparent=True, format=\"svg\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('wt fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(wtResults)*100,np.std(wtResults)*100))\n",
    "print('tbx2a fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(aResults)*100,np.std(aResults)*100))\n",
    "print('tbx2b fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(bResults)*100,np.std(bResults)*100))\n",
    "# Mann Whitney \n",
    "Au, pA = mannwhitneyu(wtResults, aResults)\n",
    "print('wt vs. tbx2a:')\n",
    "print('\\tU = {0:.3f}, p = {1:.5f}'.format(Au,pA))\n",
    "Bu, pB = mannwhitneyu(wtResults, bResults)\n",
    "print('wt vs. tbx2b:')\n",
    "print('\\tU = {0:.3f}, p = {1:.5f}'.format(Bu,pB))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# divide images into 1000 bins\n",
    "mH, mB = np.histogram(viewer.layers['M_mip'].data.flatten(), bins = 1000)\n",
    "lH, lB = np.histogram(viewer.layers['L_mip'].data.flatten(), bins = 1000)\n",
    "\n",
    "\n",
    "fH, axH = plt.subplots(figsize=(8,8))\n",
    "axH.stairs(lH,lB, fill=True, color = 'm', alpha = 0.5)\n",
    "axH.stairs(mH,mB, fill=True, color = 'g', alpha = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "# change keybind for default actions:\n",
    "@viewer.bind_key('F1', overwrite=True)\n",
    "def eraserMode(viewer):\n",
    "    viewer.layers.selection.active.mode = 'fill'\n",
    "    viewer.layers.selection.active.mode = 'paint'\n",
    "    viewer.layers.selection.active.mode = 'erase'\n",
    "    viewer.layers.selection.active.mode = 'pick'\n",
    "    viewer.layers.selection.active.mode = 'pan_zoom'\n",
    "\n",
    "# napari native functions for other actions\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\n",
    "        \"Set the currently selected label to the largest used label plus one.\"\n",
    "    ),\n",
    ")\n",
    "def new_label(layer):\n",
    "    \"\"\"Set the currently selected label to the largest used label plus one.\"\"\"\n",
    "    layer.selected_label = layer.data.max() + 1\n",
    "\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\"Decrease the currently selected label by one.\"),\n",
    ")\n",
    "def decrease_label_id(layer):\n",
    "    layer.selected_label -= 1\n",
    "\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\"Increase the currently selected label by one.\"),\n",
    ")\n",
    "def increase_label_id(layer):\n",
    "    layer.selected_label += 1\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
