{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# [NAPARI](https://napari.org/) testing\n",
    "## and adding [cellpose](https://cellpose-napari.readthedocs.io/en/latest/index.html) (installer [here](https://pypi.org/project/cellpose-napari/0.1.3/))\n",
    "\n",
    "To upgrade run in a cell:\n",
    "```python\n",
    "!pip install cellpose-napari==0.1.3\n",
    "!pip install cellpose --upgrade\n",
    "```\n",
    "\n",
    "```bash\n",
    "napari -w cellpose-napari\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic usage:\n",
    "```python\n",
    "viewer = napari.view_image(data.astronaut(), rgb=True)\n",
    "nbscreenshot(viewer)\n",
    "viewer.close()\n",
    "```\n",
    "\n",
    "example notebooks [here](https://github.com/sofroniewn/napari-training-course/blob/master/lessons/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting napari-stl-exporter\n",
      "  Downloading napari_stl_exporter-0.0.5-py3-none-any.whl (7.8 kB)\n",
      "Collecting numpy-stl\n",
      "  Downloading numpy-stl-2.16.3.tar.gz (772 kB)\n",
      "     |████████████████████████████████| 772 kB 3.4 MB/s            \n",
      "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: scikit-image in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from napari-stl-exporter) (0.18.1)\n",
      "Requirement already satisfied: numpy in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from napari-stl-exporter) (1.21.2)\n",
      "Requirement already satisfied: napari-plugin-engine>=0.1.4 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from napari-stl-exporter) (0.1.9)\n",
      "Requirement already satisfied: importlib-metadata>=1.5.0 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from napari-plugin-engine>=0.1.4->napari-stl-exporter) (4.8.1)\n",
      "Collecting python-utils>=1.6.2\n",
      "  Downloading python_utils-2.7.1-py2.py3-none-any.whl (14 kB)\n",
      "Requirement already satisfied: scipy>=1.0.1 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (1.7.1)\n",
      "Requirement already satisfied: matplotlib!=3.0.0,>=2.0.0 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (3.4.2)\n",
      "Requirement already satisfied: networkx>=2.0 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (2.6.3)\n",
      "Requirement already satisfied: pillow!=7.1.0,!=7.1.1,>=4.3.0 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (8.2.0)\n",
      "Requirement already satisfied: imageio>=2.3.0 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (2.9.0)\n",
      "Requirement already satisfied: tifffile>=2019.7.26 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (2021.7.2)\n",
      "Requirement already satisfied: PyWavelets>=1.1.1 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from scikit-image->napari-stl-exporter) (1.1.1)\n",
      "Requirement already satisfied: typing-extensions>=3.6.4 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from importlib-metadata>=1.5.0->napari-plugin-engine>=0.1.4->napari-stl-exporter) (3.10.0.2)\n",
      "Requirement already satisfied: zipp>=0.5 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from importlib-metadata>=1.5.0->napari-plugin-engine>=0.1.4->napari-stl-exporter) (3.6.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->napari-stl-exporter) (2.8.2)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->napari-stl-exporter) (0.10.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->napari-stl-exporter) (1.3.1)\n",
      "Requirement already satisfied: pyparsing>=2.2.1 in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from matplotlib!=3.0.0,>=2.0.0->scikit-image->napari-stl-exporter) (2.4.7)\n",
      "Requirement already satisfied: six in /Users/angueyraaristjm/anaconda3/lib/python3.7/site-packages (from python-utils>=1.6.2->numpy-stl->napari-stl-exporter) (1.16.0)\n",
      "Building wheels for collected packages: numpy-stl\n",
      "  Building wheel for numpy-stl (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for numpy-stl: filename=numpy_stl-2.16.3-cp37-cp37m-macosx_10_9_x86_64.whl size=18694 sha256=661dbf9f65c3e4964c199c922948afb509e97f4e3699349b89e8d8305ca1bf99\n",
      "  Stored in directory: /Users/angueyraaristjm/Library/Caches/pip/wheels/06/f4/db/7fac39962a6ba79b7e740892042332083924bff552d4bef41e\n",
      "Successfully built numpy-stl\n",
      "Installing collected packages: python-utils, numpy-stl, napari-stl-exporter\n",
      "Successfully installed napari-stl-exporter-0.0.5 numpy-stl-2.16.3 python-utils-2.7.1\n"
     ]
    }
   ],
   "source": [
    "## to install\n",
    "# !pip install napari[all]\n",
    "# !pip install QT\n",
    "# !pip install napari[pyqt5]\n",
    "# !pip install cellpose-napari==0.1.4\n",
    "# !pip install napari-nikon-nd2\n",
    "# !pip install magicgui\n",
    "# !pip install napari-pyclesperanto-assistant\n",
    "# !pip install napari-animated-gif-io\n",
    "# !pip install napari-clusters-plotter #hdbscan failed\n",
    "# !pip install napari-plot-profile\n",
    "# !pip install napari-brightness-contrast\n",
    "# !pip install napari-curtain\n",
    "# !pip install napari-3d-ortho-viewer\n",
    "# !pip install napari-manual-split-and-merge-labels\n",
    "# !pip install napari-crop\n",
    "# !pip install napari-stl-exporter\n",
    "\n",
    "## to upgrade:\n",
    "# !pip install cellpose-napari==0.1.4\n",
    "# !pip install cellpose-napari --upgrade\n",
    "# !pip install cellpose --upgrade\n",
    "# !pip install napari --upgrade\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import napari\n",
    "import cellpose_napari\n",
    "from napari.utils import nbscreenshot\n",
    "from tifffile import imread\n",
    "import numpy as np\n",
    "from scipy import ndimage\n",
    "from scipy.stats import mannwhitneyu\n",
    "import napari_nikon_nd2\n",
    "import os \n",
    "from magicgui import magicgui\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "from cmcrameri import cm #colormaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotStyle = 'dark';\n",
    "plotStyle = 'light';\n",
    "\n",
    "\n",
    "batlow = cm.vik(range(255))\n",
    "batlow = batlow[0:255:int(np.floor(255/8)),:]\n",
    "matplotlib.rcParams['axes.prop_cycle'] = matplotlib.cycler(color=batlow) \n",
    "\n",
    "#gene Colors\n",
    "zfC = {\n",
    "    'rho'  : '#7d7d7d',\n",
    "    'sws1' : '#B73AB9',\n",
    "    'sws2' : '#4364F6',\n",
    "    'mws1' : '#59CB3B',\n",
    "    'mws2' : '#59CB3B',\n",
    "    'mws3' : '#59CB3B',\n",
    "    'mws4' : '#59CB3B',\n",
    "    'lws1' : '#CE2A22',\n",
    "    'lws2' : '#CE2A22',\n",
    "    'actb2': '#926645',\n",
    "    'tbx2a': '#c92675',\n",
    "    'tbx2b': '#7526c9',\n",
    "    'six7' : '#d6ab00',\n",
    "}\n",
    "\n",
    "zfG = {\n",
    "    'wt' : '#000000',\n",
    "    'tbx2a' : '#ab266b',\n",
    "    'tbx2b' : '#421f8e',\n",
    "    'foxq2' : '#001dd6',\n",
    "}\n",
    "\n",
    "zfGm = {\n",
    "    'wt' : 'o',\n",
    "    'tbx2a' : 'P',\n",
    "    'tbx2b' : 'X',\n",
    "    'foxq2' : '^',\n",
    "}\n",
    "\n",
    "prLabel = {\n",
    "    'R'  : 'R',\n",
    "    'U' : 'UV',\n",
    "    'S' : 'S',\n",
    "    'M' : 'M',\n",
    "    'L' : 'L',\n",
    "}\n",
    "\n",
    "if plotStyle=='dark':\n",
    "    # dark background\n",
    "    params = {\"ytick.color\" : \"w\",\n",
    "              \"xtick.color\" : \"w\",\n",
    "              \"axes.labelcolor\" : \"w\",\n",
    "              \"axes.edgecolor\" : \"w\",\n",
    "             \"axes.linewidth\" : 3,\n",
    "             \"xtick.major.width\" : 3,\n",
    "             \"ytick.major.width\" : 3,\n",
    "             \"xtick.major.size\" : 8,\n",
    "             \"ytick.major.size\" : 8,\n",
    "             \"text.color\" : \"w\"}\n",
    "    plt.rcParams.update(params)\n",
    "elif plotStyle=='light':\n",
    "    # white background\n",
    "    params = {\"ytick.color\" : \"k\",\n",
    "              \"xtick.color\" : \"k\",\n",
    "              \"axes.labelcolor\" : \"k\",\n",
    "              \"axes.edgecolor\" : \"k\",\n",
    "             \"axes.linewidth\" : 3,\n",
    "             \"xtick.major.width\" : 3,\n",
    "             \"ytick.major.width\" : 3,\n",
    "             \"xtick.major.size\" : 8,\n",
    "             \"ytick.major.size\" : 8,\n",
    "             \"text.color\" : \"k\"}\n",
    "    plt.rcParams.update(params)\n",
    "font_prop = font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc')\n",
    "\n",
    "\n",
    "def formatFigureMain(figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=30)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=36)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=32)\n",
    "    axH.set_xscale('linear')\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.set_xlabel(axH.get_xlabel(), fontproperties = fontTicks)\n",
    "    axH.set_ylabel(axH.get_ylabel(), fontproperties = fontTicks)\n",
    "    return fontLabels\n",
    "\n",
    "def lighten_color(color, amount=0.5):\n",
    "    \"\"\"\n",
    "    Lightens the given color by multiplying (1-luminosity) by the given amount.\n",
    "    Input can be matplotlib color string, hex string, or RGB tuple.\n",
    "\n",
    "    Examples:\n",
    "    >> lighten_color('g', 0.3)\n",
    "    >> lighten_color('#F034A3', 0.6)\n",
    "    >> lighten_color((.3,.55,.1), 0.5)\n",
    "    \"\"\"\n",
    "    import colorsys\n",
    "    try:\n",
    "        c = matplotlib.colors.cnames[color]\n",
    "    except:\n",
    "        c = color\n",
    "    c = colorsys.rgb_to_hls(*matplotlib.colors.to_rgb(c))\n",
    "    return matplotlib.colors.rgb2hex(colorsys.hls_to_rgb(c[0], 1 - amount * (1 - c[1]), c[2]))\n",
    "\n",
    "def estimateJitter(dataArray):\n",
    "    \"\"\" creates random jitter scaled by local density of points\"\"\"\n",
    "    from scipy.stats import gaussian_kde\n",
    "    kde = gaussian_kde(dataArray)\n",
    "    density = kde(dataArray)\n",
    "    jitter = np.random.randn(len(dataArray))*density\n",
    "    return jitter\n",
    "\n",
    "# viewer = napari.Viewer()\n",
    "\n",
    "print('Opened viewer; plot style is ' + plotStyle)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Test segmentation (initial](#testSeg)\n",
    "- [Extraction of z planes for analysis](#zExtract)\n",
    "- [Segmentation with cellpose](#cellSeg)\n",
    "- [Manual correction of segmentation](#manualCuration)\n",
    "- [Create thumbnails _WIP_](#thumbnails)\n",
    "- [Quantification](#quantification)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Testing segmentation<a id='testSeg'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dPath = \"/Users/angueyraaristjm/Documents/LiImaging/A1R/zf_partial/20201009_CRtbx2a/\"\n",
    "fPathL = \"20201009_i06b_L\"\n",
    "fPathM = \"20201009_i06b_M\"\n",
    "# load the image data and inspect its shape\n",
    "zfL = imread(dPath+fPathL + \".tiff\")\n",
    "zfM = imread(dPath+fPathM + \".tiff\")\n",
    "viewer.add_image(zfL, colormap='magenta')#, contrast_limits=[196,5860]);\n",
    "viewer.add_image(zfM, colormap='green', blending='additive')#, contrast_limits=[223,6721]);\n",
    "viewer.open(dPath+fPathL + \"_seg.tiff\", name='zfL_seg', plugin='builtins');\n",
    "viewer.open(dPath+fPathM + \"_seg.tiff\", name='zfM_seg', plugin='builtins');\n",
    "# nbscreenshot(viewer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nL = len(np.unique(viewer.layers['zfL_seg'].data))-1\n",
    "print(\"Image has {0} L cones\".format(nL))\n",
    "nM = len(np.unique(viewer.layers['zfM_seg'].data))-1\n",
    "print(\"Image has {0} M cones\".format(nM))\n",
    "\n",
    "dL = viewer.layers['zfL_seg'].data\n",
    "dM = viewer.layers['zfM_seg'].data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure how to normalize across images...\n",
    "1. average GFP instensity inside M-cones? &rarr; difficult in CRISPants\n",
    "1. max GFP intensity? &rarr; could be ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# eroding segmentation because it too generous with ROIs and I want to exclude GFP diffraction\n",
    "viewer.layers['zfL_seg'].data = ndimage.grey_erosion(dL, size=(8,8))\n",
    "viewer.layers['zfM_seg'].data = ndimage.grey_erosion(dM, size=(8,8))\n",
    "\n",
    "L_avgG = np.zeros([nL,1])\n",
    "L_avgR = np.zeros([nL,1])\n",
    "# calculate mean GFP intensity inside each L cone (normalized and baseline subtracted)\n",
    "maxM = viewer.layers['zfM'].data[viewer.layers['zfL_seg'].data == 0].max()\n",
    "maxL = viewer.layers['zfL'].data[viewer.layers['zfL_seg'].data == 1].max()\n",
    "minM = viewer.layers['zfM'].data[viewer.layers['zfL_seg'].data == 0].min()\n",
    "minL = viewer.layers['zfL'].data[viewer.layers['zfL_seg'].data == 0].min()\n",
    "for roi in range(nL):\n",
    "    tempROI = viewer.layers['zfL_seg'].data == roi\n",
    "    L_avgG[roi] = (viewer.layers['zfM'].data[tempROI].mean()-minM)/(maxM-minM)\n",
    "    L_avgR[roi] = (viewer.layers['zfL'].data[tempROI].mean()-minL)/(maxL-minL)\n",
    "\n",
    "M_avgG = np.zeros([nM,1])\n",
    "M_avgR = np.zeros([nM,1])\n",
    "# calculate mean GFP intensity inside each M cone (normalized and baseline subtracted)\n",
    "for roi in range(nM):\n",
    "    tempROI = viewer.layers['zfM_seg'].data == roi\n",
    "    M_avgG[roi] = (viewer.layers['zfM'].data[tempROI].mean()-minM)/(maxM-minM)\n",
    "    M_avgR[roi] = (viewer.layers['zfL'].data[tempROI].mean()-minL)/(maxL-minL)\n",
    "\n",
    "    \n",
    "nbins = 20\n",
    "denseFlag = True\n",
    "    \n",
    "plt.figure(figsize=(8,6))\n",
    "plt.hist(L_avgG, bins=np.divide(range(0,nbins+1),nbins), density = denseFlag, color = (1.0,0.5,0.0,0.5), label=\"L_GFP\")\n",
    "plt.hist(L_avgR, bins=np.divide(range(0,nbins+1),nbins), density = denseFlag, color = (1.0,0.0,0.0,0.5), label=\"L_tdT\")\n",
    "\n",
    "plt.hist(M_avgG, bins=np.divide(range(0,nbins+1),nbins), density = denseFlag, color = (0.0,1.0,0.0,0.5), label=\"M_GFP\")\n",
    "plt.hist(M_avgR, bins=np.divide(range(0,nbins+1),nbins), density = denseFlag, color = (0.5,1.0,0.0,0.5), label=\"M_tdT\")\n",
    "\n",
    "plt.xlabel(\"Normalized intensity\", size=14)\n",
    "plt.ylabel(\"Probability\", size=14)\n",
    "plt.title(\"\")\n",
    "plt.legend(loc='upper right')\n",
    "# plt.savefig(\"overlapping_histograms_with_matplotlib_Python.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer.layers['zfM_seg'].data = dM\n",
    "viewer.layers['zfL_seg'].data = ndimage.grey_erosion(viewer.layers['zfL_seg'].data, size=(10,10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@viewer.bind_key('k')\n",
    "def toggle_sel(viewer):\n",
    "    if (viewer.layers['zfM_seg'].show_selected_label == True):\n",
    "        viewer.layers['zfM_seg'].show_selected_label = False\n",
    "    elif (viewer.layers['zfM_seg'].show_selected_label == False):\n",
    "        viewer.layers['zfM_seg'].show_selected_label = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.layers['zfM_seg'].show_selected_label=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.layers['zfM_seg'].data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Extract layers from z-stacks<a id='zExtract'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "\n",
    "# uninjected\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201009_CRtbx2a/\"\n",
    "# ### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09b\"; zlims = [39,43]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09d\"; zlims = [31,34]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10b\"; zlims = [33,36]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10d\"; zlims = [38,41]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11b\"; zlims = [25,27]; gene = 'wt'; # this one could be example\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11d\"; zlims = [31,34]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12b\"; zlims = [23,25]; gene = 'wt';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12d\"; zlims = [31,34]; gene = 'wt';\n",
    "\n",
    "# tbx2a\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201009_CRtbx2a/\"\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a';\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\"; zlims = [00,00]; gene = 'tbx2a'; #very curved stack. would require oblique cut\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\"; zlims = [30,33]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\"; zlims = [30,33]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\"; zlims = [29,32]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\"; zlims = [36,38]; gene = 'tbx2a'; # this could be example\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\"; zlims = [45,47]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\"; zlims = [20,23]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\"; zlims = [28,30]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\"; zlims = [23,25]; gene = 'tbx2a';\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\"; zlims = [36,38]; gene = 'tbx2a'; # this in lens up\n",
    "\n",
    "# #tbx2b\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201030_CRtbx2ab/\"\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\"; zlims = [24,26]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\"; zlims = [37,39]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\"; zlims = [20,24]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\"; zlims = [27,29]; gene = 'tbx2b';\n",
    "\n",
    "# uninjected\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/A1R/zf2021onwards/20210716_tbx2F0s/\"\n",
    "### 20210716: imaged 5dpf larvae to increase n specifically for tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101b\"; zlims = [00,00]; zlims2 = [9,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101d\"; zlims = [00,00]; zlims2 = [17,27]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L102b\"; zlims = [00,00]; zlims2 = [19,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103b\"; zlims = [00,00]; zlims2 = [16,22]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "\n",
    "# # tbx2a\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101b\"; zlims = [25,27]; zlims2 = [11,17]; gene = 'tbx2a'; # might be difficult to segment\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101d\"; zlims = [30,32]; zlims2 = [18,20]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102b\"; zlims = [38,41]; zlims2 = [28,33]; gene = 'tbx2a'; # lots of M cones south?\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102d\"; zlims = [38,42]; zlims2 = [21,31]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103b\"; zlims = [38,40]; zlims2 = [26,29]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103d\"; zlims = [27,29]; zlims2 = [11,13]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L104b\"; zlims = [23,25]; zlims2 = [11,14]; gene = 'tbx2a';\n",
    "### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105b\"; zlims = [21,23]; zlims2 = [14,19]; gene = 'tbx2a'; # huge rip. exclude from analysis\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105d\"; zlims = [31,33]; zlims2 = [22,27]; gene = 'tbx2a'; # this one has to be wt\n",
    "### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106b\"; zlims = [27,38]; zlims2 = [15,30]; gene = 'tbx2a'; # too curved, makes cells overlap in mip. exclude from analysis\n",
    "### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106d\"; zlims = [24,29]; zlims2 = [6,26]; gene = 'tbx2a'; # alwo too curved, makes cells overlap in mip. exclude from analysis\n",
    "\n",
    "# # tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101b\"; gene = 'tbx2b'; zlims = [27,30]; zlims2 = [13,16];\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,21]; # somewhat curved. Exclude\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102b\"; gene = 'tbx2b'; zlims = [23,27]; zlims2 = [11,16]; # this could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102d\"; gene = 'tbx2b'; zlims = [24,27]; zlims2 = [8,13]; # good too\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103b\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103d\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104b\"; gene = 'tbx2b'; zlims = [25,27]; zlims2 = [8,13];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,22];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105b\"; gene = 'tbx2b'; zlims = [26,28]; zlims2 = [9,15];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105d\"; gene = 'tbx2b'; zlims = [21,24]; zlims2 = [8,13];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106b\"; gene = 'tbx2b'; zlims = [34,37]; zlims2 = [15,22];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106d\"; gene = 'tbx2b'; zlims = [35,38]; zlims2 = [23,27];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107b\"; gene = 'tbx2b'; zlims = [26,30]; zlims2 = [13,17];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [14,17];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108b\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [16,21];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108d\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [19,22]; # really flat. could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L109b\"; gene = 'tbx2b'; zlims = [32,34]; zlims2 = [17,21];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110b\"; gene = 'tbx2b'; zlims = [24,26]; zlims2 = [8,14];\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110d\"; gene = 'tbx2b'; # side-view; really good one\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L111b\"; gene = 'tbx2b'; zlims = [30,33]; zlims2 = [15,17]; # really flat. could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L112b\"; gene = 'tbx2b'; zlims = [28,30]; zlims2 = [14,18]; # mutant? confirm genotype\n",
    "\n",
    "\n",
    "\n",
    "# create subdirectory for analysis\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "if (os.path.isdir(dOut)==False):\n",
    "    os.mkdir(dOut)\n",
    "\n",
    "print('Canvas clear')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load whole stack to figure out best planes for cellpose\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201030_CRtbx2ab/\"\n",
    "viewer.open((dPath+fPath + \".nd2\"))\n",
    "# viewer.layers.remove(viewer.layers[2]) # remove transmitted detector image\n",
    "viewer.layers.remove(viewer.layers[3]) # remove transmitted detector image\n",
    "viewer.layers[0].colormap = 'green'\n",
    "viewer.layers[0].name = 'M'\n",
    "viewer.layers[1].colormap = 'magenta'\n",
    "viewer.layers[1].name = 'L'\n",
    "viewer.layers[2].colormap = 'bop blue'\n",
    "viewer.layers[2].name = 'LWS'\n",
    "viewer.layers.select_next();\n",
    "print('Loaded!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make mips (and remove any previous ones)\n",
    "# for FPs ------- # for OSs # code for eyes with 1D4 immuno\n",
    "zlims = [28,30]; zlims2 = [14,18];\n",
    "if len(viewer.layers)>3:\n",
    "    for l in viewer.layers[3:]:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "for l in [viewer.layers[0],viewer.layers[1]]:\n",
    "    l.visible = False\n",
    "#     viewer.layers.remove(l.name + '_mip')\n",
    "    viewer.add_image(l.data[zlims[0]:zlims[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")\n",
    "    \n",
    "for l in [viewer.layers[2]]:\n",
    "    l.visible = False\n",
    "#     viewer.layers.remove(l.name + '_mip')\n",
    "    viewer.add_image(l.data[zlims2[0]:zlims2[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")\n",
    "\n",
    "# # code for eyes without immuno\n",
    "# # zlims = [45,47];\n",
    "# if len(viewer.layers)>2:\n",
    "#     for l in viewer.layers[2:]:\n",
    "#         viewer.layers.remove(l)\n",
    "\n",
    "# for l in [viewer.layers[0],viewer.layers[1]]:\n",
    "#     l.visible = False\n",
    "# #     viewer.layers.remove(l.name + '_mip')\n",
    "#     viewer.add_image(l.data[zlims[0]:zlims[1]].max(axis=0), blending='additive', colormap = l.colormap, name = l.name + \"_mip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# when things look good, save in folder for batch cellpose analysis\n",
    "l = viewer.layers['M_mip']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers['L_mip']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers['LWS_mip']; l.save(dOut + l.name + '.tiff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now go back to next file [&uarr;](#zExtract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Segmentation using cellpose<a id='cellSeg'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201009_CRtbx2a/\"\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09b\"; zlims = [39,43]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09d\"; zlims = [31,34]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10b\"; zlims = [33,36]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10d\"; zlims = [38,41]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11b\"; zlims = [25,27]; gene = 'wt'; # this one could be example\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11d\"; zlims = [31,34]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12b\"; zlims = [23,25]; gene = 'wt';\n",
    "fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12d\"; zlims = [31,34]; gene = 'wt'; \n",
    "\n",
    "# tbx2a\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201009_CRtbx2a/\"\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a';\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\"; zlims = [00,00]; gene = 'tbx2a'; #very curved stack. would require oblique cut\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\"; zlims = [30,33]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\"; zlims = [30,33]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\"; zlims = [29,32]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\"; zlims = [45,47]; gene = 'tbx2a'; # this could be example\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\"; zlims = [26,31]; gene = 'tbx2a'; \n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\"; zlims = [20,23]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\"; zlims = [28,30]; gene = 'tbx2a';\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\"; zlims = [23,25]; gene = 'tbx2a';\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\"; zlims = [36,38]; gene = 'tbx2a'; # this in lens up\n",
    "\n",
    "# # tbx2b\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201030_CRtbx2ab/\"\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\"; zlims = [24,26]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\"; zlims = [37,39]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\"; zlims = [20,24]; gene = 'tbx2b';\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\"; zlims = [27,29]; gene = 'tbx2b';\n",
    "\n",
    "# uninjected\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/A1R/zf2021onwards/20210716_tbx2F0s/\"\n",
    "### 20210716: imaged 5dpf larvae to increase n specifically for tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101b\"; zlims = [00,00]; zlims2 = [9,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101d\"; zlims = [00,00]; zlims2 = [17,27]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L102b\"; zlims = [00,00]; zlims2 = [19,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103b\"; zlims = [00,00]; zlims2 = [16,22]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "\n",
    "# # tbx2a\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101b\"; zlims = [25,27]; zlims2 = [11,17]; gene = 'tbx2a'; # might be difficult to segment\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101d\"; zlims = [30,32]; zlims2 = [18,20]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102b\"; zlims = [38,41]; zlims2 = [28,33]; gene = 'tbx2a'; # lots of M cones south?\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102d\"; zlims = [38,42]; zlims2 = [21,31]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103b\"; zlims = [38,40]; zlims2 = [26,29]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103d\"; zlims = [27,29]; zlims2 = [11,13]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L104b\"; zlims = [23,25]; zlims2 = [11,14]; gene = 'tbx2a';\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105b\"; zlims = [21,23]; zlims2 = [14,19]; gene = 'tbx2a'; # huge rip. exclude from analysis\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105d\"; zlims = [31,33]; zlims2 = [22,27]; gene = 'tbx2a'; # this one has to be wt\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106b\"; zlims = [27,38]; zlims2 = [15,30]; gene = 'tbx2a'; # too curved, makes cells overlap in mip. exclude from analysis\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106d\"; zlims = [24,29]; zlims2 = [6,26]; gene = 'tbx2a'; # alwo too curved, makes cells overlap in mip. exclude from analysis\n",
    "\n",
    "# # tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101b\"; gene = 'tbx2b'; zlims = [27,30]; zlims2 = [13,16];\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,21]; # somewhat curved. Exclude\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102b\"; gene = 'tbx2b'; zlims = [23,27]; zlims2 = [11,16]; # this could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102d\"; gene = 'tbx2b'; zlims = [24,27]; zlims2 = [8,13]; # good too\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103b\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103d\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104b\"; gene = 'tbx2b'; zlims = [25,27]; zlims2 = [8,13];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,22];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105b\"; gene = 'tbx2b'; zlims = [26,28]; zlims2 = [9,15];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105d\"; gene = 'tbx2b'; zlims = [21,24]; zlims2 = [8,13];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106b\"; gene = 'tbx2b'; zlims = [34,37]; zlims2 = [15,22];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106d\"; gene = 'tbx2b'; zlims = [35,38]; zlims2 = [23,27];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107b\"; gene = 'tbx2b'; zlims = [26,30]; zlims2 = [13,17];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [14,17];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108b\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [16,21];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108d\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [19,22]; # really flat. could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L109b\"; gene = 'tbx2b'; zlims = [32,34]; zlims2 = [17,21];\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110b\"; gene = 'tbx2b'; zlims = [24,26]; zlims2 = [8,14];\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110d\"; gene = 'tbx2b'; # side-view; really good one\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L111b\"; gene = 'tbx2b'; zlims = [30,33]; zlims2 = [15,17]; # really flat. could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L112b\"; gene = 'tbx2b'; zlims = [28,30]; zlims2 = [14,18]; # mutant? confirm genotype\n",
    "\n",
    "\n",
    "# create subdirectory for analysis\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "print('Viewer cleared...')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "\n",
    "# load mips\n",
    "viewer.open(dOut + 'M' + \"_mip.tiff\", plugin='builtins', colormap = 'green');\n",
    "viewer.open(dOut + 'L' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "print('Loaded! Now run cellpose!!!!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# after running cellpose, rename cellpose images and save before manual curation:\n",
    "baseName = 'M'\n",
    "lname = baseName + '_mip'\n",
    "viewer.layers[lname + '_cp_masks_000'].name = baseName + '_seg'\n",
    "viewer.layers[lname + '_cp_outlines_000'].name = baseName + '_outlines'\n",
    "viewer.layers[lname + '_cp_cellprob_000'].name = baseName + '_prob'\n",
    "viewer.layers[lname + '_cp_flows_000'].name = baseName + '_flows'\n",
    "\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_outlines']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_prob']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_flows']; l.save(dOut + l.name + '.tiff')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# after running cellpose, rename cellpose images and save before manual curation:\n",
    "baseName = 'L'\n",
    "lname = baseName + '_mip'\n",
    "viewer.layers[lname + '_cp_masks_000'].name = baseName + '_seg'\n",
    "viewer.layers[lname + '_cp_outlines_000'].name = baseName + '_outlines'\n",
    "viewer.layers[lname + '_cp_cellprob_000'].name = baseName + '_prob'\n",
    "viewer.layers[lname + '_cp_flows_000'].name = baseName + '_flows'\n",
    "\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_outlines']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_prob']; l.save(dOut + l.name + '.tiff')\n",
    "l = viewer.layers[baseName + '_flows']; l.save(dOut + l.name + '.tiff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Go back and run next one &uarr;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Manual correction of cellpose segmentation<a id='manualCuration'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear viewer\n",
    "viewer.layers.select_all()\n",
    "viewer.layers.remove_selected()\n",
    "# for l in viewer.layers:\n",
    "#         viewer.layers.remove(l)\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09b\"; zlims = [39,43]; gene = 'wt'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i09d\"; zlims = [31,34]; gene = 'wt'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10b\"; zlims = [33,36]; gene = 'wt'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i10d\"; zlims = [38,41]; gene = 'wt'; # done (L labels are a little generous)\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11b\"; zlims = [25,27]; gene = 'wt'; # done # this one could be example\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i11d\"; zlims = [31,34]; gene = 'wt'; # done (M and L seg are quite good; example too)\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12b\"; zlims = [23,25]; gene = 'wt'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_wt_i12d\"; zlims = [31,34]; gene = 'wt'; # done\n",
    "\n",
    "# # tbx2a\n",
    "# Extremely difficult to segment M in F0-larvae\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\"; zlims = [36,38]; gene = 'tbx2a'; # is this supposed to be d????\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\"; zlims = [00,00]; gene = 'tbx2a'; #very curved stack. would require oblique cut\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\"; zlims = [30,33]; gene = 'tbx2a'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\"; zlims = [30,33]; gene = 'tbx2a'; # done\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\"; zlims = [29,32]; gene = 'tbx2a'; # done (finished M cone segmentation too)\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\"; zlims = [45,47]; gene = 'tbx2a'; # done; this could be example\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\"; zlims = [26,31]; gene = 'tbx2a'; # done (finished M cone segmentation too)\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\"; zlims = [20,23]; gene = 'tbx2a'; # done; very clean L-cone seg\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\"; zlims = [28,30]; gene = 'tbx2a'; # done (finished M cone segmentation too)\n",
    "# fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\"; zlims = [23,25]; gene = 'tbx2a'; # done\n",
    "# #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\"; zlims = [36,38]; gene = 'tbx2a'; # this in lens up\n",
    "\n",
    "# # tbx2b\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\"; zlims = [24,26]; gene = 'tbx2b'; # done\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\"; zlims = [37,39]; gene = 'tbx2b'; # done (finished M-seg too)\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\"; zlims = [20,24]; gene = 'tbx2b'; # done\n",
    "# fPath = \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\"; zlims = [27,29]; gene = 'tbx2b';\n",
    "\n",
    "# uninjected\n",
    "dPath = \"/Volumes/angueyraNEI/LiImaging/A1R/zf2021onwards/20210716_tbx2F0s/\"\n",
    "### 20210716: imaged 5dpf larvae to increase n specifically for tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101b\"; zlims = [00,00]; zlims2 = [9,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L101d\"; zlims = [00,00]; zlims2 = [17,27]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L102b\"; zlims = [00,00]; zlims2 = [19,31]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103b\"; zlims = [00,00]; zlims2 = [16,22]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L103d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L104d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L105d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106b\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_uninj_L106d\"; zlims = [00,00]; zlims2 = [00,00]; gene = 'wt'; #segmented by Laura already, missing 1D4 mip\n",
    "\n",
    "# # tbx2a\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101b\"; zlims = [25,27]; zlims2 = [11,17]; gene = 'tbx2a'; # might be difficult to segment\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L101d\"; zlims = [30,32]; zlims2 = [18,20]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102b\"; zlims = [38,41]; zlims2 = [28,33]; gene = 'tbx2a'; # lots of M cones south?\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L102d\"; zlims = [38,42]; zlims2 = [21,31]; gene = 'tbx2a';\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103b\"; zlims = [38,40]; zlims2 = [26,29]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L103d\"; zlims = [27,29]; zlims2 = [11,13]; gene = 'tbx2a'; # this one has to be wt\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L104b\"; zlims = [23,25]; zlims2 = [11,14]; gene = 'tbx2a';\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105b\"; zlims = [21,23]; zlims2 = [14,19]; gene = 'tbx2a'; # huge rip. exclude from analysis\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L105d\"; zlims = [31,33]; zlims2 = [22,27]; gene = 'tbx2a'; # this one has to be wt\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106b\"; zlims = [27,38]; zlims2 = [15,30]; gene = 'tbx2a'; # too curved, makes cells overlap in mip. exclude from analysis\n",
    "# ### fPath = \"20210716_5dpf_m2G_t2T_tbx2a_L106d\"; zlims = [24,29]; zlims2 = [6,26]; gene = 'tbx2a'; # alwo too curved, makes cells overlap in mip. exclude from analysis\n",
    "\n",
    "# # tbx2b\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101b\"; gene = 'tbx2b'; zlims = [27,30]; zlims2 = [13,16]; #this one is messy and seems to have a lot of cones?\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L101d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,21]; # somewhat curved. Exclude\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102b\"; gene = 'tbx2b'; zlims = [23,27]; zlims2 = [11,16]; # this could be example; did M\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L102d\"; gene = 'tbx2b'; zlims = [24,27]; zlims2 = [8,13]; # good too; did M\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103b\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L103d\"; gene = 'tbx2b'; # not GFP+; exclude\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104b\"; gene = 'tbx2b'; zlims = [25,27]; zlims2 = [8,13]; # did M\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L104d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [11,22]; #did M\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105b\"; gene = 'tbx2b'; zlims = [26,28]; zlims2 = [9,15]; #did M; this segmentation is quite good\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L105d\"; gene = 'tbx2b'; zlims = [21,24]; zlims2 = [8,13]; #did M\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106b\"; gene = 'tbx2b'; zlims = [34,37]; zlims2 = [15,22]; #did M\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L106d\"; gene = 'tbx2b'; zlims = [35,38]; zlims2 = [23,27]; # note done yet\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107b\"; gene = 'tbx2b'; zlims = [26,30]; zlims2 = [13,17]; # rough segmentation; could also not be mutant.\n",
    "### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L107d\"; gene = 'tbx2b'; zlims = [29,32]; zlims2 = [14,17]; # skipping this one; hard segmentation\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108b\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [16,21]; # almost all tdTomato- cells are GFP+\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L108d\"; gene = 'tbx2b'; zlims = [31,34]; zlims2 = [19,22]; # really flat. could be example\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L109b\"; gene = 'tbx2b'; zlims = [32,34]; zlims2 = [17,21]; #v skipped M\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110b\"; gene = 'tbx2b'; zlims = [24,26]; zlims2 = [8,14];\n",
    "# # ### fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L110d\"; gene = 'tbx2b'; # side-view; really good one\n",
    "fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L111b\"; gene = 'tbx2b'; zlims = [30,33]; zlims2 = [15,17]; # really flat. really could be example\n",
    "# fPath = \"20210716_5dpf_m2G_t2T_tbx2b_L112b\"; gene = 'tbx2b'; zlims = [28,30]; zlims2 = [14,18]; # not done; mutant? confirm genotype\n",
    "\n",
    "# create subdirectory for analysis\n",
    "dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "print('Viewer cleared...')\n",
    "\n",
    "# clear key binds\n",
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    ...\n",
    "@viewer.bind_key('b', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    ...\n",
    "@viewer.bind_key('Shift-x', overwrite=True)\n",
    "def removeLabel(viewer):\n",
    "    ...\n",
    "\n",
    "# load mips\n",
    "viewer.open(dOut + 'M' + \"_mip.tiff\", plugin='builtins', colormap = 'green', blending='additive');\n",
    "viewer.open(dOut + 'L' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "viewer.open(dOut + 'M' + \"_seg.tiff\", plugin='builtins', blending='additive');\n",
    "viewer.open(dOut + 'L' + \"_seg.tiff\", plugin='builtins', blending='additive');\n",
    "viewer.layers['M_seg'].preserve_labels = True;\n",
    "viewer.layers['L_seg'].preserve_labels = True;\n",
    "\n",
    "# viewer.layers['M_seg'].contour = 5\n",
    "# viewer.layers['L_seg'].contour = 5\n",
    "\n",
    "viewer.layers['L_mip'].visible = False\n",
    "viewer.layers['L_seg'].visible = False\n",
    "\n",
    "\n",
    "print('Loaded  ' + fPath + ' !\\n Now fix it and save it!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define useful keyboard shortcuts\n",
    "\n",
    "@viewer.bind_key('Shift-x', overwrite=True)\n",
    "def removeLabel(viewer):\n",
    "    lname = 'M_seg'\n",
    "    tempd = viewer.layers[lname].data\n",
    "    tempd[tempd == viewer.layers[lname].selected_label]=0\n",
    "    viewer.layers[lname].data = tempd\n",
    "    print('Cut!')\n",
    "\n",
    "print('removeLabel on M_seg (\"Shift-X\")')\n",
    "\n",
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'M_seg'\n",
    "    if (viewer.layers[lname].preserve_labels == True):\n",
    "        viewer.layers[lname].preserve_labels = False\n",
    "    elif (viewer.layers[lname].preserve_labels == False):\n",
    "        viewer.layers[lname].preserve_labels = True\n",
    "print('toggle M_seg preserve_labels (\"k\")')\n",
    "        \n",
    "@viewer.bind_key('b', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'L_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "print('toggle L_mip visibility (\"B\")')\n",
    "\n",
    "@viewer.bind_key('n', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'M_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "print('toggle M_mip visibility (\"N\")')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load z-stack if needed\n",
    "# dPath = \"/Volumes/zfSSD/LiImaging/A1R/zf/20201030_CRtbx2ab/\"\n",
    "viewer.open((dPath+fPath + \".nd2\"))\n",
    "viewer.layers.remove(viewer.layers[fPath+' [2]']) # remove transmitted detector image\n",
    "viewer.layers.remove(viewer.layers[fPath+' [3]']) # remove immuno layer\n",
    "viewer.layers[fPath].colormap = 'magenta'\n",
    "viewer.layers[fPath].name = 'Mz'\n",
    "viewer.layers[fPath+' [1]'].colormap = 'green'\n",
    "viewer.layers[fPath+' [1]'].visible = False\n",
    "viewer.layers[fPath+' [1]'].name = 'Lz'\n",
    "viewer.layers.select_next();\n",
    "print('Loaded!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### resave curated segmentation after napari-ing around"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.layers[7]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseName = 'M'\n",
    "lname = baseName + '_seg'\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated_incomplete.tiff')\n",
    "# l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated.tiff')\n",
    "viewer.layers['L_mip'].visible = True\n",
    "viewer.layers['L_seg'].visible = True\n",
    "viewer.layers['M_seg'].contour = 0\n",
    "viewer.layers['M_mip'].visible = False\n",
    "viewer.layers['M_seg'].visible = False\n",
    "print('Done with ' + lname)\n",
    "\n",
    "@viewer.bind_key('Shift-x', overwrite=True)\n",
    "def removeLabel(viewer):\n",
    "#     lname = 'M_seg'mvvvv\n",
    "    lname = 'L_seg'\n",
    "    tempd = viewer.layers[lname].data\n",
    "    tempd[tempd == viewer.layers[lname].selected_label]=0\n",
    "    viewer.layers[lname].data = tempd\n",
    "    print('Cut!')\n",
    "\n",
    "print('removeLabel on L_seg (\"Shift-X\")')\n",
    "\n",
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'L_seg'\n",
    "    if (viewer.layers[lname].preserve_labels == True):\n",
    "        viewer.layers[lname].preserve_labels = False\n",
    "    elif (viewer.layers[lname].preserve_labels == False):\n",
    "        viewer.layers[lname].preserve_labels = True\n",
    "print('toggle L_seg preserve_labels (\"k\")')\n",
    "\n",
    "@viewer.bind_key('b', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'M_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "print('toggle L_mip visibility (\"B\")')\n",
    "\n",
    "        \n",
    "@viewer.bind_key('n', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'L_mip'\n",
    "    if (viewer.layers[lname].visible == True):\n",
    "        viewer.layers[lname].visible = False\n",
    "    elif (viewer.layers[lname].visible == False):\n",
    "        viewer.layers[lname].visible = True\n",
    "\n",
    "print('toggle L_mip visibility (\"N\")')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### resave curated segmentation after napari-ing around"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "baseName = 'L'\n",
    "lname = baseName + '_seg'\n",
    "l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated.tiff')\n",
    "print('Done with ' + lname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# baseName = 'M'\n",
    "# l = viewer.layers[baseName + '_seg']; l.save(dOut + l.name + '_curated.tiff')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reload curated segmentations after saving"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all except mips\n",
    "if len(viewer.layers)>2:\n",
    "    for l in viewer.layers[2:]:\n",
    "        viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.open(dOut + 'M' + \"_seg_curated_incomplete.tiff\", name='M_seg', plugin='builtins');\n",
    "viewer.open(dOut + 'L' + \"_seg_curated.tiff\", name='L_seg', plugin='builtins');\n",
    "viewer.layers['L_seg'].contour = 5\n",
    "\n",
    "print(fPath)\n",
    "nM = len(np.unique(viewer.layers['M_seg'].data))-1\n",
    "print(\"Image has {0} M cones\".format(nM))\n",
    "nR = len(np.unique(viewer.layers['L_seg'].data))-1\n",
    "print(\"Image has {0} L cones\".format(nR))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove all lauers\n",
    "if len(viewer.layers)>0:\n",
    "    for l in viewer.layers[0:]:\n",
    "        viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manual merging of labels\n",
    "mergeLabels = [167,168]\n",
    "# lname = 'M_seg'\n",
    "lname = 'L_seg'\n",
    "if viewer.layers[lname].visible == False:\n",
    "    print('Layer not active')\n",
    "else:\n",
    "    tempd = viewer.layers[lname].data\n",
    "    tempd[tempd == mergeLabels[1]]=mergeLabels[0]\n",
    "    viewer.layers[lname].data = tempd\n",
    "    print('Merged in ' + lname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# manually delete label\n",
    "selLabel = 49\n",
    "lname = 'L_seg'\n",
    "tempd = viewer.layers[lname].data\n",
    "tempd[tempd == selLabel]=0\n",
    "viewer.layers[lname].data = tempd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@viewer.bind_key('k', overwrite=True)\n",
    "def toggle_sel(viewer):\n",
    "    lname = 'M_seg'\n",
    "    if (viewer.layers[lname].show_selected_label == True):\n",
    "        viewer.layers[lname].show_selected_label = False\n",
    "    elif (viewer.layers[lname].show_selected_label == False):\n",
    "        viewer.layers[lname].show_selected_label = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.layers['M_seg'].show_selected_label = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## To Do: Batch export of png with mips and contours of curated segmentation<a id='thumbnails'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This could do the trick:\n",
    "```python\n",
    "def blended_img(viewer):\n",
    "    import napari\n",
    "    import numpy as np\n",
    "    \n",
    "    blended = np.zeros(viewer.layers[0].data.shape + (4,))\n",
    "    for layer in viewer.layers:\n",
    "        # normalize data by clims\n",
    "        normalized_data = (layer.data - layer.contrast_limits[0]) / (\n",
    "        layer.contrast_limits[1] - layer.contrast_limits[0]\n",
    "    )\n",
    "        colormapped_data = layer.colormap.map(normalized_data.flatten())\n",
    "        colormapped_data = colormapped_data.reshape(normalized_data.shape + (4,))\n",
    "\n",
    "        blended = blended + colormapped_data\n",
    "    \n",
    "    blended[..., 3] = 1 # set alpha channel to 1\n",
    "\n",
    "    return blended\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Quantification: GFP inside L cones<a id='quantification'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)\n",
    "\n",
    "Plan (loosely based on Mander's coefficient):\n",
    "1. Erode L cone segmentation to focus on center of cells\n",
    "2. Divide GFP fluorescence into millies as a way to normalize data between images into 0 - 1000 range\n",
    "3. Calculate for each cone what normalized average fluorescence corresponds too.\n",
    "\n",
    "Questions:\n",
    "1. How much erosion?\n",
    "2. Do I have to exclude L cones to do GFP percentile calculation? &rarr; this would assume that GFP outside L cones is unchanged, which could be tested|plotted directly.\n",
    "3. Also check if RFP distributions are unchanged?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calcMillie(viewer,lname):\n",
    "    outHist,outBins = np.histogram(viewer.layers[lname].data.flatten(), bins = 1000)\n",
    "    return outHist,outBins\n",
    "\n",
    "def GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=10):\n",
    "    # clear viewer\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "    viewer.layers.select_all()\n",
    "    viewer.layers.remove_selected()\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    print('Quantifying GFP in L cones of ' + fPath + '(' + gene + ')\\n')\n",
    "    # load mips\n",
    "    viewer.open(dOut + 'M' + \"_mip.tiff\", plugin='builtins', colormap = 'green', blending='additive');\n",
    "    viewer.open(dOut + 'L' + \"_mip.tiff\", plugin='builtins', colormap = 'magenta', blending='additive');\n",
    "    # viewer.open(dOut + 'M' + \"_seg_curated_incomplete.tiff\", name='M_seg', plugin='builtins', blending='additive');\n",
    "    viewer.open(dOut + 'L' + \"_seg_curated.tiff\", name='L_seg', plugin='builtins', blending='additive');\n",
    "    # viewer.layers['M_seg'].preserve_labels = True;\n",
    "    viewer.layers['L_seg'].preserve_labels = True;\n",
    "    viewer.layers['L_seg'].contour = 5\n",
    "    # erode L cone segmentation\n",
    "    viewer.layers['L_seg'].data = ndimage.grey_erosion(viewer.layers['L_seg'].data, size=(erosionValue,erosionValue))\n",
    "    # divide images into 1000 bins\n",
    "    mH, mB = calcMillie(viewer,'M_mip')\n",
    "    lH, lB = calcMillie(viewer,'L_mip')\n",
    "\n",
    "    lcones = np.unique(viewer.layers['L_seg'].data)\n",
    "    lcones = lcones[lcones>0]\n",
    "    lRFP = np.empty(np.shape(lcones))\n",
    "    lGFP = np.empty(np.shape(lcones))\n",
    "    lRFPsd = np.empty(np.shape(lcones))\n",
    "    lGFPsd = np.empty(np.shape(lcones))\n",
    "    lRtile = np.empty(np.shape(lcones))\n",
    "    lGtile = np.empty(np.shape(lcones))\n",
    "    \n",
    "    i=0;\n",
    "    for l in lcones:\n",
    "        lRFP[i] = np.mean(viewer.layers['L_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lGFP[i] = np.mean(viewer.layers['M_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lRFPsd[i] = np.std(viewer.layers['L_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lGFPsd[i] = np.std(viewer.layers['M_mip'].data[viewer.layers['L_seg'].data==l])\n",
    "        lRtile[i] = len(lB[lB<lRFP[i]])\n",
    "        lGtile[i] = len(mB[mB<lGFP[i]])\n",
    "        i=i+1;\n",
    "        \n",
    "    np.savez(dOut + 'quantificationGFP.npz',\n",
    "             lcones=lcones,\n",
    "             lRFP=lRFP,\n",
    "             lGFP=lGFP,\n",
    "             lRFPsd=lRFPsd,\n",
    "             lGFPsd=lGFPsd,\n",
    "             lRtile=lRtile,\n",
    "             lGtile=lGtile)\n",
    "    results = zip(['lcones','lRFP','lGFP','lRFPsd','lGFPsd','lRtile','lGtile'],\n",
    "                 [lcones,lRFP,lGFP,lRFPsd,lGFPsd,lRtile,lGtile])\n",
    "    \n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "erodeV = 10; #12;\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "for fPath in wt_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "    \n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "for fPath in tbx2a_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                 \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "\n",
    "for fPath in tbx2b_fPaths:\n",
    "    GFPanalyzer(viewer,dAnalysis,gene,fPath, erosionValue=erodeV)\n",
    "    for l in viewer.layers:\n",
    "        viewer.layers.remove(l)\n",
    "\n",
    "print('Finished analysis')\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for l in viewer.layers:\n",
    "    print(l)\n",
    "    viewer.layers.remove(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# viewer = napari.Viewer()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all results\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "                \n",
    "                \n",
    "rName = 'lGtile'\n",
    "\n",
    "wtResults = []\n",
    "i=1;\n",
    "for fPath in wt_fPaths:\n",
    "    gene = 'wt'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    wtResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "aResults = []\n",
    "i=1;\n",
    "for fPath in tbx2a_fPaths:\n",
    "    gene = 'tbx2a'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    aResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "bResults = []\n",
    "i=1;\n",
    "for fPath in tbx2b_fPaths:\n",
    "    gene = 'tbx2b'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "    bResults.append(results[rName])\n",
    "    i=i+1;\n",
    "\n",
    "\n",
    "# combine across images\n",
    "wtResults = np.concatenate(wtResults).ravel()\n",
    "aResults = np.concatenate(aResults).ravel()\n",
    "bResults = np.concatenate(bResults).ravel()\n",
    "\n",
    "# plot histogram\n",
    "pdfV = True\n",
    "# wtH, wtB = np.histogram(wtResults, bins = 50, density=pdfV)\n",
    "# aH, aB = np.histogram(aResults, bins = 50, density=pdfV)\n",
    "# bH, bB = np.histogram(bResults, bins = 50, density=pdfV)\n",
    "\n",
    "# plot histogram but first calculate a fixed bin number for all genotypes\n",
    "# allBins = np.histogram_bin_edges(np.concatenate([wtResults,aResults,bResults]).ravel(),bins='fd');\n",
    "allBins = np.arange(0,1010,10);\n",
    "allBinsC = (allBins[:-1] + allBins[1:]) / 2 # Centers\n",
    "wtH, wtB = np.histogram(wtResults, bins = allBins, density=pdfV)\n",
    "aH, aB = np.histogram(aResults, bins = allBins, density=pdfV)\n",
    "bH, bB = np.histogram(bResults, bins = allBins, density=pdfV)\n",
    "\n",
    "# # this can be used for cumulative distribution then normalized\n",
    "# wtH = np.cumsum(wtH)/np.max(np.cumsum(wtH));\n",
    "# aH = np.cumsum(aH)/np.max(np.cumsum(aH));\n",
    "# bH = np.cumsum(bH)/np.max(np.cumsum(bH));\n",
    "\n",
    "wt95CI = np.min(allBinsC[(np.cumsum(wtH))>.095])/1000;\n",
    "wGBright = np.sum(wtH[allBinsC>wt95CI*1000]);\n",
    "aGBright = np.sum(aH[allBinsC>wt95CI*1000]);\n",
    "bGBright = np.sum(bH[allBinsC>wt95CI*1000]);\n",
    "\n",
    "print('GFP threshold is {0}'.format(wt95CI))\n",
    "\n",
    "baseColor = 'k'\n",
    "if (plotStyle=='light'):\n",
    "    baseColor = 'k';\n",
    "\n",
    "# fH, (axH1,axH2,axH3) = plt.subplots(nrows=3, ncols=1, figsize=(10, 20))\n",
    "fH, (axH1,axH2,axH3) = plt.subplots(nrows=1, ncols=3, figsize=(30, 8))\n",
    "axH1.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH2.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH3.stairs(wtH,wtB/1000, fill=True, color = baseColor, alpha = 1.0)\n",
    "axH2.stairs(aH,aB/1000, fill=True, color = zfC['lws1'], alpha = 0.8)\n",
    "axH3.stairs(bH,bB/1000, fill=True, color = zfC['sws2'], alpha = 0.8)\n",
    "\n",
    "\n",
    "axH1.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "axH2.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "axH3.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "\n",
    "\n",
    "# axH.set_title('Distribution of GFP inside L cones')\n",
    "for axH in (axH1,axH2,axH3):\n",
    "    formatFigureMain(fH, axH, _)\n",
    "    axH.set_xlabel('Normalized GFP values')\n",
    "    if pdfV:\n",
    "        axH.set_ylabel('Probability density')\n",
    "    else:\n",
    "        axH.set_ylabel('Counts')\n",
    "## normalized\n",
    "# axH.stairs(wtH/np.max(wtH),wtB, fill=True, color = 'k', alpha = 0.5)\n",
    "# axH.stairs(aH/np.max(aH),aB, fill=True, color = 'r', alpha = 0.5)\n",
    "# axH.stairs(bH/np.max(bH),bB, fill=True, color = 'b', alpha = 0.5)\n",
    "\n",
    "dPath = '/Users/angueyraaristjm/Documents/LiLab/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Presentations/revealjs/resources/20211008_UCLA/'\n",
    "plt.savefig(dPath + \"CRtbx2F0_LM_hist.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "print('# cells of bright GFP cells: wt={0}, tbx2a={1}, tbx2b={2}'.format(wGBright*1000,aGBright*1000,bGBright*1000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# fH, (axH1,axH2,axH3) = plt.subplots(nrows=3, ncols=1, figsize=(10, 20))\n",
    "fH, axH = plt.subplots(nrows=1, ncols=1, figsize=(9, 6))\n",
    "\n",
    "\n",
    "# plot histogram\n",
    "pdfV = True\n",
    "# wtH, wtB = np.histogram(wtResults, bins = 50, density=pdfV)\n",
    "# aH, aB = np.histogram(aResults, bins = 50, density=pdfV)\n",
    "# bH, bB = np.histogram(bResults, bins = 50, density=pdfV)\n",
    "\n",
    "# plot histogram but first calculate a fixed bin number for all genotypes\n",
    "# allBins = np.histogram_bin_edges(np.concatenate([wtResults,aResults,bResults]).ravel(),bins='fd');\n",
    "allBins = np.arange(0,1010,10);\n",
    "allBinsC = (allBins[:-1] + allBins[1:]) / 2 # Centers\n",
    "wtH, wtB = np.histogram(wtResults, bins = allBins, density=pdfV)\n",
    "aH, aB = np.histogram(aResults, bins = allBins, density=pdfV)\n",
    "bH, bB = np.histogram(bResults, bins = allBins, density=pdfV)\n",
    "\n",
    "# # this can be used for cumulative distribution then normalized\n",
    "# wtH = np.cumsum(wtH)/np.max(np.cumsum(wtH));\n",
    "# aH = np.cumsum(aH)/np.max(np.cumsum(aH));\n",
    "# bH = np.cumsum(bH)/np.max(np.cumsum(bH));\n",
    "\n",
    "wt95CI = np.min(allBinsC[(np.cumsum(wtH))>.095])/1000;\n",
    "wGBright = np.sum(wtH[allBinsC>wt95CI*1000]);\n",
    "aGBright = np.sum(aH[allBinsC>wt95CI*1000]);\n",
    "bGBright = np.sum(bH[allBinsC>wt95CI*1000]);\n",
    "\n",
    "\n",
    "axH.stairs(wtH,wtB/1000, fill=True, color = 'k', alpha = .5)\n",
    "axH.stairs(aH,aB/1000, fill=True, color = zfC['lws1'], alpha = .5)\n",
    "axH.stairs(bH,bB/1000, fill=True, color = zfC['sws2'], alpha = .5)\n",
    "\n",
    "\n",
    "axH.plot([wt95CI,wt95CI],[0,0.016],color=baseColor)\n",
    "\n",
    "formatFigureMain(fH, axH, _)\n",
    "# axH.set_title('Distribution of GFP inside L cones')\n",
    "axH.set_xlabel('Normalized GFP values')\n",
    "if pdfV:\n",
    "    axH.set_ylabel('Probability density')\n",
    "else:\n",
    "    axH.set_ylabel('Counts')\n",
    "## normalized\n",
    "# axH.stairs(wtH/np.max(wtH),wtB, fill=True, color = 'k', alpha = 0.5)\n",
    "# axH.stairs(aH/np.max(aH),aB, fill=True, color = 'r', alpha = 0.5)\n",
    "# axH.stairs(bH/np.max(bH),bB, fill=True, color = 'b', alpha = 0.5)\n",
    "\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "dPath = '/Users/angueyraaristjm/Documents/LiLab/Conferences/2021_RCS/resources/'\n",
    "# plt.savefig(dPath + \"CRtbx2F0_LM_hist.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "print('# cells of bright GFP cells: wt={0}, tbx2a={1}, tbx2b={2}'.format(wGBright*1000,aGBright*1000,bGBright*1000))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### After establishing threshold based on all data, count number of cells exceeding threshold for each eye"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect all results\n",
    "\n",
    "# analysis directory\n",
    "dAnalysis = \"/Users/angueyraaristjm/Documents/LiImaging/Analysis/CRtbx2F0s/m2Gt2T/\"\n",
    "\n",
    "# open file\n",
    "\n",
    "# uninjected\n",
    "### 20201009: imaged 3dpf and 5dpf larvae. Analyzing 5 dpf for paper\n",
    "gene = 'wt'\n",
    "wt_fPaths = [\"20201009_5dpf_Rh22G_trb2T_wt_i09b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i09d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i10d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i11d\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12b\",\n",
    "          \"20201009_5dpf_Rh22G_trb2T_wt_i12d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L101d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L102b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L103d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L104d\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L105b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106b\",\n",
    "          \"20210716_5dpf_m2G_t2T_uninj_L106d\",\n",
    "         ]\n",
    "# tbx2a\n",
    "gene = 'tbx2a';\n",
    "tbx2a_fPaths = [\"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",\n",
    "#                 \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i07b\",  # is this supposed to be d????\n",
    "                # #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08b\", #very curved stack. would require oblique cut\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i08d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09b\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i09d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10b\", #done; this could be example\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i10d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11b\", #done; very clean L-cone seg\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i11d\",\n",
    "                \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12b\",\n",
    "                #### fPath = \"20201009_5dpf_Rh22G_trb2T_CRtbx2a_i12d\", # this in lens up\n",
    "               ]\n",
    "# tbx2b\n",
    "gene = 'tbx2b';\n",
    "tbx2b_fPaths = [\"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i07d\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08b\",\n",
    "                \"20201030_5dpf_Rh22G_trb2T_CRtbx2b_i08d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L102d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L104d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L105d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L106b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L107b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L108d\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L109b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L110b\",\n",
    "                \"20210716_5dpf_m2G_t2T_tbx2b_L111b\",\n",
    "               ]\n",
    "                \n",
    "rName = 'lGtile'\n",
    "\n",
    "wtResults = []\n",
    "i=1;\n",
    "for fPath in wt_fPaths:\n",
    "    gene = 'wt'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     wtResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    wtResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "    \n",
    "aResults = []\n",
    "i=1;\n",
    "for fPath in tbx2a_fPaths:\n",
    "    gene = 'tbx2a'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     aResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    aResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "bResults = []\n",
    "i=1;\n",
    "for fPath in tbx2b_fPaths:\n",
    "    gene = 'tbx2b'\n",
    "    dOut = dAnalysis + gene + '/' + fPath + '/'\n",
    "    results = np.load(dOut + 'quantificationGFP.npz')\n",
    "#     bResults.append(np.count_nonzero(results['lGtile']>wt95CI*1000)) #this is for absolute counts\n",
    "    bResults.append(np.divide(np.count_nonzero(results['lGtile']>wt95CI*1000),len(results['lGtile']))) #this is for fraction\n",
    "    i=i+1;\n",
    "\n",
    "geneList = ['wt','tbx2a','tbx2b']\n",
    "\n",
    "\n",
    "plotname = ''\n",
    "fH, axH = plt.subplots(figsize= [6,8])\n",
    "\n",
    "barW = np.size(geneList)+2; # bar width\n",
    "barD = np.size(geneList)+1; # bar distance whitin 1 photoreceptor subtype\n",
    "barP = -np.size(geneList)+1; # position in x-axis\n",
    "barStep = 0.9; # distance between photoreceptor groups\n",
    "barPos = []; #array to save bar positions\n",
    "\n",
    "\n",
    "jitter = estimateJitter(wtResults)/400\n",
    "pH = plt.bar([barStep+1/barD], np.mean(wtResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+1/barD], np.mean(wtResults), yerr = [[0.025],[np.std(wtResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(wtResults))*[barStep+1/barD]+jitter, wtResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['wt'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+1/barD))\n",
    "\n",
    "\n",
    "\n",
    "jitter = estimateJitter(aResults)/400\n",
    "pH = plt.bar([barStep+2/barD], np.mean(aResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+2/barD], np.mean(aResults), yerr = [[0.025],[np.std(aResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(aResults))*[barStep+2/barD]+jitter, aResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['tbx2a'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+2/barD))\n",
    "\n",
    "jitter = estimateJitter(bResults)/400\n",
    "pH = plt.bar([barStep+3/barD], np.mean(bResults), width=1/barW, color=zfC['lws1']);\n",
    "pH = plt.errorbar([barStep+3/barD], np.mean(bResults), yerr = [[0.025],[np.std(bResults)]], ecolor=lighten_color(zfC['lws1'],1), elinewidth=3, capsize=15, capthick=3, zorder=7)\n",
    "pH = plt.scatter(np.ones(len(bResults))*[barStep+3/barD]+jitter, bResults, color=lighten_color(zfC['lws1'],.75), zorder=8, marker = zfGm['tbx2b'], s=100, edgecolor='#000000', linewidth=0.5, alpha = .6);\n",
    "barPos = np.append(barPos,(barStep+3/barD))\n",
    "\n",
    "formatFigureMain(fH, axH, pH)\n",
    "\n",
    "# axH.set_ylabel('GFP+ L cones per 64 x 64 $\\mu$m$^2$', fontproperties=formatFigureMain(fH, axH, pH))\n",
    "# axH.set_ylim([0,180]);\n",
    "# stath = 170\n",
    "\n",
    "axH.set_ylabel('Fraction of GFP+ L cones', fontproperties=formatFigureMain(fH, axH, pH))\n",
    "axH.set_ylim([0,1]);\n",
    "stath = 0.85\n",
    "\n",
    "geneList = ['control','F$\\emptyset$[tbx2a]','F$\\emptyset$[tbx2b]']\n",
    "\n",
    "axH.xaxis.set_tick_params(rotation=45)\n",
    "axH.set_xticks(np.sort(barPos));\n",
    "axH.set_xticklabels(geneList);\n",
    "\n",
    "hH = plt.hlines(stath, barPos[0], barPos[1], colors='#000000',linewidth=4)\n",
    "plt.text((barPos[0]+barPos[1])/2, stath, '*', font_properties=font_prop, fontsize=36, ha='center')# plt.savefig(savePath + \"Counts_tbx2_RU.svg\", transparent=True, format=\"svg\")\n",
    "\n",
    "\n",
    "dPath = '/Users/angueyraaristjm/OneDrive - National Institutes of Health/zf/Manuscripts/2021_zfConeRNAseq/resources/'\n",
    "# dPath = '/Users/angueyraaristjm/Documents/LiLab/Presentations/revealjs/resources/20211008_UCLA/'\n",
    "plt.savefig(dPath + \"CRtbx2F0_GFPLcones.svg\", transparent=True, format=\"svg\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('wt fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(wtResults)*100,np.std(wtResults)*100))\n",
    "print('tbx2a fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(aResults)*100,np.std(aResults)*100))\n",
    "print('tbx2b fractions:')\n",
    "print('\\tmean = {0:.3f}%, s.d. = {1:.5f}%'.format(np.mean(bResults)*100,np.std(bResults)*100))\n",
    "# Mann Whitney \n",
    "Au, pA = mannwhitneyu(wtResults, aResults)\n",
    "print('wt vs. tbx2a:')\n",
    "print('\\tU = {0:.3f}, p = {1:.5f}'.format(Au,pA))\n",
    "Bu, pB = mannwhitneyu(wtResults, bResults)\n",
    "print('wt vs. tbx2b:')\n",
    "print('\\tU = {0:.3f}, p = {1:.5f}'.format(Bu,pB))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# divide images into 1000 bins\n",
    "mH, mB = np.histogram(viewer.layers['M_mip'].data.flatten(), bins = 1000)\n",
    "lH, lB = np.histogram(viewer.layers['L_mip'].data.flatten(), bins = 1000)\n",
    "\n",
    "\n",
    "fH, axH = plt.subplots(figsize=(8,8))\n",
    "axH.stairs(lH,lB, fill=True, color = 'm', alpha = 0.5)\n",
    "axH.stairs(mH,mB, fill=True, color = 'g', alpha = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "# change keybind for default actions:\n",
    "@viewer.bind_key('F1', overwrite=True)\n",
    "def eraserMode(viewer):\n",
    "    viewer.layers.selection.active.mode = 'fill'\n",
    "    viewer.layers.selection.active.mode = 'paint'\n",
    "    viewer.layers.selection.active.mode = 'erase'\n",
    "    viewer.layers.selection.active.mode = 'pick'\n",
    "    viewer.layers.selection.active.mode = 'pan_zoom'\n",
    "\n",
    "# napari native functions for other actions\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\n",
    "        \"Set the currently selected label to the largest used label plus one.\"\n",
    "    ),\n",
    ")\n",
    "def new_label(layer):\n",
    "    \"\"\"Set the currently selected label to the largest used label plus one.\"\"\"\n",
    "    layer.selected_label = layer.data.max() + 1\n",
    "\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\"Decrease the currently selected label by one.\"),\n",
    ")\n",
    "def decrease_label_id(layer):\n",
    "    layer.selected_label -= 1\n",
    "\n",
    "\n",
    "@register_label_action(\n",
    "    trans._(\"Increase the currently selected label by one.\"),\n",
    ")\n",
    "def increase_label_id(layer):\n",
    "    layer.selected_label += 1\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
