{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup for PCR for genotyping, fragment analysis, sequencing, etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Phusion PCR setup:\n",
    "    DNA (1:10) dilution = 2 uL\n",
    "    Water = 10.4 uL\n",
    "    F primer (10 uL) = 1 uL\n",
    "    R primer (10 uL) = 1 uL\n",
    "    M13-FAM primer (10 uM) = 1 uL\n",
    "    Phusion polymerase = 0.2 uL\n",
    "    5x buffer = 4 uL\n",
    "    dNTP mix = 0.4 uL\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED: 13 uL reaction for fragment analysis only\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class gene:\n",
    "    def __init__(self, gene_name, id_numbers):\n",
    "        self.name = gene_name\n",
    "        self.ids = id_numbers\n",
    "        self.n = len(self.ids.split(','))\n",
    "        self.string = self.name + ' (n='+ str(self.n) + ')'\n",
    "        \n",
    "        tested_Tms = {'tjp2a': 72, #inconsistent results here. Should order new primers. On 2017/06/07, bands present with 72C.\n",
    "                      'tjp2b': 64 ,\n",
    "                      'cadm1a': 64, #tested 67C on 2017/03/30 and bands disappeared\n",
    "                      'cadm1b': 67,\n",
    "                      'eml1': 67,\n",
    "                      'cpne3': 67,\n",
    "                      'nrxn3aS': 72,\n",
    "                      'nrxn3aL': 72,\n",
    "                      'nrxn3bS': 72,\n",
    "                      'nrxn3bL': 72,\n",
    "                      'eml1⼆': 72,\n",
    "                      'cpne3⼆': 72,\n",
    "                      'cpne3_sa9988': 67,\n",
    "                      'eml1_sa14062': 66,\n",
    "                      'nrxn3a_sa11330': 67,\n",
    "                      'nrxn3b_sa36960': 66,\n",
    "#                       'syt5a': 00, #F and R primers that don't work NEB Tm = 61; 2020/02/03: tested 59 & 67: no bands; 62: extremely faint band\n",
    "                      'syt5a': 67, #Fii and Rii New primers: NEB Tm = 62; 2020/02/03: tested 59 & 67: no bands; 62: extremely faint band\n",
    "                      'syt5b': 70, # NEB Tm_F = 74, Tm_R = 67\n",
    "                      'gnat2': 67, #NEB Tm = 64; 2020/02/03: tested 62 and 67: both work;\n",
    "                      'sema7a': 67, #NEB Tm = 62; 2020/02/03: tested 62 and 67: both work;\n",
    "                      'efna1b': 70, #NEB Tm = 63; 2020/02/03: tested 59, 62 and 67: 3 extra bands at lower weight; tested 69: much cleaner\n",
    "                      'tbx2a': 67, #NEB Tm = 59; Fi Ri primers\n",
    "                      'tbx2a_FiRii': 69, # NEB Tm_F = 73, Tm_R = 68\n",
    "                      'tbx2a_FiRiii': 70, # NEB Tm_F = 73, Tm_R = 68\n",
    "                      'tbx2a_FiiRi': 67, # NEB Tm_F = 73, Tm_R = 68\n",
    "                      'tbx2a_FiiRii': 69, # NEB Tm_F = 73, Tm_R = 68\n",
    "                      'tbx2a_FiiRiii': 70, # NEB Tm_F = 73, Tm_R = 68\n",
    "                      'tbx2b': 67, #NEB Tm = 61; \n",
    "                      'ntng2b': 67, #NEB Tm = 61; \n",
    "                      'foxq2' : 64, # tested and it works\n",
    "                      'skor1a': 67, # this is for F2R2 or F1R2 (R1 perimer is bad); tested by Laura Dec_2021\n",
    "                      'nr2f1b': 70, # NEB Tm_F = 73, Tm_R = 67 \n",
    "                      'xbp1': 67, # NEB Tm_F = 73, Tm_R = 68 tested by Laura Dec_2021\n",
    "                      'lrrfip1a': 67, # tested by Laura Dec_2021\n",
    "                      'sall1a': 67, # NEB Tm_F = 74, Tm_R = 67\n",
    "                      'tgif': 67, # NEB Tm_F = 76, Tm_R = 66\n",
    "                      'tefa': 67, # NEB Tm_F = 75, Tm_R = 67\n",
    "                      'tefb': 67, # NEB Tm_F = 74, Tm_R = 67\n",
    "                      'lhx1a': 67, # NEB Tm_F = 74, Tm_R = 65\n",
    "                      'gal4': 64, #untested, without PIG and M13 it's 56\n",
    "                      'gfp': 72, #untested, without PIG and M13 it's 70\n",
    "                      'arr3a': 67, #untested, without PIG and M13 it's 60\n",
    "                      'arr3b': 67, #untested, without PIG and M13 it's 60\n",
    "                     }\n",
    "        self.Tm = tested_Tms.get(self.name, 'default')\n",
    "        \n",
    "        tested_seqPrimer = {\n",
    "                        'tjp2a': 'none', #inconsistent results here. Should order new primers. On 2017/06/07, bands present with 72C.\n",
    "                        'tjp2b': 'none' ,\n",
    "                        'cadm1a': 'F',\n",
    "                        'cadm1b': 'none',\n",
    "                        'eml1': 'none',\n",
    "                        'cpne3': 'none',\n",
    "                        'nrxn3aS': 'none',\n",
    "                        'nrxn3aL': 'none',\n",
    "                        'nrxn3bS': 'none',\n",
    "                        'nrxn3bL': 'none',\n",
    "                        'eml1⼆': 'none',\n",
    "                        'cpne3⼆': 'none',\n",
    "                        'cpne3_sa9988': 'none',\n",
    "                        'eml1_sa14062': 'none',\n",
    "                        'nrxn3a_sa11330': 'R',\n",
    "                        'nrxn3b_sa36960': 'none',\n",
    "                        'syt5a':  'idk', # trying Fii (200/07/31)\n",
    "                        'syt5b': 'idk', # \n",
    "                        'gnat2': 'idk',\n",
    "                        'sema7a': 'idk', # trying F (200/07/31)\n",
    "                        'efna1b': 'idk',\n",
    "                        'tbx2a': 'F',\n",
    "                        'tbx2b': 'R',\n",
    "                        'ntng2': 'idk',\n",
    "                        'foxq2': 'R', # tested on July2021\n",
    "                        'skor1a': 'idk', # \n",
    "                        'nr2f1b': 'idk', # \n",
    "                        'xbp1': 'idk', # \n",
    "                        'lrrfip1a': 'idk', # \n",
    "                        'sall1a': 'idk', # \n",
    "                        'tgif': 'idk', # \n",
    "                        'tefa': 'idk', # \n",
    "                        'tefb': 'idk', # \n",
    "                        'lhx1a': 'idk', #\n",
    "                        'gal4': 'idk', #\n",
    "                        'gfp': 'idk', #\n",
    "                        'arr3a': 'idk', #\n",
    "                        'arr3b': 'idk', #\n",
    "                     }\n",
    "        self.seqPrimer = tested_seqPrimer.get(self.name, 'default')\n",
    "        \n",
    "        expectedSize = {\n",
    "                        'tjp2a': 'idk', #M13 and PIG seems to add 20bp in fragment analysis, so these values are PCR size + 20bp or value from fragment analysis\n",
    "                        'tjp2b': 'idk' ,\n",
    "                        'cadm1a': 'idk',\n",
    "                        'cadm1b': 'idk',\n",
    "                        'eml1': 'idk',\n",
    "                        'cpne3': 'idk',\n",
    "                        'nrxn3aS': 'idk',\n",
    "                        'nrxn3aL': 'idk',\n",
    "                        'nrxn3bS': 'idk',\n",
    "                        'nrxn3bL': 'idk',\n",
    "                        'eml1⼆': 'idk',\n",
    "                        'cpne3⼆': 'idk',\n",
    "                        'cpne3_sa9988': 'idk',\n",
    "                        'eml1_sa14062': 'idk',\n",
    "                        'nrxn3a_sa11330': 'idk',\n",
    "                        'nrxn3b_sa36960': 'idk',\n",
    "                        'syt5a':  'idk', \n",
    "                        'syt5b': 343,  \n",
    "                        'gnat2': 'idk',\n",
    "                        'sema7a': 358,\n",
    "                        'efna1b': 'idk',\n",
    "                        'tbx2a': 488,\n",
    "                        'tbx2a_FiRii': 532,\n",
    "                        'tbx2a_FiRiii': 578,\n",
    "                        'tbx2a_FiiRi': 402,\n",
    "                        'tbx2a_FiiRii': 452,\n",
    "                        'tbx2a_FiiRiii': 498,\n",
    "                        'tbx2b': 332,\n",
    "                        'ntng2': 'idk',\n",
    "                        'foxq2': 426, \n",
    "                        'skor1a': 349, #assuming F2R2\n",
    "                        'nr2f1b': 412,\n",
    "                        'xbp1': 289,\n",
    "                        'lrrfip1a': 281,\n",
    "                        'sall1a': 179, # \n",
    "                        'tgif': 270, # \n",
    "                        'tefa': 322, # \n",
    "                        'tefb': 354, # \n",
    "                        'lhx1a': 375, #\n",
    "                        'gal4': 204, #\n",
    "                        'gfp': 204, #\n",
    "                        'arr3a': 283, #\n",
    "                        'arr3b': 465, #\n",
    "                     }\n",
    "        self.expectedSize = expectedSize.get(self.name, 'default')\n",
    "\n",
    "def array2list(thearray):\n",
    "    d=(\",\".join(map(str,thearray)))\n",
    "    return d\n",
    "# NORMAL Phusion\n",
    "# volumes  = [10.4, 4, 0.4, 1, 1, 1, 0.2];\n",
    "# names = ['Water','Buffer5x','dNTP mix','FWD','REV','M13-FAM','Phusion'];\n",
    "# MASTER MIX Phusion\n",
    "# volumes  = [5, 1, 1, 1, 10];\n",
    "# MASTER MIX Phusion using Kate Pinter's ratios (May, 2021)\n",
    "volumes  = [6.42, 0.316, 0.632, 0.632, 10];\n",
    "volumes = np.append(volumes,np.sum(volumes)); alertString = \"20 uL reaction for default PCR reactions\";\n",
    "volumes=np.ndarray.tolist(np.multiply(volumes,2/3));  alertString = \"13 uL reaction for fragment analysis only\";# for Fo-screen and no sequencing plans\n",
    "names = ['Water','FWD','REV','M13-FAM','PhusionMix2x','Total'];\n",
    "PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "\n",
    "### test PCR (half-volume reactions, no M13-FAM)\n",
    "### NORMAL Phusion half-volume\n",
    "# volumes  = [11.4, 4, 0.4, 1, 1, 0.2];\n",
    "# volumes=np.ndarray.tolist(np.divide(volumes,2))\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','Buffer5x','dNTP mix','FWD','REV','Phusion','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "# OR\n",
    "### MASTER MIX Phusion three quarters-volume no M13-FAM\n",
    "# volumes  = [6, 1, 1, 10];\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,3/4))\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','FWD','REV','PhusionMix2x','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "\n",
    "# ### For sequencing\n",
    "# volumes  = [6, 1, 1, 10];\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,1/2)); alertString = \"10 uL reaction for sequencing only\";\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','FWD','REV','PhusionMix2x','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "\n",
    "print('LOADED: ' + alertString)\n",
    "# volumes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>-</th>\n",
       "      <th>1x</th>\n",
       "      <th>tbx2a (n=24)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Water</th>\n",
       "      <td>4.28</td>\n",
       "      <td>112.99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FWD</th>\n",
       "      <td>0.21</td>\n",
       "      <td>5.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REV</th>\n",
       "      <td>0.42</td>\n",
       "      <td>11.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M13-FAM</th>\n",
       "      <td>0.42</td>\n",
       "      <td>11.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PhusionMix2x</th>\n",
       "      <td>6.67</td>\n",
       "      <td>176.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Total</th>\n",
       "      <td>12.00</td>\n",
       "      <td>316.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "-                1x  tbx2a (n=24)\n",
       "Water          4.28        112.99\n",
       "FWD            0.21          5.56\n",
       "REV            0.42         11.12\n",
       "M13-FAM        0.42         11.12\n",
       "PhusionMix2x   6.67        176.00\n",
       "Total         12.00        316.80"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "#######\n",
    "genes=[];\n",
    "\n",
    "\n",
    "genes.append(gene('tbx2a',(array2list(np.arange(0,8*3)+1))));\n",
    "# genes.append(gene('tbx2b',(array2list(np.arange(0,8*10)+1))));\n",
    "# genes.append(gene('tgif',(array2list(np.arange(0,8)+1))));\n",
    "# genes.append(gene('tefa',(array2list(np.arange(0,8)+1))));\n",
    "# genes.append(gene('tefb',(array2list(np.arange(0,8)+1))));\n",
    "# genes.append(gene('lhx1a',(array2list(np.arange(0,8)+1))));\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# genes.append(gene('eml1ii',(array2list(np.arange(0,12,1)))));\n",
    "\n",
    "PCR = pd.DataFrame.copy(PhusionDefault)\n",
    "for g in genes:\n",
    "    PCR[g.string] = round(pd.Series(PCR['1x'] * (g.n*1*1.1)),2)\n",
    "PCR.columns.name = '-'\n",
    "PCR.round(decimals=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>id#</th>\n",
       "      <th>Tm (°C)</th>\n",
       "      <th>seqPrimer</th>\n",
       "      <th>expectedSize (bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>tbx2a</th>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>67</td>\n",
       "      <td>F</td>\n",
       "      <td>488</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene                                                 id#  Tm (°C) seqPrimer  \\\n",
       "tbx2a  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...       67         F   \n",
       "\n",
       "gene   expectedSize (bp)  \n",
       "tbx2a                488  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [];\n",
    "ids = [];\n",
    "tms = [];\n",
    "seqPrimer =[];\n",
    "expectedSize =[];\n",
    "for g in genes:\n",
    "    names.append(g.name)\n",
    "    ids.append(g.ids)\n",
    "    tms.append(g.Tm)\n",
    "    seqPrimer.append(g.seqPrimer)\n",
    "    expectedSize.append(g.expectedSize)\n",
    "\n",
    "IDs = pd.DataFrame({'id#': ids, 'Tm (°C)' : tms, 'seqPrimer' : seqPrimer, 'expectedSize (bp)' : expectedSize}, index=names)\n",
    "IDs.columns.name = 'gene'\n",
    "#IDs[['id#','Tm (°C)']]\n",
    "IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|              |    1x |   tbx2a (n=24) |\n",
      "|:-------------|------:|---------------:|\n",
      "| Water        |  4.28 |         112.99 |\n",
      "| FWD          |  0.21 |           5.56 |\n",
      "| REV          |  0.42 |          11.12 |\n",
      "| M13-FAM      |  0.42 |          11.12 |\n",
      "| PhusionMix2x |  6.67 |         176    |\n",
      "|    |\n",
      "| **Total**        | 12    |         316.8  |\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "t = PCR.round(decimals=2).to_markdown()\n",
    "t = t.replace('Total','   |\\n| **Total**')\n",
    "print(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|       | id#                                                            |   Tm (°C) | seqPrimer   |   expectedSize (bp) |\n",
      "|:------|:---------------------------------------------------------------|----------:|:------------|--------------------:|\n",
      "| tbx2a | 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24 |        67 | F           |                 488 |\n"
     ]
    }
   ],
   "source": [
    "print(IDs.to_markdown())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "64/4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
