{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DNA lysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Phusion PCR setup:\n",
    "    DNA (1:10) dilution = 2 uL\n",
    "    Water = 10.4 uL\n",
    "    F primer (10 uL) = 1 uL\n",
    "    R primer (10 uL) = 1 uL\n",
    "    M13-FAM primer (10 uM) = 1 uL\n",
    "    Phusion polymerase = 0.2 uL\n",
    "    5x buffer = 4 uL\n",
    "    dNTP mix = 0.4 uL\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "class gene:\n",
    "    def __init__(self, gene_name, id_numbers):\n",
    "        self.name = gene_name\n",
    "        self.ids = id_numbers\n",
    "        self.n = len(self.ids.split(','))\n",
    "        self.string = self.name + ' (n='+ str(self.n) + ')'\n",
    "        \n",
    "        tested_Tms = {'tjp2a': 72, #inconsistent results here. Should order new primers. On 2017/06/07, bands present with 72C.\n",
    "                      'tjp2b': 64 ,\n",
    "                      'cadm1a': 64, #tested 67C on 2017/03/30 and bands disappeared\n",
    "                      'cadm1b': 67,\n",
    "                      'eml1': 67,\n",
    "                      'cpne3': 67,\n",
    "                      'nrxn3aS': 72,\n",
    "                      'nrxn3aL': 72,\n",
    "                      'nrxn3bS': 72,\n",
    "                      'nrxn3bL': 72,\n",
    "                      'eml1⼆': 72,\n",
    "                      'cpne3⼆': 72,\n",
    "                      'cpne3_sa9988': 67,\n",
    "                      'eml1_sa14062': 66,\n",
    "                      'nrxn3a_sa11330': 67,\n",
    "                      'nrxn3b_sa36960': 66,\n",
    "#                       'syt5a': 00, #F and R primers that don't work NEB Tm = 61; 2020/02/03: tested 59 & 67: no bands; 62: extremely faint band\n",
    "                      'syt5a': 62, #Fii and Rii New primers: NEB Tm = 62; 2020/02/03: tested 59 & 67: no bands; 62: extremely faint band\n",
    "                      'gnat2': 67, #NEB Tm = 64; 2020/02/03: tested 62 and 67: both work;\n",
    "                      'sema7a': 67, #NEB Tm = 62; 2020/02/03: tested 62 and 67: both work;\n",
    "                      'efna1b': 70, #NEB Tm = 63; 2020/02/03: tested 59, 62 and 67: 3 extra bands at lower weight; tested 69: much cleaner\n",
    "                      'tbx2a': 62, #NEB Tm = 59; 2020/06/26: tested 62 and 67: only 62 has good bands; 67 had weak bands in some samples, but not all;\n",
    "                      'tbx2b': 62, #NEB Tm = 61; 2020/06/26: tested 62 and 67: only 62 has good bands; 67 had weak bands in some samples, but not all;\n",
    "                      'ntng2': 62, #NEB Tm = 61; 2020/06/26: tested 62 and 67: only 62 has good bands; 67 had weak bands in some samples, but not all;\n",
    "                     }\n",
    "        self.Tm = tested_Tms.get(self.name, 'default')\n",
    "        \n",
    "        tested_seqPrimer = {'tjp2a': 'none', #inconsistent results here. Should order new primers. On 2017/06/07, bands present with 72C.\n",
    "                      'tjp2b': 'none' ,\n",
    "                      'cadm1a': 'F',\n",
    "                      'cadm1b': 'none',\n",
    "                      'eml1': 'none',\n",
    "                      'cpne3': 'none',\n",
    "                      'nrxn3aS': 'none',\n",
    "                      'nrxn3aL': 'none',\n",
    "                      'nrxn3bS': 'none',\n",
    "                      'nrxn3bL': 'none',\n",
    "                      'eml1⼆': 'none',\n",
    "                      'cpne3⼆': 'none',\n",
    "                      'cpne3_sa9988': 'none',\n",
    "                      'eml1_sa14062': 'none',\n",
    "                      'nrxn3a_sa11330': 'R',\n",
    "                      'nrxn3b_sa36960': 'none',\n",
    "                      'syt5a':  'idk',\n",
    "                      'gnat2': 'idk',\n",
    "                      'sema7a': 'idk',\n",
    "                      'efna1b': 'idk',\n",
    "                      'tbx2a': 'idk',\n",
    "                      'tbx2b': 'idk',\n",
    "                      'ntng2': 'idk',\n",
    "                     }\n",
    "        self.seqPrimer = tested_seqPrimer.get(self.name, 'default')\n",
    "\n",
    "def array2list(thearray):\n",
    "    d=(\",\".join(map(str,thearray)))\n",
    "    return d\n",
    "# NORMAL Phusion\n",
    "# volumes  = [10.4, 4, 0.4, 1, 1, 1, 0.2];\n",
    "# names = ['Water','Buffer5x','dNTP mix','FWD','REV','M13-FAM','Phusion'];\n",
    "# MASTER MIX Phusion\n",
    "volumes  = [5, 1, 1, 1, 10];\n",
    "volumes = np.append(volumes,np.sum(volumes));\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,1/2)); # for Fo-screen and no sequencing plans\n",
    "names = ['Water','FWD','REV','M13-FAM','PhusionMix2x','Total'];\n",
    "PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "\n",
    "### test PCR (half-volume reactions, no M13-FAM)\n",
    "### NORMAL Phusion half-volume\n",
    "# volumes  = [11.4, 4, 0.4, 1, 1, 0.2];\n",
    "# volumes=np.ndarray.tolist(np.divide(volumes,2))\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','Buffer5x','dNTP mix','FWD','REV','Phusion','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "# OR\n",
    "### MASTER MIX Phusion three quarters-volume no M13-FAM\n",
    "# volumes  = [6, 1, 1, 10];\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,3/4))\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','FWD','REV','PhusionMix2x','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes=[];\n",
    "# genes.append(gene('tjp2a',('2Si_1,2Si_2,2Sii_1,2Sii_2,21,22')));\n",
    "# genes.append(gene('tjp2b',('21,22')));\n",
    "# genes.append(gene('cadm1a',('1a,1b,1c,1d,1e,1f,2a,2b,2c,2d,2e,2f')));\n",
    "# genes.append(gene('cadm1a',('2')));\n",
    "# genes.append(gene('cadm1b',('19')));\n",
    "# genes.append(gene('cpne3⼆',('1a,1b,2')));\n",
    "# genes.append(gene('nrxn3a',('1,2')));\n",
    "# genes.append(gene('nrxn3b',('1,2')));\n",
    "# genes.append(gene('eml1⼆',('1a,1b')));\n",
    "# genes.append(gene('cpne3⼆',('9a,9b,10a,10b,11a,11b,12,13a,13b')));\n",
    "genes.append(gene('gnat2',('1,2,3,4,5,6')));\n",
    "genes.append(gene('sema7a',('1,2')));\n",
    "genes.append(gene('eml1⼆',('1,2,3,4')));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>-</th>\n",
       "      <th>1x</th>\n",
       "      <th>sema7a (n=16)</th>\n",
       "      <th>gnat2 (n=8)</th>\n",
       "      <th>syt5a (n=32)</th>\n",
       "      <th>efna1b (n=24)</th>\n",
       "      <th>syt5a (n=16)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Water</td>\n",
       "      <td>2.5</td>\n",
       "      <td>44.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>44.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FWD</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8.8</td>\n",
       "      <td>4.4</td>\n",
       "      <td>17.6</td>\n",
       "      <td>13.2</td>\n",
       "      <td>8.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REV</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8.8</td>\n",
       "      <td>4.4</td>\n",
       "      <td>17.6</td>\n",
       "      <td>13.2</td>\n",
       "      <td>8.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>M13-FAM</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8.8</td>\n",
       "      <td>4.4</td>\n",
       "      <td>17.6</td>\n",
       "      <td>13.2</td>\n",
       "      <td>8.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PhusionMix2x</td>\n",
       "      <td>5.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>132.0</td>\n",
       "      <td>88.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Total</td>\n",
       "      <td>9.0</td>\n",
       "      <td>158.4</td>\n",
       "      <td>79.2</td>\n",
       "      <td>316.8</td>\n",
       "      <td>237.6</td>\n",
       "      <td>158.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "-              1x  sema7a (n=16)  gnat2 (n=8)  syt5a (n=32)  efna1b (n=24)  \\\n",
       "Water         2.5           44.0         22.0          88.0           66.0   \n",
       "FWD           0.5            8.8          4.4          17.6           13.2   \n",
       "REV           0.5            8.8          4.4          17.6           13.2   \n",
       "M13-FAM       0.5            8.8          4.4          17.6           13.2   \n",
       "PhusionMix2x  5.0           88.0         44.0         176.0          132.0   \n",
       "Total         9.0          158.4         79.2         316.8          237.6   \n",
       "\n",
       "-             syt5a (n=16)  \n",
       "Water                 44.0  \n",
       "FWD                    8.8  \n",
       "REV                    8.8  \n",
       "M13-FAM                8.8  \n",
       "PhusionMix2x          88.0  \n",
       "Total                158.4  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### For sequencing\n",
    "# volumes  = [6, 1, 1, 10];\n",
    "\n",
    "# ### MASTER MIX Phusion three quarters-volume no M13-FAM\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,3/4))\n",
    "\n",
    "### MASTER MIX Phusion half-volume no M13-FAM\n",
    "# volumes=np.ndarray.tolist(np.multiply(volumes,1/2))\n",
    "# volumes = np.append(volumes,np.sum(volumes));\n",
    "# names = ['Water','FWD','REV','PhusionMix2x','Total'];\n",
    "# PhusionDefault = pd.DataFrame(data=volumes, index=names, columns=['1x']);\n",
    "\n",
    "#######\n",
    "genes=[];\n",
    "# genes.append(gene('cpne3⼆',('1E,1H,1WT')));\n",
    "# genes.append(gene('cpne3_sa9988',(array2list(np.arange(1,25,1)))));\n",
    "# genes.append(gene('tjp2a',(array2list(np.arange(1,25,1)))));\n",
    "# genes.append(gene('eml1_sa14062',('1wt,2wt,3wt')));\n",
    "# genes.append(gene('eml1_FR',(array2list(np.arange(9,13,1)))));\n",
    "# genes.append(gene('eml1_FRii',(array2list(np.arange(9,13,1)))));\n",
    "# genes.append(gene('eml1_FiiR',(array2list(np.arange(9,13,1)))));\n",
    "# genes.append(gene('eml1_FiiRii',(array2list(np.arange(9,13,1)))));\n",
    "# genes.append(gene('tjp2a_FR',(array2list(np.arange(17,21,1)))));\n",
    "# genes.append(gene('tjp2a_FRii',(array2list(np.arange(17,21,1)))));\n",
    "# genes.append(gene('tjp2a_FiiRii',(array2list(np.arange(17,21,1)))));\n",
    "# genes.append(gene('NoDNA',(array2list(np.arange(0,1,1)))));\n",
    "# genes.append(gene('Nothing',(array2list(np.arange(0,1,1)))));\n",
    "# genes.append(gene('nrxn3a_sa11330',(array2list(np.arange(1,19,1)))));\n",
    "# genes.append(gene('eml1⼆',(array2list(np.arange(1,13,1)))));\n",
    "# genes.append(gene('nrxn3a_sa11330',(array2list(np.arange(0,40,1)))));\n",
    "# genes.append(gene('nrxn3b_sa36960',(array2list(np.arange(1,17,1)))));\n",
    "# genes.append(gene('cadm1a',(array2list(np.arange(1,43,1)))));\n",
    "# genes.append(gene('nrxn3a_sa11330',(array2list(np.arange(1,8,1)))));\n",
    "\n",
    "# genes.append(gene('nrxn3b_sa36960',('1wt,2wt,3wt')));\n",
    "# genes.append(gene('cadm1a',('16B,10A,10B,10C,10D,10E,10F,10G,10H,10I,10J,10K,10L,10M,10N,10O,10P,10Q,10R,10S,10T')));\n",
    "# genes.append(gene('tjp2a',('8D,8E,8F,8G,14G,14H,18F,18G,18H,18I,18wt,2A,2B,2C,2D,2E,2F,2G,2H,2I,2J,2K,2L,2M,2N,2O,2P,2Q,2R,2S,2T')));\n",
    "# genes.append(gene('tjp2b',('8I,18I,18wt')));\n",
    "# genes.append(gene('nrxn3aL',('7F,FG')));\n",
    "\n",
    "genes.append(gene('sema7a',array2list(np.arange(0,8*2,1))));\n",
    "genes.append(gene('gnat2',array2list(np.arange(0,8,1))));\n",
    "genes.append(gene('syt5a',array2list(np.arange(0,8*4,1))));\n",
    "genes.append(gene('efna1b',array2list(np.arange(0,8*3,1))));\n",
    "genes.append(gene('syt5a',array2list(np.arange(0,8*2,1))));\n",
    "\n",
    "\n",
    "PCR = pd.DataFrame.copy(PhusionDefault)\n",
    "for g in genes:\n",
    "    PCR[g.string] = round(pd.Series(PCR['1x'] * (g.n*1*1.1)),2)\n",
    "PCR.columns.name = '-'\n",
    "PCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "40"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "24+16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>-</th>\n",
       "      <th>1x</th>\n",
       "      <th>gnat2 (n=6)</th>\n",
       "      <th>sema7a (n=2)</th>\n",
       "      <th>eml1⼆ (n=4)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Water</td>\n",
       "      <td>5</td>\n",
       "      <td>264.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>176.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FWD</td>\n",
       "      <td>1</td>\n",
       "      <td>52.8</td>\n",
       "      <td>17.6</td>\n",
       "      <td>35.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>REV</td>\n",
       "      <td>1</td>\n",
       "      <td>52.8</td>\n",
       "      <td>17.6</td>\n",
       "      <td>35.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>M13-FAM</td>\n",
       "      <td>1</td>\n",
       "      <td>52.8</td>\n",
       "      <td>17.6</td>\n",
       "      <td>35.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>PhusionMix2x</td>\n",
       "      <td>10</td>\n",
       "      <td>528.0</td>\n",
       "      <td>176.0</td>\n",
       "      <td>352.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Total</td>\n",
       "      <td>18</td>\n",
       "      <td>950.4</td>\n",
       "      <td>316.8</td>\n",
       "      <td>633.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "-             1x  gnat2 (n=6)  sema7a (n=2)  eml1⼆ (n=4)\n",
       "Water          5        264.0          88.0        176.0\n",
       "FWD            1         52.8          17.6         35.2\n",
       "REV            1         52.8          17.6         35.2\n",
       "M13-FAM        1         52.8          17.6         35.2\n",
       "PhusionMix2x  10        528.0         176.0        352.0\n",
       "Total         18        950.4         316.8        633.6"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# PCR table for fragment analysis screen (8 larvae per founder)\n",
    "PCR = pd.DataFrame.copy(PhusionDefault)\n",
    "for g in genes:\n",
    "    PCR[g.string] = pd.Series(PCR['1x'] * round((g.n*8*1.1),1))\n",
    "PCR.columns.name = '-'\n",
    "PCR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene</th>\n",
       "      <th>id#</th>\n",
       "      <th>Tm (°C)</th>\n",
       "      <th>seqPrimer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>sema7a</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15</td>\n",
       "      <td>67</td>\n",
       "      <td>idk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>gnat2</td>\n",
       "      <td>0,1,2,3,4,5,6,7</td>\n",
       "      <td>67</td>\n",
       "      <td>idk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>syt5a</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...</td>\n",
       "      <td>62</td>\n",
       "      <td>idk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>efna1b</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...</td>\n",
       "      <td>70</td>\n",
       "      <td>idk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>syt5a</td>\n",
       "      <td>0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15</td>\n",
       "      <td>62</td>\n",
       "      <td>idk</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "gene                                                  id#  Tm (°C) seqPrimer\n",
       "sema7a              0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15       67       idk\n",
       "gnat2                                     0,1,2,3,4,5,6,7       67       idk\n",
       "syt5a   0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...       62       idk\n",
       "efna1b  0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18...       70       idk\n",
       "syt5a               0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15       62       idk"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "names = [];\n",
    "ids = [];\n",
    "tms = [];\n",
    "seqPrimer =[];\n",
    "for g in genes:\n",
    "    names.append(g.name)\n",
    "    ids.append(g.ids)\n",
    "    tms.append(g.Tm)\n",
    "    seqPrimer.append(g.seqPrimer)\n",
    "\n",
    "IDs = pd.DataFrame({'id#': ids, 'Tm (°C)' : tms, 'seqPrimer' : seqPrimer}, index=names)\n",
    "IDs.columns.name = 'gene'\n",
    "#IDs[['id#','Tm (°C)']]\n",
    "IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<table border=\"1\" class=\"dataframe\">\\n  <thead>\\n    <tr style=\"text-align: right;\">\\n      <th>-</th>\\n      <th>1x</th>\\n      <th>tjp2a\\n (n=1)</th>\\n      <th>tjp2b\\n (n=1)</th>\\n      <th>cadm1a\\n (n=2)</th>\\n      <th>cadm1b\\n (n=2)</th>\\n      <th>nrxn3aL\\n (n=1)</th>\\n      <th>nrxn3bL\\n (n=1)</th>\\n    </tr>\\n  </thead>\\n  <tbody>\\n    <tr>\\n      <th>Water</th>\\n      <td>5</td>\\n      <td>42.0</td>\\n      <td>42.0</td>\\n      <td>84.0</td>\\n      <td>84.0</td>\\n      <td>42.0</td>\\n      <td>42.0</td>\\n    </tr>\\n    <tr>\\n      <th>FWD</th>\\n      <td>1</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n      <td>16.8</td>\\n      <td>16.8</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n    </tr>\\n    <tr>\\n      <th>REV</th>\\n      <td>1</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n      <td>16.8</td>\\n      <td>16.8</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n    </tr>\\n    <tr>\\n      <th>M13-FAM</th>\\n      <td>1</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n      <td>16.8</td>\\n      <td>16.8</td>\\n      <td>8.4</td>\\n      <td>8.4</td>\\n    </tr>\\n    <tr>\\n      <th>PhusionMix2x</th>\\n      <td>10</td>\\n      <td>84.0</td>\\n      <td>84.0</td>\\n      <td>168.0</td>\\n      <td>168.0</td>\\n      <td>84.0</td>\\n      <td>84.0</td>\\n    </tr>\\n  </tbody>\\n</table>'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Need to create html output into a letter sized\n",
    "PCR.to_html()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'<table border=\"1\" class=\"dataframe\">\\n  <thead>\\n    <tr style=\"text-align: right;\">\\n      <th>gene</th>\\n      <th>Tm (°C)</th>\\n      <th>id#</th>\\n    </tr>\\n  </thead>\\n  <tbody>\\n    <tr>\\n      <th>tjp2a</th>\\n      <td>72</td>\\n      <td>14</td>\\n    </tr>\\n    <tr>\\n      <th>tjp2b</th>\\n      <td>64</td>\\n      <td>14</td>\\n    </tr>\\n    <tr>\\n      <th>cadm1a</th>\\n      <td>64</td>\\n      <td>16, 17</td>\\n    </tr>\\n    <tr>\\n      <th>cadm1b</th>\\n      <td>67</td>\\n      <td>16, 17</td>\\n    </tr>\\n    <tr>\\n      <th>nrxn3aL</th>\\n      <td>72</td>\\n      <td>17</td>\\n    </tr>\\n    <tr>\\n      <th>nrxn3bL</th>\\n      <td>72</td>\\n      <td>17</td>\\n    </tr>\\n  </tbody>\\n</table>'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "IDs.to_html()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 9 10 11 12 13 14 15]\n"
     ]
    }
   ],
   "source": [
    "a=np.arange(9,16)\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "np.arange?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
